
library(Seurat)
library('dplyr')
library(GSA)
library(fgsea)
library("clustifyr")
library(clusterProfiler)
library(pathview)
library(CellChat)
library(Nebulosa)
library(tidyverse)
library(Nebulosa)
library(ggpointdensity)


#MDS_NBM_all.RDS is generated by integration of scRNA-seq files generated as described below
#_NBM1
NBM1_niche_immune <- CreateSeuratObject(counts = Read10X(data.dir = "~/single_cell_sequencing/run2/NBM-1_Niche_Immune/outs/filtered_feature_bc_matrix/") , min.cells = 3 , min.features = 200 , project = 'NBM1_niche_immune')
NBM1_CD34_Myeloid <- CreateSeuratObject(counts = Read10X(data.dir = "~/single_cell_sequencing/run2/NBM-1_CD34_Myeloid/outs/filtered_feature_bc_matrix/") , min.cells = 3 , min.features = 200 , project = 'NBM1_CD34_Myeloid')



NBM1_niche_immune[["percent.mt"]] <- PercentageFeatureSet(NBM1_niche_immune, pattern = "^MT-");
NBM1_niche_immune[["nRatio"]]=NBM1_niche_immune$nCount_RNA/NBM1_niche_immune$nFeature_RNA;    
VlnPlot(NBM1_niche_immune, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3);
CombinePlots(plots = list(FeatureScatter(NBM1_niche_immune, feature1 = "nCount_RNA", feature2 = "percent.mt"), FeatureScatter(NBM1_niche_immune, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")));
plot(density(NBM1_niche_immune$nRatio,to = 8));
NBM1_niche_immune<-subset(NBM1_niche_immune,subset = nRatio <=3.5 & percent.mt < 5);
VlnPlot(NBM1_niche_immune, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3);
CombinePlots(plots = list(FeatureScatter(NBM1_niche_immune, feature1 = "nCount_RNA", feature2 = "percent.mt"), FeatureScatter(NBM1_niche_immune, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")));
plot(density(NBM1_niche_immune$nRatio,to = 8));

NBM1_CD34_Myeloid[["percent.mt"]] <- PercentageFeatureSet(NBM1_CD34_Myeloid, pattern = "^MT-");
NBM1_CD34_Myeloid[["nRatio"]]=NBM1_CD34_Myeloid$nCount_RNA/NBM1_CD34_Myeloid$nFeature_RNA;    
VlnPlot(NBM1_CD34_Myeloid, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3);
CombinePlots(plots = list(FeatureScatter(NBM1_CD34_Myeloid, feature1 = "nCount_RNA", feature2 = "percent.mt"), FeatureScatter(NBM1_CD34_Myeloid, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")));
plot(density(NBM1_CD34_Myeloid$nRatio,to = 8));
NBM1_CD34_Myeloid<-subset(NBM1_CD34_Myeloid,subset = nRatio <=5 & percent.mt < 5);
VlnPlot(NBM1_CD34_Myeloid, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3);
CombinePlots(plots = list(FeatureScatter(NBM1_CD34_Myeloid, feature1 = "nCount_RNA", feature2 = "percent.mt"), FeatureScatter(NBM1_CD34_Myeloid, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")));
plot(density(NBM1_CD34_Myeloid$nRatio,to = 8));


NBM1_merge<-merge(NBM1_niche_immune, y = NBM1_CD34_Myeloid, add.cell.ids = c("Immune_niche","CD34_Myeloid"), project = "NBM1")

NBM1_merge <- FindVariableFeatures(NBM1_merge, selection.method = "vst", nfeatures = 5000);
NBM1_merge <- NormalizeData(NBM1_merge, normalization.method = "LogNormalize", scale.factor = 10000); 

NBM1_merge[["Group"]]<-"NBM"
NBM1_merge[["Sample"]]<-"NBM1"
##################
#Continue the same procedual for each samples individually (NBM1-4, MDS1-5)
#Integratation NBM_all
#Integration of NBM


NBM_list<-list(NBM1_merge,NBM2_merge,NBM3_merge,NBM4_merge,NBM5_merge)


features_NBM <- SelectIntegrationFeatures(object.list = NBM_list) #Integratation Ctrl
NBM_anchors <- FindIntegrationAnchors(object.list = NBM_list, anchor.features = features_NBM, dims=1:30)
NBM_all <- IntegrateData(anchorset = NBM_anchors,dims=1:30)

NBM_all[["Group"]]<-"NBM"

#Integration of MDS



MDS_list<-list(MDS1_merge,MDS2_merge,MDS3_merge,MDS4_merge,MDS5_merge)


features_MDS <- SelectIntegrationFeatures(object.list = MDS_list) #Integratation Ctrl
MDS_anchors <- FindIntegrationAnchors(object.list = MDS_list, anchor.features = features_MDS, dims=1:30)
MDS_all <- IntegrateData(anchorset = MDS_anchors,dims=1:30)

MDS_all[["Group"]]<-"MDS"

#Integration of MDS and NBM


NBM_MDS_list<-list(NBM_all,MDS_all)


features_NBM_MDS <- SelectIntegrationFeatures(object.list = MDS_list) #Integratation Ctrl
NBM_MDS_anchors <- FindIntegrationAnchors(object.list = NBM_MDS_list, anchor.features = features_NBM_MDS, dims=1:30)
MDS_NBM_all <- IntegrateData(anchorset = NBM_MDS_anchors,dims=1:30)

MDS_NBM_all[["Group"]]<-"MDS"



#Clussification

MDS_NBM_all <- ScaleData(MDS_NBM_all,verbose = FALSE);
MDS_NBM_all <- RunPCA(MDS_NBM_all, npcs = 30,verbose = FALSE);

nDims=20
MDS_NBM_all <- FindNeighbors(MDS_NBM_all, reduction = "pca", dims = 1:30)
MDS_NBM_all <- FindClusters(MDS_NBM_all, resolution = 0.4);
MDS_NBM_all <- RunUMAP(MDS_NBM_all,dims = 1:nDims);
Idents(MDS_NBM_all)<-"seurat_clusters"
DimPlot(MDS_NBM_all, reduction = "umap", label=T,split.by="Group",pt.size = 0.25)

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes


MDS_NBM_all <- CellCycleScoring(MDS_NBM_all, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

MDS_NBM_all$Proliferation_score <- MDS_NBM_all$S.Score - MDS_NBM_all$G2M.Score


MDS_NBM_all <- ScaleData(MDS_NBM_all,vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(MDS_NBM_all));
MDS_NBM_all <- RunPCA(MDS_NBM_all, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_all)
nDims=15
MDS_NBM_all <- FindNeighbors(MDS_NBM_all, reduction = "pca", dims = 1:30)
MDS_NBM_all <- FindClusters(MDS_NBM_all, resolution = 0.4);
MDS_NBM_all <- RunUMAP(MDS_NBM_all,dims = 1:nDims);
Idents(MDS_NBM_all)<-"seurat_clusters"
DimPlot(MDS_NBM_all, reduction = "umap", label=T,split.by="Sample",pt.size = 0.25)


saveRDS(MDS_NBM_all,"scRNAseq_dataset/RDSfile/chenMDS/MDS single cell/MDS_NBM_all.RDS")
######################################3



MDS_NBM_all<-readRDS("scRNAseq_dataset/RDSfile/chenMDS/MDS single cell/MDS_NBM_all.RDS")


#MDS_NBM_all_mt<-readRDS("~/single_cell_sequencing/MDS single cell//MDS_NBM_all_clustify_mutant_WT.RDS")


hallmakers<-GSA.read.gmt("~/single_cell_sequencing/GSEA/hallmakers_genesymble.gmt")
C2<-GSA.read.gmt("~/single_cell_sequencing/GSEA/C2.gmt")
C5<-GSA.read.gmt("~/single_cell_sequencing/GSEA/C5_BP.gmt")

hallmakers_set<-hallmakers$genesets
names(hallmakers_set)<-hallmakers$geneset.names

C2_set<-C2$genesets
names(C2_set)<-C2$geneset.names

C5_set<-C5$genesets
names(C5_set)<-C5$geneset.names


mouse_human_genes = read.csv("http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt",sep="\t")
gene_list<-c("Cdk6","H2-K1")
convert_mouse_to_human <- function(gene_list){
  
output = c()
  
for(gene in gene_list){
    class_key = (mouse_human_genes %>% filter(Symbol == gene & Common.Organism.Name=="mouse, laboratory"))[['DB.Class.Key']]
    if(!identical(class_key, integer(0)) ){
      human_genes = (mouse_human_genes %>% filter(DB.Class.Key == class_key & Common.Organism.Name=="human"))[,"Symbol"]
      for(human_gene in human_genes){
        output = append(output,human_gene)
      }
    }
  }
  
  print (output)
}

convert_mouse_to_human("H2-Eb1")
HSC_output<-read_xlsx("GDCdata/low_output_HSC_signature.xlsx") #low_output_HSC_signature.xlsx is downloaded from Rodriguez-Fraticelli et al, DOI:10.1038/s41586-020-2503-6

low_output<-convert_mouse_to_human(HSC_output$`Low-output`[1:200])
low_output<-unique(low_output)
high_output<-convert_mouse_to_human(HSC_output$`High-output`[1:200])
high_output<-unique(high_output)

Multilineage<-convert_mouse_to_human(HSC_output$Multilineage[1:200])
Multilineage<-unique(Multilineage)
MK_bias<-convert_mouse_to_human(HSC_output$Mk_bias[1:200])
MK_bias<-unique(MK_bias)

HSC_output_set<-list(high_output,low_output,Multilineage,MK_bias)
names(HSC_output_set)<-c("high_output","low_output","Multilineage","MK_bias")


densityplot<-function(dataset=dataset,reduction=NULL,assay="RNA",features=features,split.by=NULL){
  DefaultAssay(dataset)<-assay
  density_plot<-plot_density(dataset, features,reduction=reduction)
  density_plot$data<-cbind(density_plot$data,dataset@meta.data[split.by])
  if(is.null(split.by)){density_plot}else{density_plot+facet_wrap(.~Group)}
}


run_pseudobulkDE_new <- function(data,
                                 contrasts,
                                 design,
                                 sample_col,
                                 condition_col,
                                 cluster_col,
                                 extr_col = NULL, # + Extra
                                 batch_correction=F,
                                 do_prefiltering = TRUE) {
  
  
  # Pull the necessary columns from the metadata table
  meta_data<-data@meta.data
  anno_raw <- meta_data[, c(sample_col, condition_col, extr_col)]
  
  # Pull names from the additional column(s)
  extr_name <- colnames(anno_raw)[3:ncol(anno_raw)]
  # Replace forbidden chrs and add a chr in columns if values start with int
  anno_raw <- apply(anno_raw, 2, function(x) {
    x <- ifelse(grepl('^[0-9]', x), paste0('x', x), x)
    gsub('-|_', '.', x)
  })
  
  
  # Create cell-level annoation dataframe with columns as factors
  anno_cell <- data.frame(
    samples          = anno_raw[, sample_col],
    condition        = anno_raw[, condition_col],
    Extra            = anno_raw[, extr_col],
    row.names        = rownames(meta_data),
    stringsAsFactors = TRUE
  )
  # Add column name(s) to the additional column(s)
  
  # Add factor column that is the conjunction of samples and conditions
  anno_cell$sc <- factor(paste0(anno_cell$samples, '.', anno_cell$condition))
  # split the contrast into a strings
  contrast <- strsplit(contrasts, ',')[[1]]
  
  # Turn the design parameter into a formula object
  design <- formula(design)
  
  # Create groups that will be used to aggregate the expr. in the sparse matrix
  
  agg_g <- S4Vectors::DataFrame(anno_cell[, 'sc'])
  
  # Aggregate the RNA counts by summation; making a sample-level count-matrix
  data[["Sample1"]]<-paste(data[[sample_col]][,1],data[[condition_col]][,1])
  
  agg_m_seurat <- AggregateExpression(data, assays = "RNA", return.seurat = T, group.by = "Sample1")
  
  agg_m<-agg_m_seurat@assays$RNA$counts
  colnames(agg_m) <- as.character(sapply(colnames(agg_m), function(x) {
    x <- ifelse(grepl('^[0-9]', x), paste0('x', x), x)
    x<-gsub('-|_', '.', x)
    gsub(' ', '.', x)
  }))
  
  # Simply transpose agg.m
  
  # Create the intial dataframe used by DESeq2 for determining the conditions
  anno_con <- unique(anno_cell[,-c(1)])
  
  
  # Subset the condition annotation based on the condition being present in the contrast of interest
  anno_con <- anno_con[anno_con$condition %in% contrast, ]
  
  
  # Create the rownames of the condition annotation by pasting samples and condition
  rownames(anno_con) <- anno_con$sc
  
  # change the name of the agg.m object to adhere to the rest of the script
  counts_obj <- agg_m
  
  
  # Subset the full gene-count matrix based on the samples in the condition annotation
  count_name <- colnames(counts_obj)
  anno_name <- rownames(anno_con)
  counts_obj <- counts_obj[, which(count_name %in% anno_name)]
  
  # Reorder the condition annotation based on the column names in counts_obj
  anno_con <- anno_con[match(colnames(counts_obj), anno_name), ]
  
  # Batch effect correction using RUSeq
  dds <- DESeq2::DESeqDataSetFromMatrix(counts_obj, colData = anno_con,
                                        design = design)
  
  
  # Remove outliers, i.e. genes only expressed in a few samples
  if (do_prefiltering) {
    
    message('Info: perform pre-filtering')
    
    con_id <- anno_con[, contrast[1]] == contrast[2]
    
    fragments <- DESeq2::fpm(dds, robust = TRUE)
    
    keep_g1 <- rowSums(fragments[, con_id] >= 2) >= round(sum(con_id) / 2)
    keep_g2 <- rowSums(fragments[, !con_id] >= 2) >= round(sum(!con_id) / 2)
    
    keep_tot <- keep_g1 | keep_g2
    
    dds <- DESeq2::DESeq(dds[keep_tot, ], parallel = FALSE)
    
  } else {
    dds <- DESeq2::DESeq(dds, parallel = FALSE, minReplicatesForReplace = Inf)
  }
  
  # Compute normalized counts
  norm_counts <- as.matrix(DESeq2::counts(dds, normalized = TRUE))
  
  # Generate result table and compute shrunken log fold change
  res <- DESeq2::results(dds, contrast = contrast)
  res_lfc <- DESeq2::lfcShrink(dds, contrast = contrast, type = 'normal',
                               quiet = TRUE)
  res_lfc_ashr <- DESeq2::lfcShrink(dds, contrast = contrast, type = 'ashr',
                                    quiet = TRUE)
  
  # Perpare all the output in a list
  output_l <- list(
    dds          = dds,
    norm_counts  = norm_counts,
    res          = res,
    res_lfc      = res_lfc,
    res_lfc_ashr = res_lfc_ashr
  )
  # Return the DE analyis output list
  Output<-data.frame(cbind(norm_counts,res_lfc_ashr))
  return(Output)
}

#preprocess
###########################

#barcode mutant cells
mutation_code_file<-list.files("~/single_cell_sequencing/MDS single cell/Mutation_state",full.names = T)
mutation_code_name<-list.files("~/single_cell_sequencing/MDS single cell/Mutation_state",full.names = F)
mutation_code<-lapply(mutation_code_file,function(x){
  read.table(x,sep="\t",header=T)
})
mutation_code_name<-str_split(mutation_code_name,pattern="\\.",simplify = T)[,1]


names(mutation_code)<-mutation_code_name

mutation_code<-lapply(mutation_code,function(x){
  x%>%mutate(Start=as.character(Start))%>%mutate(Gene=case_when(str_detect(x$Start,"1974")~"SF3B1",str_detect(x$Start,"2523")~"DNMT31",str_detect(x$Start,"1052")~"TET2"
  ))%>%mutate(State=case_when(.$altCount>=1~"Mutant",.$altCount==0 & .$refCount>=1~"WT"))
})





for (i in seq(1:length(mutation_code))){
  mutation_code[[i]]$cellBarcodes<-paste0(names(mutation_code)[i],"_",mutation_code[[i]]$cellBarcodes,"_1")
  mutation_code[[i]]$Sample<-names(mutation_code)[i]
}

mutation_code_merge<-mutation_code[[1]]

for (i in 2:length(mutation_code)){
  mutation_code_merge<-rbind(mutation_code_merge,mutation_code[[i]])
}

mutation_code_merge_1<-mutation_code_merge%>%dplyr::select(Start,Sample,cellBarcodes,State)%>%dcast(cellBarcodes+Sample~Start)
mutation_code_merge_1$mutant_state<-apply(mutation_code_merge_1, 1, function(x){
  ifelse("Mutant" %in% x,"Mutant",NA)})
mutation_code_merge_1%>%filter(mutant_state=="Mutant")%>%group_by(Sample)%>%summarize(n=n())

mutation_code_remerge<-mutation_code_merge_1

names(mutation_code_remerge)

mutation_code_TET2<-data.frame(bar_code=mutation_code_remerge$cellBarcodes, TET2=NA)
mutation_code_TET2$TET2<-apply(mutation_code_remerge[c(3:7)],1,function(x){
  a<-ifelse("Mutant" %in% x,"Mutant",NA)
  a<-ifelse("WT" %in% x & !"Mutant" %in% x, "WT", a)
  return(a)
  })

mutation_code_SF3B1<-data.frame(bar_code=mutation_code_remerge$cellBarcodes, SF3B1=NA)
mutation_code_SF3B1$SF3B1<-apply(mutation_code_remerge[c(8:9)],1,function(x){
  a<-ifelse("Mutant" %in% x,"Mutant",NA)
  a<-ifelse("WT" %in% x & !"Mutant" %in% x, "WT", a)
  return(a)
})


mutation_code_DNMT31<-data.frame(bar_code=mutation_code_remerge$cellBarcodes, DNMT31=NA)
mutation_code_DNMT31$DNMT31<-apply(mutation_code_remerge[c(10)],1,function(x){
  a<-ifelse("Mutant" %in% x,"Mutant",NA)
  a<-ifelse("WT" %in% x & !"Mutant" %in% x, "WT", a)
  return(a)
})


mutation_code_merge_final<-mutation_code_SF3B1%>%full_join(mutation_code_TET2)%>%full_join(mutation_code_DNMT31)


mutation_code_merge_final$Mutant_state<-apply(mutation_code_merge_final[-1],1,function(x){
  a<-ifelse("Mutant" %in% x, "Mutant",NA)
  return(a)
})

mutation_code_merge_final$Mutant_state<-ifelse(is.na(mutation_code_merge_final$Mutant_state) & mutation_code_merge_final$SF3B1 =="WT", "WT", mutation_code_merge_final$Mutant_state)


length(which(mutation_code_merge_final$Mutant_state=="WT"))

mutation_code_merge_final$bar_code<-tolower(mutation_code_merge_final$bar_code)



MDS_NBM_all_metadata<-MDS_NBM_all@meta.data

bar_code<-str_split(row.names(MDS_NBM_all_metadata),"_",simplify =T)[,3]

bar_code<-paste0(MDS_NBM_all_metadata$orig.ident,"_",bar_code,"_1")

MDS_NBM_all_metadata["bar_code"]<-bar_code

metadata_MDS124<-MDS_NBM_all_metadata[which(MDS_NBM_all_metadata$orig.ident=="MDS_124"),]
MDS_NBM_all_metadata$bar_code[which(MDS_NBM_all_metadata$orig.ident=="MDS_124")]<-paste0("MDS_124_",str_split(row.names(metadata_MDS124),"_",simplify = T)[,1],"_1")
  
head(MDS_NBM_all_metadata[which(MDS_NBM_all_metadata$orig.ident=="MDS_124"),])
MDS_NBM_all_metadata$bar_code<-tolower(MDS_NBM_all_metadata$bar_code)


MDS_NBM_all_metadata_mt<-MDS_NBM_all_metadata%>%left_join(mutation_code_merge_final)
row.names(MDS_NBM_all_metadata_mt)<-row.names(MDS_NBM_all_metadata)

MDS_NBM_all_metadata_mt[which(MDS_NBM_all_metadata_mt$orig.ident=="MDS_10209_niche_immune"),]%>%filter(Mutant_state=="Mutant")




View(MDS_NBM_all_metadata_mt)

MDS_NBM_all_metadata_mt%>%filter(Mutant_state=="Mutant")%>%group_by(orig.ident)%>%summarize(n=n())
head(MDS_NBM_all_metadata_mt%>%filter(Mutant_state=="Mutant"))
length(which(MDS_NBM_all_metadata_mt$Mutant_state=="Mutant"))


MDS_NBM_all_mt<-AddMetaData(MDS_NBM_all,metadata=MDS_NBM_all_metadata_mt)
head(MDS_NBM_all_mt@meta.data)

MDS_NBM_all_mt$SF3B1<-ifelse(MDS_NBM_all_mt$SF3B1=="Mutant","Mutant", "Unknown")
MDS_NBM_all_mt$TET2<-ifelse(MDS_NBM_all_mt$TET2=="Mutant","Mutant", "Unknown")
MDS_NBM_all_mt$DNMT31<-ifelse(MDS_NBM_all_mt$DNMT31=="Mutant","Mutant", "Unknown")

MDS_NBM_all_mt[["Mutation"]]<-"Unknown"
MDS_NBM_all_mt[["Mutation"]]<-ifelse(MDS_NBM_all_mt$SF3B1=="Mutant","Mutant",MDS_NBM_all_mt$Mutation)
MDS_NBM_all_mt[["Mutation"]]<-ifelse(MDS_NBM_all_mt$TET2=="Mutant","Mutant",MDS_NBM_all_mt$Mutation)
MDS_NBM_all_mt[["Mutation"]]<-ifelse(MDS_NBM_all_mt$DNMT31=="Mutant","Mutant",MDS_NBM_all_mt$Mutation)

DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"), label=T,split.by = "Sample",reduction = "umap", cells.highlight = names(MDS_NBM_all_mt$TET2[MDS_NBM_all_mt$TET2=="Mutant"]), cols.highlight = "tomato4",ncol=2,repel = T)

MT_ratio<-subset(MDS_NBM_all_mt,subset=Group=="MDS")@meta.data%>%group_by(Mutation,type_re,Sample)%>%summarize(n=n())%>%group_by(type_re,Sample)%>%
  summarize(Mutation=Mutation,perc=n/sum(n)*100)

#Note that type_re was defined in the code of cell annotation in the lower part of the script
MT_ratio$type_re<-factor(MT_ratio$type_re,levels=c("BMSCs", "ECs", "HSC/MPPs","LMPPs","CLPs","MEPs","GMPs","Monoblasts",
                                                                     "CD14+ Mono","CD16+ Mono", "DCs","Erythroid progenitors","Mks",
                                                                     "CD4","CD8", "NK", "B", "Plasma cells"))
MT_ratio%>%complete(Mutation,fill = list(perc = 0))%>%filter(Mutation=="Mutant")%>%ggplot(aes(perc+0.01,type_re))+geom_col(fill="tomato4")+theme_bw()+labs(y="",x="Size")

MT_ratio%>%complete(Mutation,fill = list(perc = 0))%>%filter(Mutation=="Mutant")%>%filter(type_re%in%c("HSC/MPPs","LMPPs","CLPs","MEPs","GMPs"))%>%
  ggplot(aes(type_re,perc))+geom_boxplot(outlier.size = -1)+geom_jitter()+theme_bw()+labs(y="",x="Size")


Idents(MDS_NBM_all_mt)<-"seurat_clusters"


#By RaCHseq

RaCH_list<-list.files("scRNAseq data transfer Lanpeng Chen/LRMDS/Mutant cell barcodes/RaCHseq",full.names = T)
RaCH_name<-list.files("scRNAseq data transfer Lanpeng Chen/LRMDS/Mutant cell barcodes/RaCHseq")
names(RaCH_list)<-RaCH_name

RaCH_list<-lapply(RaCH_list, function(x){
  x=data.frame(read.table(x))})

for (i in 1:length(RaCH_list)){
  RaCH_list[[i]]["bar_code"]<- tolower(paste0(sub("_pos.txt", replacement="",RaCH_name)[i],"_",RaCH_list[[i]]$V1,"_1"))
  RaCH_list[[i]]<-RaCH_list[[i]]["bar_code"]
}


RaCH_df<-unlist(RaCH_list)

MDS_NBM_all_mt[["RaCH"]]<-"Unknown"


MDS_NBM_all_mt$RaCH<-ifelse(MDS_NBM_all_mt$bar_code %in% RaCH_df, "SF3B1m", "Unknown")

DimPlot(subset(MDS_NBM_all_mt,subset=Group=="NBM"),
        cells.highlight = names(MDS_NBM_all_mt$RaCH[MDS_NBM_all_mt$RaCH=="SF3B1m"]),raster = F)

DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"), label=T,split.by = "Group",reduction = "umap", cells.highlight = names(MDS_NBM_all_mt$RaCH[MDS_NBM_all_mt$RaCH=="SF3B1m"]), cols.highlight = "tomato4",repel = T)

# visualization and analysis of mutant cells
MDS_umap<-DimPlot(subset(MDS_NBM_all_mt,cells=names(MDS_NBM_all_mt$Group[MDS_NBM_all_mt$Group=="MDS"])))
MDS_umap<-MDS_umap$data
MDS_umap_MT<-DimPlot(subset(MDS_NBM_all_mt,cells=names(MDS_NBM_all_mt$Mutation[MDS_NBM_all_mt$Mutation=="Mutant"])))$data
MDS_umap%>%ggplot(aes(MDS_umap$UMAP_1,MDS_umap$UMAP_2))+geom_point(color="grey80")+geom_point(data=MDS_umap_MT,mapping=aes(UMAP_1,UMAP_2),color="tomato4",size=0.5)+theme_classic()

MT_proportion<-MDS_NBM_all_mt@meta.data%>%filter(Group!="NBM")#%>%filter(Sample!="MDS_337")

MT_proportion_SF3B1<-MT_proportion%>%group_by(type_re,SF3B1,Sample)%>%summarize(n=n())%>%
  group_by(Sample)%>%summarize(type_re,SF3B1=SF3B1,perc_SF3B1=n/sum(n))%>%complete(type_re,SF3B1,fill=list(perc_SF3B1=0))%>%
  filter(SF3B1=="Mutant")

MT_proportion_TET2<-MT_proportion%>%group_by(type_re,TET2,Sample)%>%summarize(n=n())%>%
  group_by(Sample)%>%summarize(type_re,TET2=TET2,perc_TET2=n/sum(n))%>%complete(type_re,TET2,fill=list(perc_TET2=0))%>%
  filter(TET2=="Mutant")

MT_proportion_DNMT31<-MT_proportion%>%group_by(type_re,DNMT31,Sample)%>%summarize(n=n())%>%
  group_by(Sample)%>%summarize(type_re,DNMT31=DNMT31,perc_DNMT31=n/sum(n))%>%complete(type_re,DNMT31,fill=list(perc_DNMT31=0))%>%
  filter(DNMT31=="Mutant")

MT_proportion_all<-MT_proportion_DNMT31%>%full_join(MT_proportion_TET2,by=c("Sample","type_re"))%>%
  full_join(MT_proportion_SF3B1,by=c("Sample","type_re"))

MT_proportion_all<-apply(MT_proportion_all,c(1,2),function(x){
  x<-ifelse(is.na(x),0,x)
x})
MT_proportion_all<-data.frame(MT_proportion_all)

MT_proportion_all<-MT_proportion_all%>%melt(id.vars=c("Sample","type_re"),
                                            measure.vars=c("perc_DNMT31","perc_TET2","perc_SF3B1"),
                                            value.name="Perc",variable.name="Mutation")

MT_proportion_all$type_re<-factor(MT_proportion_all$type_re,levels=c("BMSCs", "ECs", "HSC/MPPs","LMPPs","CLPs","MEPs","GMPs","Monoblasts",
                                                                     "CD14+ Mono","CD16+ Mono", "DCs","Erythroid progenitors","Mks",
                                                                     "CD4","CD8", "NK", "B", "Plasma cells"))
MT_proportion_all%>%filter(type_re%in%c("HSC/MPPs","LMPPs","CLPs","MEPs","GMPs"))%>%ggplot(aes(type_re,as.numeric(Perc)*100,fill=Mutation))+geom_bar(stat="identity",position="stack")+facet_wrap(.~Sample,ncol=1,scales = "free_y")+
  theme_classic()+theme(axis.text.x=element_text(angle=45,hjust=0.8,vjust=0.9), strip.background = element_blank())+
  scale_fill_manual(values=c("grey80","steelblue4","tomato4"))+labs(y="Percentage")

MT_proportion_all%>%filter(type_re%in%c("HSC/MPPs","LMPPs","CLPs","MEPs","GMPs"))%>%ggplot(aes(type_re,as.numeric(Perc)*100,fill=Mutation))




#################################
#Annotation, all cell types
##############

Idents(MDS_NBM_all_mt)<-"seurat_clusters"
MDS_NBM_all_mt$Group<-factor(MDS_NBM_all_mt$Group,levels=c("NBM","LRMDS"))

DefaultAssay(MDS_NBM_all_mt)<-"RNA"
DimPlot(MDS_NBM_all_mt, reduction = "umap", label=T,pt.size = 0.25)
MDS_NBM_all_mt<-FindVariableFeatures(MDS_NBM_all_mt)

MDS_NBM_all_mt <- clustify(
  input = MDS_NBM_all_mt,
  cluster_col = "seurat_clusters",
  ref_mat = cbmc_ref,seurat_out = TRUE) 

Idents(MDS_NBM_all_mt)<-"type"
DimPlot(MDS_NBM_all_mt, reduction = "umap",split.by = "Sample" ,label=T,raster=T)

FeaturePlot(MDS_NBM_all_mt,reduction = "umap", label=F,split.by="Group",
            features=c("CD3D","CD3B","NKG7","NCAM1"),pt.size = 0.25,min.cutoff = 0,order=T,cols = c("grey", "steelblue4"))&NoLegend()&theme(axis.title=element_blank());

FeaturePlot(MDS_NBM_all_mt,reduction = "umap", label=T,split.by="Group",
            features=c("ESAM"),pt.size = 0.25,min.cutoff = 0,order=T,cols = c("grey", "steelblue4"))&NoLegend()&theme(axis.title=element_blank());


FeaturePlot(MDS_NBM_all_mt,reduction = "umap", label=F,split.by="Group",
            features=c("KAT6A"),min.cutoff = 0.25,order=T)&NoLegend()&theme(axis.title=element_blank());

DotPlot(subset(MDS_NBM_all_mt,subset=Group=="NBM"), feature=c("KAT6A"))&
  coord_flip()&theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1),legend.position = "right")

VlnPlot(subset(MDS_NBM_all_mt,subset=Group=="NBM"), feature=c("KAT6A"))&
  theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1),legend.position = "right")

FeaturePlot(subset(MDS_NBM_all_mt,subset=Group=="NBM"),reduction = "umap", label=T,
            features=c("KAT6A"),min.cutoff = 0.25,order=T)&NoLegend()&theme(axis.title=element_blank());



Idents(MDS_NBM_all_mt)<-"seurat_clusters"
DimPlot(MDS_NBM_all_mt, reduction = "umap", split.by = "Sample",label=T,pt.size = 0.25)

MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$seurat_clusters %in% c("5")]<-"BMSCs"
MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$seurat_clusters %in% c("20")]<-"ECs"
MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$seurat_clusters %in% c("9","10")]<-"Plasma cells"
MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$type=="CD34+"]<-"HSPC(CD34+)"
MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$type=="pDCs"]<-"DCs"
MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$seurat_clusters %in% c("1","3")]<-"T cells"
MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$seurat_clusters %in% c("11")]<-"NK"

#MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$type=="NK"]<-"NK and T cells"
#MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$type=="NK"]<-"NK and T cells"

MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$seurat_clusters %in% c("7")]<-"MEPs"
MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$seurat_clusters %in% c("6")]<-"Erythroid progenitors"
MDS_NBM_all_mt@meta.data$type[MDS_NBM_all_mt@meta.data$seurat_clusters %in% c("16")]<-"Mks"


#In order to further annotate HSPCs and Immune cells in the whole atlas, I use  meta.date from the HSPC and immune datasets
MDS_NBM_HSPC_M<-readRDS("~/single_cell_sequencing/MDS single cell/MDS_NBM_HSPC_M.RDS")
MDS_NBM_Immune<-readRDS("~/single_cell_sequencing/MDS single cell/MDS_NBM_Immune.RDS")


HSPC_meta<-MDS_NBM_HSPC_M@meta.data
HSPC_meta$type<-HSPC_meta$HSPC_type
HSPC_meta<-HSPC_meta["type"]



Immune_meta<-MDS_NBM_Immune@meta.data
Immune_meta$type<-Immune_meta$Immune_type
Immune_meta<-Immune_meta["type"]

MDS_NBM_all_mt_meta<-MDS_NBM_all_mt@meta.data
MDS_NBM_all_mt_meta_re<-MDS_NBM_all_mt_meta[!row.names(MDS_NBM_all_mt_meta)%in%c(row.names(HSPC_meta),row.names(Immune_meta)),]

MDS_NBM_all_mt_meta_re<-MDS_NBM_all_mt_meta_re["type"]

MDS_NBM_all_mt_meta_re<-rbind(MDS_NBM_all_mt_meta_re,HSPC_meta,Immune_meta)
MDS_NBM_all_mt_meta_re_1<-as.matrix(MDS_NBM_all_mt_meta_re)
MDS_NBM_all_mt_meta_re_1<-MDS_NBM_all_mt_meta_re_1[row.names(MDS_NBM_all_mt@meta.data),]

MDS_NBM_all_mt[["type_re"]]<-MDS_NBM_all_mt$type
MDS_NBM_all_mt$type_re<-MDS_NBM_all_mt_meta_re_1
Idents(MDS_NBM_all_mt)<-"type_re"
MDS_NBM_all_mt<-subset(MDS_NBM_all_mt,ident="HSPC(CD34+)",invert=T) #remove non-annotated cells (satellite cells)

DimPlot(MDS_NBM_all_mt, split.by = "Group", raster=F,label = T,repel=F)&NoLegend()



MDS_NBM_all_MDS345<-subset(MDS_NBM_all_mt,subset=Sample=="MDS345")

saveRDS(MDS_NBM_all_mt,"~/single_cell_sequencing/MDS single cell//MDS_NBM_all_clustify_mutant_WT.RDS")
MDS_NBM_all_mt<-readRDS("scRNAseq_dataset/RDSfile/chenMDS/MDS single cell/MDS_NBM_all_clustify_mutant_WT.RDS")

summary(factor(MDS_NBM_all_mt$Group))

Idents(MDS_NBM_all_mt)<-"type_re"
DimPlot(MDS_NBM_all_mt, reduction = "umap",split.by = "Group" ,label=T,pt.size = 0.25)
MDS_NBM_all_mt<-subset(MDS_NBM_all_mt,ident="T cells",invert=T)

MDS_NBM_all_mt$type_re<-ifelse(MDS_NBM_all_mt$type_re=="DC","DCs",MDS_NBM_all_mt$type_re)
MDS_NBM_all_mt$type_re<-factor(MDS_NBM_all_mt$type_re,levels=c("BMSCs","ECs","HSC/MPPs","LMPPs","GMPs","Monoblasts","DCs","CD14+ Mono","CD16+ Mono", "MEPs","Erythroid progenitors","Mks",
                                                         "CLPs","NK","CD4","CD8","B","Plasma cells"))
Idents(MDS_NBM_all_mt)<-"type_re"
DimPlot(MDS_NBM_all_mt, reduction = "umap",split.by = "Group" ,label=T,pt.size = 0.25)

MDS_NBM_all_mt$Group<-factor(MDS_NBM_all_mt$Group,levels=c("NBM","MDS"))

DotPlot(MDS_NBM_all_mt,features=c("CD55","ECM1","GDF15","HDGF","IGFBP7","IL17RA","IL18","IL6ST","IRS2","LEPR","NECTIN2","PDZK1IP1",
           "PLGRKT","PLXNA1","PVR","TF","TGFB1","TGFBR1","TMPO","TEML2"),split.by = "Group",cols=c("red","red"))&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))

DotPlot(subset(MDS_NBM_all_mt,subset=Group=="NBM"),features=c("PLCG1"),cols=c("grey","tomato4"))&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))&coord_flip()

VlnPlot(subset(MDS_NBM_all_mt,subset=Group=="NBM"),features=c("PLCG1"),cols=c("grey","tomato4"))&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))&coord_flip()


MDS_NBM_all_mt[["type_re1"]]<-as.character(MDS_NBM_all_mt$type_re)
MDS_NBM_all_mt$type_re1<-ifelse(MDS_NBM_all_mt$type_re1=="CD16+ Mono", "Monocyte",MDS_NBM_all_mt$type_re1)
MDS_NBM_all_mt$type_re1<-ifelse(MDS_NBM_all_mt$type_re1=="CD14+ Mono", "Monocyte",MDS_NBM_all_mt$type_re1)
MDS_NBM_all_mt$type_re1<-factor(MDS_NBM_all_mt$type_re1,levels=c("BMSCs","ECs","HSC/MPPs","LMPPs","GMPs","Monoblasts","DCs","Monocyte", "MEPs","Erythroid progenitors","Mks",
                                                               "CLPs","CD4","CD8","NK","B","Plasma cells"))

Idents(MDS_NBM_all_mt)<-"type_re1"

DotPlot(MDS_NBM_all_mt,features=c("TNFRSF1A","TNFRSF1B","IFNGR1","IFNGR2"),split.by = "Group",cols=c("tomato4","tomato4"))&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))&coord_flip()

DotPlot(MDS_NBM_all_mt,features=c("TNF","IFNG"),split.by = "Group",cols=c("tomato4","tomato4"))&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))&coord_flip()

DotPlot(MDS_NBM_all_mt,features=c("ESAM"),split.by = "Group",cols=c("tomato4","tomato4"))&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))&coord_flip()



DimPlot(subset(MDS_NBM_all_mt,cells=names(MDS_NBM_all_mt$Sample)[MDS_NBM_all_mt$Sample!="MDS_124"]), reduction = "umap", label=T,split.by = "Group",,repel=T,raster=F,
        cols=c("seagreen4","cyan3","azure4","lightsteelblue3","burlywood2","lightsalmon","red3","tomato2","tan1","indianred1","red4","steelblue4","khaki4",
               "brown3","brown4","tan4","azure4","grey"),ncol=1)


DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"), reduction = "umap", label=T,cells.highlight = names(MDS_NBM_all_mt$Mutation[MDS_NBM_all_mt$Mutation=="Mutant"]),
        cols.highlight = "tomato4",pt.size = 0.25,split.by = "Group",ncol=1)



FeaturePlot(MDS_NBM_all_mt,reduction = "umap", label=F,split.by="Group",features=c("CD3D","CD8A","NKG7"),
            pt.size = 0.25,min.cutoff = 0.25,order = T,cols=c("grey","steelblue4"))&theme(axis.title = element_blank());


FeaturePlot(subset(MDS_NBM_all_mt,ident=c("BMSCs","ECs"),invert=T),reduction = "umap", label=T,split.by="Group",features=c("PLCG2"),
            pt.size = 0.25,min.cutoff = 1,order = T)&theme(axis.title = element_blank());


Idents(MDS_NBM_all_mt)<-"type_re"
DefaultAssay(MDS_NBM_all_mt)<-"RNA"

DotPlot(MDS_NBM_all_mt, feature=c("IL15","IL18"),split.by = "Group", assay="RNA", cols=rep("steelblue4",12))&
  coord_flip()&theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))


MDS_NBM_all_markers<-FindAllMarkers(subset(MDS_NBM_all_mt,cells=names(MDS_NBM_all_mt$Group[MDS_NBM_all_mt$Group=="NBM"])),only.pos = T)


MDS_NBM_all_markers_top<-MDS_NBM_all_markers%>%filter(p_val_adj<0.05)%>%group_by(cluster)%>%top_n(n=50,wt=avg_log2FC)

DefaultAssay(MDS_NBM_all_mt)<-"RNA"

genelist<-c("LEPR","NGFR","PRRX1","KITLG","CXCL12","IL7","CDH5","PECAM1","CD34","KIT","FLT3","HLF","MEIS1","CRHBP","LYZ","ELANE","MPO","AZU1","CD14","ITGAM","S100A9","FCGR3A","GATA1","HBD","CA1","GP9","CD3D","CD7","IL7R","CD4","CD8A","NKG7","GNLY","CD19","MS4A1","IGHM","CD38")
genelist<-c("IL1B","TNFRSF1A","IL15","PF4","CCL2","CCL4")

MDS_NBM_all_marker<-DotPlot(MDS_NBM_all_mt,features = unique(MDS_NBM_all_markers_top$gene),split.by="Group")$data
MDS_NBM_all_marker<-DotPlot(MDS_NBM_all_mt,features = genelist,split.by="Group",assay="RNA")$data
MDS_NBM_all_marker$gene<-row.names(MDS_NBM_all_marker)
MDS_NBM_all_marker$Cell_type<-str_split(MDS_NBM_all_marker$id,pattern="_",simplify = T)[,1]
MDS_NBM_all_marker$Group<-str_split(MDS_NBM_all_marker$id,pattern="_",simplify = T)[,2]
MDS_NBM_all_marker$Cell_type<-factor(MDS_NBM_all_marker$Cell_type,levels=c(levels(MDS_NBM_all_mt)))
MDS_NBM_all_marker$gene<-factor(MDS_NBM_all_marker$features.plot,levels=rev(unique(MDS_NBM_all_markers_top$gene)))

MDS_NBM_all_marker$Group<-factor(MDS_NBM_all_marker$Group,levels=c("NBM","LRMDS"))

MDS_NBM_all_marker%>%ggplot(aes(Cell_type,features.plot ,fill=avg.exp.scaled))+
  geom_raster()+scale_fill_gradient2(low="blue",mid="white",high="yellow")+theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))+theme(axis.text.y=element_blank())+facet_wrap(.~Group)+theme(strip.background = element_blank())


MDS_NBM_all_marker["GroupType"]<-paste0(MDS_NBM_all_marker$Cell_type,"_",MDS_NBM_all_marker$Group)
MDS_NBM_all_marker%>%ggplot(aes(Cell_type,features.plot ,fill=avg.exp.scaled))+
  geom_raster()+scale_fill_viridis_c()+theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))+facet_wrap(.~Group)+theme(strip.background = element_blank())

MDS_NBM_all_marker%>%ggplot(aes(GroupType,features.plot ,fill=avg.exp.scaled))+
  geom_raster()+scale_fill_viridis_c()+theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))+theme(strip.background = element_blank())


DotPlot(MDS_NBM_all_mt,features=c("PDGFRA","LEPR","LPL","BGLAP","RUNX1","SPP1","SOX9","ACAN","COL2A1","S100A4","SEMA3C",
                                 "NES","CSPG4","ACTA2"),
        assay="RNA")&
  theme(axis.title = element_blank())&coord_flip()&theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))


NBM_all<-subset(MDS_NBM_all_mt,subset=Group=="NBM")

VlnPlot(subset(NBM_all,idents=c("BMSCs","ECs"),invert=T),features=c("PRKCB"),pt.size = 0)&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))


Plot<-FeaturePlot(subset(NBM_all,ident=c("BMSCs","ECs"),invert=T),reduction = "umap", label=T,features=c("CD34"),
            pt.size = 0.25,min.cutoff = 0.25,order = T)&theme(axis.title = element_blank());

Plot$data%>%ggplot(aes(UMAP_1,UMAP_2,colour = CD34))+geom_pointdensity(size = 0.25) +
  scale_color_viridis_c() + # Use a color-blind friendly palette
  labs(title = "UMAP with Gene Expression Density") +
  theme_minimal()



densityplot(MDS_NBM_all_mt,features="TNF")

plot_density(MDS_NBM_all_mt, features="PLCG2",reduction="umap")

densityplot(MDS_NBM_all_mt,features="PLCG2",split.by = "Group")


densityplot(subset(MDS_NBM_all_mt,subset=Group=="NBM"),features="KAT6A")

#################
#Next, I can analyze specific cell types by reclustering the cells
#################
#BMSCs (Not used in my analysis)
####################
MDS_NBM_BMSC<-subset(MDS_NBM_all_mt,ident="BMSCs") #(Generated using the code in MDS_niche_merge (original code was modified in the new analysis in that script))

MDS_NBM_BMSC<-readRDS("~/single_cell_sequencing/MDS single cell/BMSC_all2.RDS")
DimPlot(MDS_NBM_BMSC, reduction = "umap", label=T,pt.size = 0.25,split.by = "Sample")&NoLegend()

FeaturePlot(MDS_NBM_BMSC,reduction = "umap", label=F,split.by="Group",features=c("NFKBIA","CXCL8","ANXA1"),pt.size = 0.25,min.cutoff = 0.25,order = T)&
  theme(axis.title = element_blank());

FeaturePlot(MDS_NBM_BMSC,reduction = "umap", label=F,split.by="Group",features=c("KITLG","LEPR","ANGPT1"),pt.size = 0.25,min.cutoff = 0.25,order = T)&
  theme(axis.title = element_blank());

NBM_BMSC<-subset(MDS_NBM_BMSC,cells=names(MDS_NBM_BMSC$Group[MDS_NBM_BMSC$Group=="NBM"]))

DefaultAssay(MDS_NBM_BMSC)<-"integrated"
FindVariableFeatures(MDS_NBM_BMSC)
MDS_NBM_BMSC <- ScaleData(MDS_NBM_BMSC,verbose = FALSE);
MDS_NBM_BMSC <- RunPCA(MDS_NBM_BMSC, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_BMSC)
nDims=8
MDS_NBM_BMSC <- FindNeighbors(MDS_NBM_BMSC, reduction = "pca", dims = 1:nDims)
MDS_NBM_BMSC <- FindClusters(MDS_NBM_BMSC, resolution = 0.3);
MDS_NBM_BMSC <- RunUMAP(MDS_NBM_BMSC,dims = 1:nDims);

DimPlot(MDS_NBM_BMSC, reduction = "umap", label=T,pt.size = 0.25,split.by = "Group")&NoLegend()
Idents(MDS_NBM_BMSC)<-"seurat_clusters"
MDS_NBM_BMSC_markers<-FindAllMarkers(MDS_NBM_BMSC,assay="RNA",only.pos = T)

FeaturePlot(MDS_NBM_BMSC,reduction = "umap", label=F,split.by="Group",features=c("LEPR","KITLG","CXCL12"),pt.size = 0.25,min.cutoff = 0.25,order = T)&
  theme(axis.title = element_blank());

VlnPlot(MDS_NBM_BMSC, features=c("LEPR","KITLG","CXCL12","IL7"),
        ,pt.size=0,assay="RNA",split.by = "Group")&theme(axis.title.x = element_blank())

VlnPlot(MDS_NBM_BMSC, features=c("NFKBIA","CD44","CXCL8","IL6"),
        ,pt.size=0,assay="RNA",split.by = "Group")&theme(axis.title.x = element_blank())



MDS_NBM_BMSC[["BMSC_cluster"]]<-MDS_NBM_BMSC$seurat_clusters
MDS_NBM_BMSC[["BMSC_cluster"]]<-ifelse(MDS_NBM_BMSC$seurat_clusters=="3","BMSC-0",MDS_NBM_BMSC$BMSC_cluster)
MDS_NBM_BMSC[["BMSC_cluster"]]<-ifelse(MDS_NBM_BMSC$seurat_clusters %in% c("0"),"BMSC-1",MDS_NBM_BMSC$BMSC_cluster)
MDS_NBM_BMSC[["BMSC_cluster"]]<-ifelse(MDS_NBM_BMSC$seurat_clusters=="1","BMSC-2",MDS_NBM_BMSC$BMSC_cluster)
MDS_NBM_BMSC[["BMSC_cluster"]]<-ifelse(MDS_NBM_BMSC$seurat_clusters=="2","BMSC-3",MDS_NBM_BMSC$BMSC_cluster)

MDS_NBM_BMSC$BMSC_cluster<-factor(MDS_NBM_BMSC$BMSC_cluster,levels=c("BMSC-0","BMSC-1","BMSC-2","BMSC-3"))
Idents(MDS_NBM_BMSC)<-"BMSC_cluster"

DimPlot(MDS_NBM_BMSC,split.by = "Group" ,reduction = "umap", label=F,pt.size = 0.25,repel=F,
        cols=c("grey40","steelblue4","khaki4","tomato4",
               "steelblue1","red3","burlywood3","tomato4","tan4","azure4","royalblue2","steelblue3","steelblue","tomato"))

MDS_NBM_BMSC<-subset


DefaultAssay(MDS_NBM_BMSC)<-"RNA"
FeaturePlot(MDS_NBM_BMSC,reduction = "umap", label=F,split.by="Group",features=c("CXCL12","KITLG","IL7"),
            cols=c("grey80","steelblue4"),pt.size = 0.25,min.cutoff = 0.25,order = T)&theme(axis.title = element_blank());

FeaturePlot(MDS_NBM_BMSC,reduction = "umap", label=F,split.by="Group",features=c("S100A4","S100A6"),
            cols=c("grey80","steelblue4"),pt.size = 0.25,min.cutoff = 0.25,order = T)&theme(axis.title = element_blank());


VlnPlot(MDS_NBM_BMSC, features=c("NFKBIA","ANXA1","CXCL8"),split.by = "Group", cols=c(rep("steelblue4",1),rep("tomato4",1)),pt.size=0,assay="RNA")
VlnPlot(MDS_NBM_BMSC, features=c("IGF1"),split.by = "Group", cols=c(rep("steelblue4",1),rep("tomato4",1)),pt.size=0,assay="RNA")


VlnPlot(MDS_NBM_BMSC, features=c("CXCL12","ANGPT1","KITLG","IL7","NFKBIA","ANXA1","CD44","CXCL8"),ncol=4,
        split.by = "Group",pt.size=0,cols=c(rep("steelblue4",1),rep("tomato4",1)))&theme(axis.title = element_blank())

VlnPlot(MDS_NBM_BMSC, features=c("APOE"),
        split.by = "Group",pt.size=0,cols=c(rep("steelblue4",1),rep("tomato4",1)))&theme(axis.title = element_blank())


VlnPlot(MDS_NBM_BMSC, features=c("CXCL12","KITLG","IL7","IGF2"),
        group.by = "Sample",pt.size=0,cols=c(rep("tomato4",4),rep("steelblue4",4)))&theme(axis.title.x = element_blank())

DotPlot(MDS_NBM_BMSC,feature=c("LEPR"))

BMSC_markers<-FindAllMarkers(MDS_NBM_BMSC,assay="RNA",only.pos = T)

MDS_NBM_BMSC_proportion<-MDS_NBM_BMSC@meta.data%>%group_by(Sample, BMSC_cluster)%>%summarize(n=n())%>%group_by(Sample)%>%
  summarize(BMSC_cluster=BMSC_cluster, perc=n/sum(n))%>%complete(BMSC_cluster,fill = list(n=0,perc=0))

MDS_NBM_BMSC_proportion["Group"]<-ifelse(grepl("MDS", MDS_NBM_BMSC_proportion$Sample),"MDS","NBM")
MDS_NBM_BMSC_proportion$Group<-factor(MDS_NBM_BMSC_proportion$Group,levels=c("NBM","MDS"))


MDS_NBM_BMSC_proportion%>%ggplot(aes(Group, perc,color=Group))+geom_boxplot()+geom_jitter(width=0.1)+theme_classic()+
  facet_wrap(.~BMSC_cluster, scale="free",ncol=2)+theme(strip.background = element_blank(),strip.text =element_text(size=10))+
  scale_color_manual(values=c("steelblue4","tomato4"))+NoLegend()+stat_compare_means(method="t.test")


MDS_NBM_BMSC_proportion%>%ggplot(aes(Sample, perc,fill=BMSC_cluster))+geom_bar(stat="identity")+theme_classic()+
  scale_fill_manual(values=c("grey40","steelblue4","khaki4","tomato4",
                             "steelblue1","red3","burlywood3","tomato4","tan4","azure4","royalblue2","steelblue3","steelblue","tomato"))


MDS_NBM_BMSC_seq<-run_pseudobulkDE(meta_data=MDS_NBM_BMSC@meta.data,expr_mat=MDS_NBM_BMSC@assays$RNA@counts,
                                  contrasts='condition,MDS,NBM',design='~condition',
                                  sample_col="Sample", condition_col="Group",cluster_col="seurat_clusters",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)

MDS_NBM_BMSC_seq["Gene"]<-row.names(MDS_NBM_BMSC_seq)
MDS_NBM_BMSC_seq$pvalue<-ifelse(MDS_NBM_BMSC_seq$pvalue<1e-30,1e-30,MDS_NBM_BMSC_seq$pvalue)
MDS_NBM_BMSC_seq["rank"]<- -sign(MDS_NBM_BMSC_seq$log2FoldChange)*log2(MDS_NBM_BMSC_seq$pvalue)
MDS_NBM_BMSC_seq["Enrichment"]<-ifelse(MDS_NBM_BMSC_seq$rank>0,"MDS","NBM")
MDS_NBM_BMSC_seq["Enrichment"]<-factor(MDS_NBM_BMSC_seq$Enrichment,levels=c("NBM","MDS"))

MDS_NBM_BMSC_seq_sig<-MDS_NBM_BMSC_seq[which(MDS_NBM_BMSC_seq$padj<0.05),]

MDS_NBM_BMSC_seq_top<-MDS_NBM_BMSC_seq_sig%>%filter(abs(log2FoldChange)>0.585)

MDS_NBM_BMSC_seq_sig_lables<-MDS_NBM_BMSC_seq_sig%>%filter(Gene %in% c("BMP4","GADD45A","GADD45B","GDF15","IRF1","CDKN1A","BMP5","RELA","ANGPT1","KITLG","IL7","LEPR","LPL","ANXA1","NFKBIA","TNFAIP8","CD44","CXCL8"))

MDS_NBM_BMSC_seq%>%ggplot(aes(log2FoldChange,-log10(padj)))+geom_point(size=0.1,col="grey")+
  geom_point(data=MDS_NBM_BMSC_seq_top,mapping=aes(log2FoldChange,-log10(padj)),size=0.5,alpha=0.3,col="grey40")+
  geom_vline(xintercept= c(0.5,-0.5),col="grey",type=2)+geom_hline(yintercept = -log10(0.05),col="grey",type=2)+
  geom_text_repel(MDS_NBM_BMSC_seq_sig_lables,mapping=aes(log2FoldChange,-log10(padj),label=Gene,col=Enrichment),max.overlaps = 20)+
  theme_bw()+scale_color_manual(values=c("steelblue4","tomato4"))

rank_BMSC<-MDS_NBM_BMSC_seq$rank
names(rank_BMSC)<-MDS_NBM_BMSC_seq$Gene
rank_BMSC<-rank_BMSC[!is.na(rank_BMSC)]

MDS_NBM_BMSC_Hall <- fgsea(hallmakers_set, rank_BMSC, minSize=15, maxSize = 500, nperm=10000)
head(MDS_NBM_BMSC_Hall[order(padj), ], n=10)
MDS_NBM_BMSC_Hall_sig<-MDS_NBM_BMSC_Hall%>%filter(padj<0.05)%>%arrange(NES)

MDS_NBM_BMSC_Hall_sig[["Enrichment"]]<-ifelse(MDS_NBM_BMSC_Hall_sig$NES<0,"NBM","MDS")
MDS_NBM_BMSC_Hall_sig<-MDS_NBM_BMSC_Hall_sig%>%group_by(Enrichment)%>%top_n(n=10,wt=abs(NES))
MDS_NBM_BMSC_Hall_sig<-MDS_NBM_BMSC_Hall_sig%>%arrange(NES)
MDS_NBM_BMSC_Hall_sig<-MDS_NBM_BMSC_Hall_sig%>%mutate(pathway=factor(pathway, levels=MDS_NBM_BMSC_Hall_sig$pathway))
MDS_NBM_BMSC_Hall_sig%>%ggplot(aes(NES,pathway,fill=Enrichment))+geom_col()+theme_classic()+scale_fill_manual(values=c("tomato4","steelblue4"))+NoLegend()


MDS_NBM_BMSC_score<-AddModuleScore(MDS_NBM_BMSC,features=hallmakers_set["HALLMARK_TNFA_SIGNALING_VIA_NFKB"], name="NFkB_signaling")
MDS_NBM_BMSC_score<-AddModuleScore(MDS_NBM_BMSC_score,features=hallmakers_set["HALLMARK_INFLAMMATORY_RESPONSE"], name="Inflammatory_response")

VlnPlot(MDS_NBM_BMSC_score, features=c("Inflammatory_response1"),
        group.by = "Sample",pt.size=0,cols=c(rep("tomato4",4),rep("steelblue4",4)))&theme(axis.title.x = element_blank())

#########
#Immune cells preprocessing
##############
Ident(MDS_NBM_Immune)<-"type_re"
MDS_NBM_Immune<-subset(MDS_NBM_all_mt, ident=c("T cells", "NK"))
MDS_NBM_Immune$Group<-ifelse(MDS_NBM_Immune$Group=="NBM","NBM","LRMDS")

MDS_NBM_Immune <- ScaleData(MDS_NBM_Immune,verbose = FALSE);
MDS_NBM_Immune <- RunPCA(MDS_NBM_Immune, npcs = 30,verbose = FALSE);

ElbowPlot(MDS_NBM_Immune)
nDims=15
MDS_NBM_Immune <- FindNeighbors(MDS_NBM_Immune, reduction = "pca", dims = 1:nDims)
MDS_NBM_Immune <- FindClusters(MDS_NBM_Immune, resolution = 0.5);
MDS_NBM_Immune <- RunUMAP(MDS_NBM_Immune,dims = 1:nDims);

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

MDS_NBM_Immune <- CellCycleScoring(MDS_NBM_Immune, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

MDS_NBM_Immune$Proliferation_score <- MDS_NBM_Immune$S.Score - MDS_NBM_Immune$G2M.Score
DefaultAssay(MDS_NBM_Immune)<-"integrated"

MDS_NBM_Immune <- ScaleData(MDS_NBM_Immune,vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(MDS_NBM_Immune));
MDS_NBM_Immune <- RunPCA(MDS_NBM_Immune, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_Immune)
nDims=10
MDS_NBM_Immune <- FindNeighbors(MDS_NBM_Immune, reduction = "pca", dims = 1:nDims)
MDS_NBM_Immune <- FindClusters(MDS_NBM_Immune, resolution = 0.6);
MDS_NBM_Immune <- RunUMAP(MDS_NBM_Immune,dims = 1:nDims);

MDS_NBM_Immune$umap_1<-MDS_NBM_Immune@reductions$umap@cell.embeddings[,1]
MDS_NBM_Immune$umap_2<-MDS_NBM_Immune@reductions$umap@cell.embeddings[,2]
######
#Immune cells analysis
##########
MDS_NBM_Immune<-readRDS("scRNAseq_dataset/RDSfile/chenMDS/MDS single cell/MDS_NBM_Immune.RDS")

Idents(MDS_NBM_Immune)<-"seurat_clusters"
DimPlot(MDS_NBM_Immune, reduction = "umap", split.by="Sample", label=T ,pt.size = 0.3)
FeaturePlot(MDS_NBM_Immune, features=c("ITGAE"),order=T,min.cutoff = 0.25,reduction = "umap", split.by="Group", label=F ,pt.size = 0.1)
FeaturePlot(MDS_NBM_Immune, features=c("SELL","CCR7"),order=T,min.cutoff = 0.25,reduction = "umap", split.by="Group", label=T ,pt.size = 0.1)
FeaturePlot(MDS_NBM_Immune, features=c("CD3D","CD4","CD8A","CD8B"),order=T,min.cutoff = 0.25,reduction = "umap", split.by="Group", label=T ,pt.size = 0.1)
FeaturePlot(MDS_NBM_Immune, features=c("CXCR3","TBX21","IL18","IFNG"),order=T,min.cutoff = 0.25,reduction = "umap", split.by="Group", label=F ,pt.size = 0.1)
  #TH1
FeaturePlot(MDS_NBM_Immune, features=c("CCR6","IL6","IL1B","TGFB1"),order=T,min.cutoff = 0.25,reduction = "umap", split.by="Group", label=T ,pt.size = 0.1)
  #TH17
FeaturePlot(MDS_NBM_Immune, features=c("FOXP3","CTLA4"),order=T,min.cutoff = 0.25,reduction = "umap", split.by="Group", label=T ,pt.size = 0.1)
  #Treg "TRGC1","TRDC"
FeaturePlot(MDS_NBM_Immune, features=c("CCR4","CCR8","IL4","GATA3"),order=T,min.cutoff = 0.25,reduction = "umap", split.by="Group", label=T ,pt.size = 0.1)
  #Th2
FeaturePlot(MDS_NBM_Immune, features=c("PRF1","GZMB"),order=T,min.cutoff = 0.25,reduction = "umap", split.by="Group", label=T ,pt.size = 0.1)
  #T cytotoxic FCGR3A 
FeaturePlot(MDS_NBM_Immune, features=c("FCGR3A","NCAM1","CD3D"),order=T,min.cutoff = 0.25,reduction = "umap", split.by="Group", label=F ,pt.size = 0.1)
  #NK

Immune_meta<-MDS_NBM_Immune@meta.data

MDS_NBM_Immune[["Immune_type"]]<-"Unknown"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters=="7")]<-"CD8 Naive"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters=="0")]<-"CD4 Naive"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters %in% c("5"))]<-"Th1/17"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters %in% c("1"))]<-"Th2"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters %in% c("9"))]<-"Treg"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters %in% c("2"))]<-"CD8 EM"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters %in% c("3"))]<-"CD8 TEMRA"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters %in% c("8","10"))]<-"T unknown"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters %in% c("4"))]<-"NK CD56dim"
MDS_NBM_Immune$Immune_type[which(MDS_NBM_Immune$seurat_clusters %in% c("6"))]<-"NK CD56high"

MDS_NBM_Immune$type_re<-"CD4"
MDS_NBM_Immune$type_re[grepl("NK",MDS_NBM_Immune$Immune_type)]<-"NK"
MDS_NBM_Immune$type_re[grepl("CD8",MDS_NBM_Immune$Immune_type)]<-"CD8"

for (i in 1:nrow(MDS_NBM_Immune@meta.data)){
  if (names(MDS_NBM_Immune$Immune_type)[i] %in% names(MDS_NBM_CD8$Immune_type)){
    MDS_NBM_Immune$Immune_type[i]<-MDS_NBM_CD8$Immune_type[which(names(MDS_NBM_CD8$Immune_type)==names(MDS_NBM_Immune$Immune_type)[i])]
  }
}

Idents(MDS_NBM_Immune)<-"Immune_type"
DimPlot(MDS_NBM_Immune, reduction = "umap", split.by="Group", label=T ,pt.size = 0.3)

Immune_marker<-FindAllMarkers(subset(MDS_NBM_Immune,subset=Group=="MDS"),asslay="RNA",only.pos = T)
Immune_marker_top<-Immune_marker%>%group_by(cluster)%>%top_n(n=30,wt=avg_log2FC)%>%ungroup()


#MDS_NBM_Immune<-subset(MDS_NBM_Immune, cells=names(MDS_NBM_Immune$Sample[MDS_NBM_Immune$Sample=="MDS_SF3B1"]),invert=T)
MDS_NBM_Immune[["Group_cluster"]]<-paste0(MDS_NBM_Immune$Immune_type,
                                        "-",MDS_NBM_Immune$Group)
Idents(MDS_NBM_Immune)<-"Immune_type"
DefaultAssay(MDS_NBM_Immune)<-"RNA"
MDS_NBM_Immune$Group<-factor(MDS_NBM_Immune$Group,levels=c("NBM","LRMDS"))
DimPlot(MDS_NBM_Immune, reduction = "umap", label=T,pt.size = 0.25,split.by = "Group",,repel=T,raster=FALSE,
        cols=c("seagreen4","cyan3","azure4","red3","burlywood2","lightsteelblue3","lightsalmon","tan1","indianred1","red4","steelblue4","khaki4",
               "brown3","brown4","tomato4","tan4","azure4"))
MDS_NBM_Immune$Group<-factor(MDS_NBM_Immune$Group,levels=c("NBM","MDS"))
FeaturePlot(MDS_NBM_Immune, features=c("CCR6"),order=T,min.cutoff = 0.25,split.by = "Group")
FeaturePlot(MDS_NBM_Immune, features=c("PDCD1","CD274","LAG3","HAVCR2","CTLA4","TOX1"),order=T,min.cutoff = 0.25,split.by = "Group")

FeaturePlot(MDS_NBM_Immune, features=c("CD69","HLA-DRA","TNFRSF9"),order=T,min.cutoff = 0.25,split.by = "Group")

FeaturePlot(MDS_NBM_Immune, features=c("CD8B"),label=T,pt.size=0.25,order=T,min.cutoff = 0.25,split.by = "Group")&theme(axis.title = element_blank())

DotPlot(MDS_NBM_Immune, assay="RNA",features=c("PDCD1","CD274","LAG3","HAVCR2","TOX1"),split.by = "Group",cols=rep("blue",12))
DotPlot(MDS_NBM_Immune, assay="RNA",features=c("HLA-DRA","TNFRSF9"),split.by = "Group",cols=rep("blue",2))
DotPlot(MDS_NBM_Immune, assay="RNA",features=c("NCAM1"),split.by = "Group",cols=rep("blue",2))
DotPlot(MDS_NBM_Immune, assay="RNA",features=c("IFNG","TNF"),split.by = "Group",cols=rep("tomato4",2))&coord_flip()&theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))

Immune_marker<-FindAllMarkers(subset(MDS_NBM_Immune,subset=Group=="MDS"),asslay="RNA",only.pos = T)
write.xlsx(Immune_marker,"single_cell_sequencing/MDS single cell/Summary/T and NK cells/Markers_of_immune_subpopulations_MDS.xlsx")

VlnPlot(MDS_NBM_Immune,features=c("ITGAE"),split.by = "Group",assay="RNA")
Immune_proportion<-MDS_NBM_Immune@meta.data%>%group_by(Sample, Immune_type)%>%summarize(n=n())%>%group_by(Sample)%>%
  summarize(Immune_type=Immune_type, Percentage=n/sum(n)*100)

Immune_proportion["Group"]<-ifelse(grepl("MDS", Immune_proportion$Sample),"LRMDS","NBM")
Immune_proportion$Group<-factor(Immune_proportion$Group,levels=c("NBM","LRMDS"))

P<-Immune_proportion%>%ggplot(aes(Group, Percentage,color=Group))+geom_boxplot(outlier.size = -1)+geom_jitter(width=0.1)+theme_classic()+
  facet_wrap(.~Immune_type, ncol=3,scale="free_x")+theme(strip.background = element_blank(),strip.text =element_text(size=10))+
  scale_color_manual(values=c("steelblue4","tomato4"))

P+stat_compare_means(method = "t.test",label="p.format")



MDS_NBM_Immune_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_Immune@meta.data,expr_mat=MDS_NBM_Immune@assays$RNA@counts,
                                    contrasts='condition,MDS,NBM',design='~condition',
                                    sample_col="Sample", condition_col="Group",cluster_col="Immune_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)
MDS_NBM_Immune_deseq["Gene"]<-row.names(MDS_NBM_Immune_deseq)


MDS_NBM_Immune$type<-ifelse(MDS_NBM_Immune$Immune_type %in% c("CD8 GZMK+", "CD8 GZMB+", "CD8 CCR7+"), "CD8T",MDS_NBM_Immune$type)
MDS_NBM_Immune$type<-ifelse(MDS_NBM_Immune$Immune_type %in% c("CD4 CCR7+", "Th1/17", "Th2","CD4 GZMB+"), "CD4T",MDS_NBM_Immune$type)

Immune_proportion<-MDS_NBM_Immune@meta.data%>%filter(type %in% c("CD4T","CD8T"))%>%group_by(Sample, type)%>%summarize(n=n())%>%group_by(Sample)%>%
  summarize(Immune_type=type, Fraction=n/sum(n))
Immune_proportion["Group"]<-ifelse(grepl("MDS", Immune_proportion$Sample),"MDS","NBM")
Immune_proportion$Group<-factor(Immune_proportion$Group,levels=c("NBM","MDS"))
Immune_proportion<-Immune_proportion%>%dcast(Sample+Group~Immune_type,value.var="Fraction")
Immune_proportion["ratio"]<-Immune_proportion$CD8T/Immune_proportion$CD4T

P<-Immune_proportion%>%ggplot(aes(Group, ratio,color=Group))+geom_boxplot(outlier.size = -1)+geom_jitter(width=0.1)+theme_classic()+
  theme(strip.background = element_blank(),strip.text =element_text(size=10))+
  scale_color_manual(values=c("steelblue4","tomato4"))

P+stat_compare_means(method = "t.test",size=3)



Idents(MDS_NBM_Immune)<-"type"

DotPlot(MDS_NBM_Immune,assay="RNA",features=c("TNFRSF1A"),split.by = "Group",cols=rep("blue",2))

plot_density(MDS_NBM_Immune, "CD3D")
#CD8+


MDS_NBM_CD8<-subset(MDS_NBM_Immune, ident=c("CD8 TEMRA", "CD8 EM", "CD8 CCR7+"))




DefaultAssay(MDS_NBM_CD8)<-"integrated"

FindVariableFeatures(MDS_NBM_CD8)
MDS_NBM_CD8 <- ScaleData(MDS_NBM_CD8,features = rownames(MDS_NBM_CD8));
MDS_marker<-AverageExpression(subset(MDS_NBM_CD8,Group=="MDS"))$RNA
MDS_marker["GZMB",]
MDS_marker["IFNG",]




DefaultAssay(MDS_NBM_CD8)<-"integrated"
FindVariableFeatures(MDS_NBM_CD8)
MDS_NBM_CD8 <- RunPCA(MDS_NBM_CD8, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_CD8)
set.seed(50)
nDims=8
MDS_NBM_CD8 <- FindNeighbors(MDS_NBM_CD8, reduction = "pca", dims = 1:nDims)
MDS_NBM_CD8 <- FindClusters(MDS_NBM_CD8, resolution = 0.5);
MDS_NBM_CD8 <- RunUMAP(MDS_NBM_CD8,dims = 1:nDims);
Idents(MDS_NBM_CD8)<-"CD8_type"

Idents(MDS_NBM_CD8)<-"seurat_clusters"

DimPlot(MDS_NBM_CD8, reduction = "umap",repel=T,split.by="Group", ncol=4,label=T ,pt.size = 0.1)

FeaturePlot(MDS_NBM_CD8,reduction = "umap", split.by = "Group", features=c("IFNG"),pt.size=0.2,
            cols=c("grey","tomato4"),min.cutoff=0.25, order=T,ncol=2)&theme(axis.title=element_blank())

MDS_NBM_CD8[["CD8_type"]]<-MDS_NBM_CD8$Immune_type


MDS_NBM_CD8$CD8_type[MDS_NBM_CD8$seurat_clusters%in% c("3","5","6")]<-"IFNG_EM"



#densityplot(MDS_NBM_CD8,features="IFNG",split.by="Group")&facet_wrap(.~Group)
#densityplot(MDS_NBM_CD8,features="ENTPD1",split.by="Group")&facet_wrap(.~Group)


CD8_proportion<-MDS_NBM_CD8@meta.data%>%group_by(Sample, CD8_type)%>%summarize(n=n())%>%group_by(Sample)%>%
  summarize(CD8_type=CD8_type, perc=n/sum(n)*100)
CD8_proportion["Group"]<-ifelse(grepl("MDS", CD8_proportion$Sample),"MDS","NBM")
CD8_proportion$Group<-factor(CD8_proportion$Group,levels=c("NBM","MDS"))
CD8_proportion%>%ggplot(aes(Sample, perc,fill=CD8_type))+geom_bar(stat="identity")+theme_classic()
CD8_proportion%>%ggplot(aes(Group, perc,color=Group))+geom_boxplot(width=0.5)+geom_jitter(width=0.1)+theme_classic()+
  facet_wrap(.~CD8_type,scales="free")+theme(strip.background = element_blank(),strip.text =element_text(size=12))+
  scale_color_manual(values=c("steelblue4","tomato4"))+theme(legend.position = "none")+stat_compare_means(method = "t.test")








CD8_marker<-FindAllMarkers(subset(MDS_NBM_CD8,subset=Group=="MDS"),asslay="RNA",only.pos = T)
CD8_marker_top<-CD8_marker%>%group_by(cluster)%>%top_n(n=30,wt=avg_log2FC)%>%ungroup()

#MDS_NBM_CD8<-subset(MDS_NBM_CD8,cells=names(MDS_NBM_CD8$Sample)[MDS_NBM_CD8$Sample!="MDS_SF3B1"])
MDS_NBM_CD8$Group<-factor(MDS_NBM_CD8$Group, levels=c("NBM","MDS"))
MDS_NBM_CD8$Immune_type<-factor(MDS_NBM_CD8$Immune_type, levels=c("CD8 CCR7+","CD8 GZMK+","CD8 GZMB+"))

Idents(MDS_NBM_CD8)<-"Immune_type"
DimPlot(MDS_NBM_CD8, reduction = "umap",repel=T,split.by="Group", ncol=4,label=T ,pt.size = 0.1,
        cols=c("steelblue4","tomato4","darkgreen"))

MDS_NBM_CD8$Immune_type<-ifelse(MDS_NBM_CD8$seurat_clusters=="3","CD4 GZMB+",MDS_NBM_CD8$Immune_type)

MDS_NBM_CD8[["Immune_type_group"]]<-paste0(MDS_NBM_CD8$Immune_type,"_",MDS_NBM_CD8$Group)
Idents(MDS_NBM_CD8)<-"Immune_type"

VlnPlot(MDS_NBM_CD8, feature=c("CCR7","PRF1","GZMB","GZMK"),split.by = "Group",ncol=1,assay="RNA",pt.size=0, cols=c("steelblue4","tomato4"))&theme(axis.title = element_blank())



gene_list<-c("TNF","IFNG","CCL3","CCL4","TNFSF10","TGFB1","TNFRSF9","CD69","HLA-DRA","CTLA4","LAG3","PDCD1","HAVCR2","GZMB","PRF1","GNLY","FASLG","CD28","IL7R","CCR7","SELL","TCF7")
CD8_markers<- AverageExpression(MDS_NBM_CD8,features=gene_list,assay="RNA")$RNA
CD8_markers<-CD8_markers[,order(colnames(CD8_markers),decreasing = T)]
pheatmap(CD8_markers[,c("NBM-CD8 CCR7+","MDS-CD8 CCR7+","NBM-CD8 GZMK+",
                        "MDS-CD8 GZMK+","NBM-CD8 GZMB+","MDS-CD8 GZMB+")],scale="row",cluster_cols = F,cluster_rows = F,colorRampPalette(c("blue","white","yellow"))(30))

CD8_marker<-FindAllMarkers(subset(MDS_NBM_CD8,subset=Group=="MDS"),asslay="RNA",only.pos = T)

write.xlsx(CD8_marker,"single_cell_sequencing/MDS single cell/Summary/T and NK cells/Markers_of_CD8_subpopulaitons_MDS.xlsx")


CD8_marker_top<-CD8_marker%>%group_by(cluster)%>%top_n(n=30,wt=avg_log2FC)%>%ungroup()
MDS_NBM_CD8[["Group_cluster"]]<-paste0(MDS_NBM_CD8$Immune_type,"-",MDS_NBM_CD8$Immune_type)
MDS_NBM_CD8$Group_cluster<-factor(MDS_NBM_CD8$Group_cluster,levels=c("NBM-CD8 CCR7+","MDS-CD8 CCR7+","NBM-CD8 GZMK+",
                                                                     "MDS-CD8 GZMK+","NBM-CD8 GZMB+","MDS-CD8 GZMB+"))
Idents(MDS_NBM_CD8)<-"Group_cluster"
Idents(MDS_NBM_CD8)<-"Immune_type"

DefaultAssay(MDS_NBM_CD8)<-"RNA"
DotPlot(MDS_NBM_CD8,assay="RNA",features=c("TNF","IFNG","TNFSF10"),cols=c("grey","tomato4"))

DoHeatmap(MDS_NBM_CD8,features = CD8_marker_top$gene)
CD8_marker_top_expression<-AverageExpression(MDS_NBM_CD8,features=CD8_marker_top$gene)$RNA
CD8_marker_top_expression<-CD8_marker_top_expression[,c("NBM-CD8 CCR7+","MDS-CD8 CCR7+","NBM-CD8 GZMK+",
                                                       "MDS-CD8 GZMK+","NBM-CD8 GZMB+","MDS-CD8 GZMB+")]
pheatmap(CD8_marker_top_expression,scale="row",cluster_cols = F,cluster_rows = F)

FeaturePlot(MDS_NBM_CD8,reduction = "umap",  split.by = "Group", features=c("CTLA4","HAVCR2","PDCD1","LAG3", "CD226"),min.cutoff=0.25, sort=T,pt.size=0.15)
FeaturePlot(MDS_NBM_CD8,reduction = "umap",  split.by = "Group", features=c("IFIT1","IFIT2","IFIT3"),min.cutoff=0.25, sort=T)
FeaturePlot(MDS_NBM_CD8,reduction = "umap",  split.by = "Group", features=c("XCL1","XCL2","ITGB1","ZNF683"),min.cutoff=0.25, sort=T) #Tissue resident
FeaturePlot(MDS_NBM_CD8,reduction = "umap", split.by = "Group", features=c("IFNG","TNF","CCL3"),pt.size=0.2,
            cols=c("grey","tomato4"),min.cutoff=0.25, order=T,ncol=2)&theme(axis.title=element_blank())

FeaturePlot(MDS_NBM_CD8, split.by = "Group", features=c("CTLA4","PDCD1","LAG3","IL10"),min.cutoff=0.1, sort=T,pt.size=0.15)

FeaturePlot(MDS_NBM_CD8, split.by = "Group", features=c("CCR7","PRF1","GZMK","GZMB"), 
            cols=c("grey","blue"),min.cutoff=0.25,order=T,pt.size=0.2)&theme(axis.title=element_blank())


FeaturePlot(MDS_NBM_CD8, split.by = "Group", features=c("GPR56"),min.cutoff=0.25, order=T)

FeaturePlot(MDS_NBM_CD8, split.by = "Group_cluster", features=c("ITGAE","ENTPD1"),pt.size=0.2, min.cutoff=0.25, sort=T) #Tissue resident

DotPlot(MDS_NBM_CD8, features=c("TNF","IFNG","TGFB1"))

DotPlot(MDS_NBM_CD8, features=c("HAVCR2"),split.by = "Sample",cols=rep("tomato4",30))

FeaturePlot(MDS_NBM_CD8, split.by = "Group", features=c("ENTPD1"),min.cutoff=0.25,pt.size=0.2,order=T,cols=c("grey","tomato4"))&
  theme(axis.title = element_blank())

densityplot(dataset=MDS_NBM_CD8,features="TNF",split.by="Group")



CD8_proportion<-MDS_NBM_CD8@meta.data%>%group_by(Sample, Immune_type)%>%summarize(n=n())%>%group_by(Sample)%>%
  summarize(Immune_type=Immune_type, Fraction=n/sum(n))
CD8_proportion["Group"]<-ifelse(grepl("MDS", CD8_proportion$Sample),"MDS","NBM")
CD8_proportion$Group<-factor(CD8_proportion$Group,levels=c("NBM","MDS"))
CD8_proportion$Immune_type<-factor(CD8_proportion$Immune_type,levels=c("CD8 CCR7+","CD8 GZMK+","CD8 GZMB+"))

CD8_proportion%>%ggplot(aes(Sample, Fraction,fill=Immune_type))+geom_bar(stat="identity")+theme_classic()

P<-CD8_proportion%>%ggplot(aes(Group, Fraction,color=Group))+geom_boxplot(width=0.5)+geom_jitter(width=0.1)+theme_classic()+
  facet_wrap(.~Immune_type,ncol=3,scales="free")+theme(strip.background = element_blank(),strip.text =element_text(size=12))+
  scale_color_manual(values=c("steelblue3","tomato4"))+theme(legend.position = "none")

P+stat_compare_means(method = "t.test",size=3)

VlnPlot(MDS_NBM_CD8, feature="TBX21",split.by = "Sample", assay="RNA")

CD8_heatmap<-DotPlot(MDS_NBM_CD8, group.by = "Sample",features=c("HAVCR2","PDCD1","CD274","CD40LG"), assay="RNA")$data
CD8_heatmap<-DotPlot(MDS_NBM_CD8, group.by = "Sample",features=c("PRF1","GZMB","FASLG"), assay="RNA")$data
CD8_heatmap<-DotPlot(MDS_NBM_CD8, group.by = "Sample",features=c("TNF","IFNG","TNFSF10","TGFB1","CCL3","TNFRSF9","CD69","HLA-DRA","CTLA4","LAG3","IL10","IL7R","CCR7","SELL","TCF7"), assay="RNA")$data

gene_list<-c("TNF","IFNG","TNFSF10","TGFB1","CCL3","TNFRSF9","CD69","HLA-DRA","CTLA4","LAG3","IL10","IL7R","CCR7","SELL","TCF7")
CD8_markers<- AverageExpression(MDS_NBM_CD8,group.by = "Sample",features=gene_list,assay="RNA")$RNA
CD8_markers<-CD8_markers[,order(colnames(CD8_markers),decreasing = T)]
pheatmap(CD8_markers,scale="row",cluster_cols = F,cluster_rows = F,colorRampPalette(c("blue","red"))(20))




Marker_GZMK<-FindMarkers(MDS_NBM_CD8, ident.1="CD8 GZMK+", assay="RNA")


MDS_NBM_CD8_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_CD8@meta.data,expr_mat=MDS_NBM_CD8@assays$RNA@counts,
                                       contrasts='condition,MDS,NBM',design='~condition',
                                       sample_col="Sample", condition_col="Group",cluster_col="Immune_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)
MDS_NBM_CD8_deseq["Gene"]<-row.names(MDS_NBM_CD8_deseq)






MDS_NBM_CD8_deseq_sig<-MDS_NBM_CD8_deseq%>%filter(padj<0.05)%>%filter(abs(log2FoldChange)>0.585)
MDS_NBM_CD8_deseq_sig["regulation"]<-ifelse(MDS_NBM_CD8_deseq_sig$log2FoldChange>0,"Up","Down")

MDS_NBM_CD8_deseq%>%ggplot(aes(log2FoldChange,-log2(padj)))+geom_point(color="grey",size=0.15)+
  geom_text_repel(data=MDS_NBM_CD8_deseq_sig,mapping=aes(log2FoldChange,-log2(padj),label=Gene,color=regulation),max.overlaps=60)+
  geom_point(data=MDS_NBM_CD8_deseq_sig,mapping=aes(log2FoldChange,-log2(padj)),color="grey40",size=0.25)+theme_bw()+
  geom_hline(yintercept= -log2(0.05),color="grey")+
  geom_vline(xintercept= 0.585,color="grey")+geom_vline(xintercept= -0.585,color="grey")+
  scale_color_manual(values=c("steelblue4","tomato4"))

writexl::write_xlsx(MDS_NBM_CD8_deseq,"single_cell_sequencing/MDS single cell/MDS_NBM_CD8_deseq.xlsx")
MDS_NBM_CD8_deseq<-read_xlsx("single_cell_sequencing/MDS single cell/MDS_NBM_CD8_deseq.xlsx")

gene_list<-c("TCF7","CCR7","IL7R","SELL") 
gene_list<-c("TNFRSF9","CD69","HLA-DRA","TFRC") #c("TNFRSF9","CD69","TFRC","HLA-DRA") activation marker
gene_list<-c("TNF","IFNG","TNFSF10","TGFB1") #c("TNFRSF9","CD69","TFRC","HLA-DRA") activation marker

gene_list<-c("MKI67","PRF1","GZMB") #c("TNFRSF9","CD69","TFRC","HLA-DRA") activation marker

gene_list<-c("TNF","IFNG","TNFSF10","TGFB1","CCL3","TNFRSF9","CD69","HLA-DRA","CTLA4","LAG3","IL10","IL7R","CCR7","SELL","TCF7")

MDS_NBM_CD8_deseq_gene<-MDS_NBM_CD8_deseq
row.names(MDS_NBM_CD8_deseq_gene)<-MDS_NBM_CD8_deseq_gene$Gene
MDS_NBM_CD8_deseq_gene<-data.frame(t(scale(t(MDS_NBM_CD8_deseq_gene[gene_list,1:9]))))

MDS_NBM_CD8_deseq_annotation<-data.frame(group=ifelse(grepl("NBM",names(MDS_NBM_CD8_deseq_gene)),"NBM","MDS"),row.names = names(MDS_NBM_CD8_deseq_gene))
annotation_col<-list(group=c("MDS"="tomato4","NBM"="steelblue4"))
pheatmap(MDS_NBM_CD8_deseq_gene,clustering_method = "ward.D",annotation_colors =annotation_col ,
         annotation=MDS_NBM_CD8_deseq_annotation,cluster_rows = F,color = colorRampPalette(c("blue","white","yellow2"))(30))

MDS_NBM_CD8_deseq_gene<-MDS_NBM_CD8_deseq
row.names(MDS_NBM_CD8_deseq_gene)<-MDS_NBM_CD8_deseq_gene$Gene
MDS_NBM_CD8_deseq_gene<-data.frame(t(MDS_NBM_CD8_deseq_gene[gene_list,1:9]))
MDS_NBM_CD8_deseq_gene["Group"]<-ifelse(grepl("NBM",rownames(MDS_NBM_CD8_deseq_gene)),"NBM","MDS")
MDS_NBM_CD8_deseq_gene<-melt(MDS_NBM_CD8_deseq_gene, id.vars="Group",variable.name="Gene")
MDS_NBM_CD8_deseq_gene$Group<-factor(MDS_NBM_CD8_deseq_gene$Group,levels=c("NBM","MDS"))
MDS_NBM_CD8_deseq_gene%>%ggplot(aes(Group,value,color=Group))+geom_boxplot(outlier.size = -1)+geom_jitter(width=0.5)+
  facet_wrap(.~Gene,ncol=length(gene_list),scales="free_y")+scale_color_manual(values=c("steelblue3","tomato4"))+theme_classic()+
  theme(strip.background = element_blank(),strip.text = element_text(size=10))

Idents(MDS_NBM_CD8)<-"Immune_type"


VlnPlot(MDS_NBM_CD8, feature=c("CCR7","PRF1","GZMK","GZMB"),split.by = "Group", ncol=1,assay="RNA",pt.size=0,cols=c("steelblue4","tomato4"))&theme(axis.title = element_blank())


VlnPlot(MDS_NBM_CD8, feature=c("IL2RA","CD69","HLA-DRA","CD44"),split.by = "Group", assay="RNA",pt.size=0)


DotPlot(MDS_NBM_CD8, feature=c("SLC2A1"),split.by = "Group", assay="RNA", cols=c("blue","blue"))
DotPlot(MDS_NBM_CD8, feature=c("CD27","CD69","GZMB","GZMK"),split.by = "Sample", assay="RNA", cols=rep("blue",12))
DotPlot(MDS_NBM_CD8, feature=c("CTLA4","PDCD1","LAG3"),split.by = "Sample", assay="RNA", cols=rep("blue",12))
DotPlot(MDS_NBM_CD8, feature=c("CTLA4","PDCD1","LAG3"),split.by = "Sample", assay="RNA", cols=rep("blue",12))
DotPlot(MDS_NBM_CD8, feature=c("IFNG","TNF","TNFSF10"),split.by = "Group", assay="RNA", cols=rep("tomato4",2))
DotPlot(MDS_NBM_CD8, feature=c("CTLA4","PDCD1","HAVCR2"), assay="RNA")
DotPlot(MDS_NBM_CD8, feature=c("CTLA4","LAG3"), assay="RNA")



MDS_NBM_CD8_deseq$pvalue[which(MDS_NBM_CD8_deseq$pvalue==0)]<- 3e-103
MDS_NBM_CD8_deseq["rank"]<-  -sign(MDS_NBM_CD8_deseq$log2FoldChange)*log10(MDS_NBM_CD8_deseq$pvalue)
rank_MDS_NBM_CD8_deseq<-MDS_NBM_CD8_deseq$rank
names(rank_MDS_NBM_CD8_deseq)<-row.names(MDS_NBM_CD8_deseq)
rank_MDS_NBM_CD8_deseq<-rank_MDS_NBM_CD8_deseq[!is.na(rank_MDS_NBM_CD8_deseq)]

MDS_CD8_Hall <- fgsea(hallmakers_set, rank_MDS_NBM_CD8_deseq, minSize=15, maxSize = 500, nperm=10000)
head(MDS_CD8_Hall[order(padj), ], n=10)
MDS_CD8_Hall_sig<-MDS_CD8_Hall%>%filter(padj<0.05)%>%arrange(desc(NES))


MDS_CD8_C2 <- fgsea(C2_set, rank_MDS_NBM_CD8_deseq, minSize=15, maxSize = 500, nperm=10000)
head(MDS_CD8_C2[order(padj), ], n=10)
MDS_CD8_C2_sig<-MDS_CD8_C2%>%filter(padj<0.25)%>%arrange(desc(NES))

MDS_CD8_C5 <- fgsea(C5_set, rank_MDS_NBM_CD8_deseq, minSize=15, maxSize = 500, nperm=10000)
head(MDS_CD8_C5[order(padj), ], n=10)
MDS_CD8_C5_sig<-MDS_CD8_C5%>%filter(padj<0.05)%>%arrange(desc(NES))
MDS_CD8_C5_sig$pathway<-tolower(MDS_CD8_C5_sig$pathway)
MDS_CD8_C5_sig$pathway<-gsub("gobp_",replacement = "GO:",MDS_CD8_C5_sig$pathway)
MDS_CD8_C5_sig_Tcell<-MDS_CD8_C5_sig[grepl("_t_cell",MDS_CD8_C5_sig$pathway),]%>%arrange(NES)%>%
  mutate(pathway=factor(pathway,levels=(pathway)))

MDS_CD8_C5_sig_Tcell%>%ggplot(aes(NES,pathway,fill= -padj))+
  geom_col()+theme_classic()+scale_fill_gradient(high="red",low="blue")




MDS_NBM_CD8_GZMK_deseq<-run_pseudobulkDE(meta_data=subset(MDS_NBM_CD8,ident="CD8 GZMK+")@meta.data,expr_mat=subset(MDS_NBM_CD8,ident="CD8 GZMK+")@assays$RNA@counts,
                                        contrasts='condition,MDS,NBM',design='~condition',
                                        sample_col="Sample", condition_col="Group",cluster_col="Immune_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)

MDS_NBM_CD8_GZMK_deseq["Gene"]<-row.names(MDS_NBM_CD8_GZMK_deseq)

# CD8 Tef

MDS_NBM_CD8_Tef_deseq<-run_pseudobulkDE(meta_data=subset(MDS_NBM_CD8,ident="CD8 GZMB+")@meta.data,expr_mat=subset(MDS_NBM_CD8,ident="CD8 GZMB+")@assays$RNA@counts,
                                    contrasts='condition,MDS,NBM',design='~condition',
                                    sample_col="Sample", condition_col="Group",cluster_col="Immune_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)
MDS_NBM_CD8_Tef_deseq["Gene"]<-row.names(MDS_NBM_CD8_Tef_deseq)

#cd8 naive

MDS_NBM_CD8_naive_deseq<-run_pseudobulkDE(meta_data=subset(MDS_NBM_CD8,ident="CD8 CCR7+")@meta.data,expr_mat=subset(MDS_NBM_CD8,ident="CD8 CCR7+")@assays$RNA@counts,
                                        contrasts='condition,MDS,NBM',design='~condition',
                                        sample_col="Sample", condition_col="Group",cluster_col="Immune_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)
MDS_NBM_CD8_naive_deseq["Gene"]<-row.names(MDS_NBM_CD8_naive_deseq)

#CD4+

MDS_NBM_CD4<-subset(MDS_NBM_Immune, ident=c("CD4 CCR7+", "Th1/17", "Th2","CD4 GZMB+"))
DefaultAssay(MDS_NBM_CD4)<-"integrated"

FindVariableFeatures(MDS_NBM_CD4)
MDS_NBM_CD4 <- ScaleData(MDS_NBM_CD4,vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(MDS_NBM_CD4));
MDS_NBM_CD4 <- RunPCA(MDS_NBM_CD4, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_CD4)
nDims=10
MDS_NBM_CD4 <- FindNeighbors(MDS_NBM_CD4, reduction = "pca", dims = 1:nDims)
MDS_NBM_CD4 <- FindClusters(MDS_NBM_CD4, resolution = 0.3);
MDS_NBM_CD4 <- RunUMAP(MDS_NBM_CD4,dims = 1:nDims);
Idents(MDS_NBM_CD4)<-"Immune_type"

MDS_NBM_CD4$Group<-factor(MDS_NBM_CD4$Group, levels=c("NBM","MDS"))

DimPlot(MDS_NBM_CD4, reduction = "umap",repel=T,split.by="Sample", ncol=4,label=T ,pt.size = 0.1)&NoLegend()

FeaturePlot(MDS_NBM_CD4, split.by = "Group", features=c("IFNG","TNF"),min.cutoff=0.25,pt.size=0.2,sort=T,label=T,cols=c("grey","tomato4"))&
  theme(axis.title = element_blank())


CD4_heatmap<-DotPlot(MDS_NBM_CD4, group.by = "Sample",features=c("TNF","IFNG","TNFSF10","TGFB1","CCL3","TNFRSF9","CD69","HLA-DRA","CTLA4","LAG3","IL7R","CCR7","SELL","TCF7"), assay="RNA")$data
CD4_heatmap$id<-as.character(CD4_heatmap$id)
CD4_heatmap$id<-factor(CD4_heatmap$id,levels=c(unique(CD4_heatmap$id[grepl("^NBM",CD4_heatmap$id)]),unique(CD4_heatmap$id[grepl("^MDS",CD4_heatmap$id)])))
CD4_heatmap%>%ggplot(aes(id,features.plot,fill=avg.exp.scaled))+
  geom_raster()+scale_fill_gradient(low="blue",high="red")+theme_classic()+labs(y="")+theme(axis.text.x=element_text(angle = 45,hjust=1,vjust=1))




CD4_proportion<-MDS_NBM_CD4@meta.data%>%group_by(Sample, Immune_type)%>%summarize(n=n())%>%group_by(Sample)%>%
  summarize(Immune_type=Immune_type, Fraction=n/sum(n))
CD4_proportion["Group"]<-ifelse(grepl("MDS", CD4_proportion$Sample),"MDS","NBM")
CD4_proportion$Group<-factor(CD4_proportion$Group,levels=c("NBM","MDS"))

P<-CD4_proportion%>%ggplot(aes(Group, Fraction,color=Group))+geom_boxplot()+geom_jitter(width=0.1)+theme_classic()+
  facet_wrap(.~Immune_type)+theme(strip.background = element_blank(),strip.text =element_text(size=12))

P+stat_compare_means(method = "t.test",size=3)+scale_color_manual(values=c("steelblue3","tomato4"))

MDS_NBM_CD4_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_CD4@meta.data,expr_mat=MDS_NBM_CD4@assays$RNA@counts,
                                        contrasts='condition,MDS,NBM',design='~condition',
                                        sample_col="Sample", condition_col="Group",cluster_col="Immune_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)
MDS_NBM_CD4_deseq["Gene"]<-row.names(MDS_NBM_CD4_deseq)

 write_xlsx(MDS_NBM_CD4_deseq,"CD4_MDS_vs_NBM.xlsx")

DotPlot(MDS_NBM_CD4, feature=c("CD69"),split.by = "Group", assay="RNA", cols=rep("blue",8))
DotPlot(MDS_NBM_CD4, features=c("HLA-DRA","TNFRSF9"),split.by = "Sample",cols=rep("blue",10))

FeaturePlot(MDS_NBM_CD4, split.by = "Group", features=c("CTLA4"),min.cutoff=0.25, sort=T)
VlnPlot(MDS_NBM_CD4,split.by = "Sample", features=c("CD69"))

gene_list<-c("TNFRSF9","CD69","HLA-DRA","TFRC") #c("TNFRSF9","CD69","TFRC","HLA-DRA") activation marker
gene_list<-c("TNF","IFNG","TNFSF10","TGFB1") #c("TNFRSF9","CD69","TFRC","HLA-DRA") activation marker


MDS_NBM_CD4_deseq_gene<-MDS_NBM_CD4_deseq
row.names(MDS_NBM_CD4_deseq_gene)<-MDS_NBM_CD4_deseq_gene$Gene
MDS_NBM_CD4_deseq_gene<-data.frame(t(MDS_NBM_CD4_deseq_gene[gene_list,1:9]))
MDS_NBM_CD4_deseq_gene["Group"]<-ifelse(grepl("NBM",rownames(MDS_NBM_CD4_deseq_gene)),"NBM","MDS")
MDS_NBM_CD4_deseq_gene<-melt(MDS_NBM_CD4_deseq_gene, id.vars="Group",variable.name="Gene")
MDS_NBM_CD4_deseq_gene$Group<-factor(MDS_NBM_CD4_deseq_gene$Group,levels=c("NBM","MDS"))
MDS_NBM_CD4_deseq_gene%>%ggplot(aes(Group,value,color=Group))+geom_boxplot(outlier.size = -1)+geom_jitter(width=0.5)+
  facet_wrap(.~Gene,ncol=length(gene_list),scales="free_y")+scale_color_manual(values=c("steelblue3","tomato4"))+theme_classic()+
  theme(strip.background = element_blank(),strip.text = element_text(size=10))

MDS_NBM_CD4_deseq$pvalue[which(MDS_NBM_CD4_deseq$pvalue==0)]<- 3e-103
MDS_NBM_CD4_deseq["rank"]<-  -sign(MDS_NBM_CD4_deseq$log2FoldChange)*log10(MDS_NBM_CD4_deseq$pvalue)
rank_MDS_NBM_CD4_deseq<-MDS_NBM_CD4_deseq$rank
names(rank_MDS_NBM_CD4_deseq)<-row.names(MDS_NBM_CD4_deseq)
rank_MDS_NBM_CD4_deseq<-rank_MDS_NBM_CD4_deseq[!is.na(rank_MDS_NBM_CD4_deseq)]

MDS_CD4_Hall <- fgsea(hallmakers_set, rank_MDS_NBM_CD4_deseq, minSize=15, maxSize = 500, nperm=10000)
head(MDS_CD4_Hall[order(padj), ], n=10)
MDS_CD4_Hall_sig<-MDS_CD4_Hall%>%filter(padj<0.05)%>%arrange(desc(NES))

MDS_CD4_C2 <- fgsea(C2_set, rank_MDS_NBM_CD4_deseq, minSize=15, maxSize = 500, nperm=10000)
head(MDS_CD4_C2[order(padj), ], n=10)
MDS_CD4_C2_sig<-MDS_CD4_C2%>%filter(padj<0.25)%>%arrange(desc(NES))

MDS_CD4_C5 <- fgsea(C5_set, rank_MDS_NBM_CD4_deseq, minSize=15, maxSize = 500, nperm=10000)
head(MDS_CD4_C5[order(padj), ], n=10)
MDS_CD4_C5_sig<-MDS_CD4_C5%>%filter(padj<0.25)%>%arrange(desc(NES))


MDS_CD4_C5_sig$pathway<-tolower(MDS_CD4_C5_sig$pathway)
MDS_CD4_C5_sig$pathway<-gsub("gobp_",replacement = "GO:",MDS_CD4_C5_sig$pathway)
MDS_CD4_C5_sig_Tcell<-MDS_CD4_C5_sig[grepl("_t_cell",MDS_CD4_C5_sig$pathway),]%>%arrange(NES)%>%
  mutate(pathway=factor(pathway,levels=(pathway)))

MDS_CD4_C5_sig_Tcell%>%ggplot(aes(NES,pathway,fill= -padj))+
  geom_col()+theme_classic()+scale_fill_gradient(high="red",low="blue")

#Treg

MDS_NBM_Treg<-subset(MDS_NBM_Immune, ident=c("Treg"))
DefaultAssay(MDS_NBM_Treg)<-"integrated"

FindVariableFeatures(MDS_NBM_Treg)
MDS_NBM_Treg <- ScaleData(MDS_NBM_Treg,vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(MDS_NBM_CD8));
MDS_NBM_Treg <- RunPCA(MDS_NBM_Treg, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_Treg)
nDims=5
MDS_NBM_Treg <- FindNeighbors(MDS_NBM_Treg, reduction = "pca", dims = 1:nDims)
MDS_NBM_Treg <- FindClusters(MDS_NBM_Treg, resolution = 0.3);
MDS_NBM_Treg <- RunUMAP(MDS_NBM_Treg,dims = 1:nDims);
Idents(MDS_NBM_Treg)<-"seurat_clusters"

MDS_NBM_Treg<-subset(MDS_NBM_Treg,cells=names(MDS_NBM_Treg$Sample)[MDS_NBM_Treg$Sample!="MDS_SF3B1"])
MDS_NBM_Treg$Group<-factor(MDS_NBM_Treg$Group, levels=c("NBM","MDS"))

DimPlot(MDS_NBM_Treg, reduction = "umap",repel=T,split.by="Sample", ncol=4,label=T)&NoLegend()


MDS_NBM_Treg_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_Treg@meta.data,expr_mat=MDS_NBM_Treg@assays$RNA@counts,
                                    contrasts='condition,MDS,NBM',design='~condition',
                                    sample_col="Sample", condition_col="Group",cluster_col="Immune_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)
MDS_NBM_Treg_deseq["Gene"]<-row.names(MDS_NBM_Treg_deseq)

write_xlsx(MDS_NBM_Treg_deseq,"Treg_MDS_vs_NBM.xlsx")

#NK
Idents(MDS_NBM_Immune)<-"Immune_type"
MDS_NBM_NK<-subset(MDS_NBM_Immune, ident=c("NK CD56dim","NK CD56high"))
DefaultAssay(MDS_NBM_NK)<-"integrated"

MDS_NBM_NK<-subset(MDS_NBM_NK, ident=c("2","3"),invert=T)

FindVariableFeatures(MDS_NBM_NK)
MDS_NBM_NK <- ScaleData(MDS_NBM_NK,vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(MDS_NBM_Immune));
MDS_NBM_NK <- RunPCA(MDS_NBM_NK, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_NK)
nDims=10
MDS_NBM_NK <- FindNeighbors(MDS_NBM_NK, reduction = "pca", dims = 1:nDims)
MDS_NBM_NK <- FindClusters(MDS_NBM_NK, resolution = 0.3);
MDS_NBM_NK <- RunUMAP(MDS_NBM_NK,dims = 1:nDims);
Idents(MDS_NBM_NK)<-"Immune_type"


MDS_NBM_NK$UMAP_1<-MDS_NBM_NK@reductions$umap@cell.embeddings[,1]
MDS_NBM_NK$UMAP_2<-MDS_NBM_NK@reductions$umap@cell.embeddings[,2]

DimPlot(MDS_NBM_NK, reduction = "umap",repel=T,split.by="Sample", ncol=4,label=T)&NoLegend()


MT_prediction<-read.table("~/single_cell_sequencing/MDS single cell/MRLC_classification.tsv",header=T)

MT_prediction$Sample<-gsub("_HM","",MT_prediction$sample)
MT_prediction$Sample<-gsub("_NI","",MT_prediction$Sample)

MT_prediction["type"]<-ifelse(grepl("NI",MT_prediction$sample),"niche_immune","")
MT_prediction["type"]<-ifelse(grepl("HM",MT_prediction$sample),"cd34_myeloid",MT_prediction$type)




MT_prediction["bar_code"]<-tolower(paste0(MT_prediction$Sample,"_",MT_prediction$type,"_",MT_prediction$barcode,"_1"))
MT_prediction["bar_code"]<-gsub("__","_",MT_prediction$bar_code)

head(MDS_NBM_NK$bar_code %in% MT_prediction$bar_code)
sum(MDS_NBM_NK$bar_code %in% MT_prediction$bar_code)

MT_prediction<-MT_prediction[c("Prediction","bar_code")]
metadata_MDS_NBM_NK<-MDS_NBM_NK@meta.data

metadata_MDS_NBM_NK<-metadata_MDS_NBM_NK%>%left_join(MT_prediction)
row.names(metadata_MDS_NBM_NK)<-metadata_MDS_NBM_NK$bar_code
metadata_MDS_NBM_NK<-metadata_MDS_NBM_NK[MDS_NBM_NK$bar_code,]
MDS_NBM_NK[["SF3B1_prediction"]]<-metadata_MDS_NBM_NK$Prediction
MDS_NBM_NK[["SF3B1_prediction"]]<-ifelse(MDS_NBM_NK$SF3B1_prediction,"MT","Unknown")
MDS_NBM_NK[["SF3B1_prediction"]]<-ifelse(grepl("NBM",MDS_NBM_NK$Group),"NBM",MDS_NBM_NK$SF3B1_prediction)
MDS_NBM_NK$SF3B1_prediction<-factor(MDS_NBM_NK$SF3B1_prediction,levels=c("NBM","MT","Unknown"))

MDS_NBM_NK<-subset(MDS_NBM_NK,subset=seurat_clusters=="4",invert=T)

MDS_NBM_NK[["NK_type"]]<-MDS_NBM_NK$Immune_type
MDS_NBM_NK$NK_type<-ifelse(MDS_NBM_NK$seurat_clusters %in% c("1"), "NK CD16-_1",MDS_NBM_NK$NK_type)
MDS_NBM_NK$NK_type<-ifelse(MDS_NBM_NK$seurat_clusters %in% c("2"), "NK CD16-_2",MDS_NBM_NK$NK_type)

MDS_NBM_NK$NK_type<-ifelse(MDS_NBM_NK$seurat_clusters %in% c("0"), "NK CD16+",MDS_NBM_NK$NK_type)
MDS_NBM_NK$NK_type<-ifelse(MDS_NBM_NK$seurat_clusters %in% c("3"), "NK CD16+ IFNy+",MDS_NBM_NK$NK_type)
MDS_NBM_NK$NK_type<-factor(MDS_NBM_NK$NK_type)
Idents(MDS_NBM_NK)<-"NK_type"




#MDS_NBM_NK<-subset(MDS_NBM_NK,cells=names(MDS_NBM_NK$Sample)[MDS_NBM_NK$Sample!="MDS_SF3B1"])
MDS_NBM_NK$Group<-factor(MDS_NBM_NK$Group, levels=c("NBM","MDS"))

MDS_NBM_NK

Idents(MDS_NBM_NK)<-"NK_type"

DimPlot(MDS_NBM_NK, reduction = "umap",repel=T,split.by="Group", ncol=4,label=T,cols=c("steelblue3","steelblue4","tomato4","darkgreen"))


DimPlot(subset(MDS_NBM_NK,subset=Group=="MDS"),cells.highlight=names(MDS_NBM_NK$SF3B1_prediction[MDS_NBM_NK$SF3B1_prediction=="MT"]), reduction = "umap",repel=T,label=T)


DimPlot(subset(MDS_NBM_NK,subset=Group=="NBM"), reduction = "umap",repel=T,split.by="Group", ncol=4,label=T,cols=c("steelblue3","steelblue4","tomato4","darkgreen"))


NK_proportion<-MDS_NBM_NK@meta.data%>%group_by(Sample,Group, NK_type)%>%summarize(n=n())%>%group_by(Group,Sample)%>%
  summarize(NK_type=NK_type, perc=n/sum(n))

NK_proportion$Group<-factor(NK_proportion$Group,levels=c("NBM","MDS"))


P<-NK_proportion%>%ggplot(aes(Group, perc,color=Group))+geom_boxplot()+geom_jitter(width=0.1)+theme_classic()+
  facet_wrap(.~NK_type,scales="free_x")+theme(strip.background = element_blank(),strip.text =element_text(size=12))

P+stat_compare_means(method = "t.test",size=3)+scale_color_manual(values=c("steelblue3","tomato4"))

NK_markers<-FindAllMarkers(subset(MDS_NBM_NK,subset=Group=="NBM"),assay="RNA",only.pos = T)
write.xlsx(NK_markers,"single_cell_sequencing/MDS single cell/Summary/T and NK cells/Markers_of_NK_subpopulaitons_NBM.xlsx")



NK_markers_top<-NK_markers%>%group_by(cluster)%>%top_n(n=30,wt= -p_val_adj)
DoHeatmap(subset(MDS_NBM_NK,subset=Group=="NBM"),features=NK_markers_top$gene,raster = T)


FeaturePlot(MDS_NBM_NK, split.by = "Group", features=c("FCGR3A","TNF","GZMB","GZMK","IL7R"),min.cutoff=0.1, order=T,label=T)
FeaturePlot(MDS_NBM_NK, split.by = "Group", features=c("GZMB","GZMK"),min.cutoff=0.1, order=T,label=T)
FeaturePlot(MDS_NBM_NK, split.by = "Group", features=c("IFNG","TNF"),min.cutoff=0.25, order=T,label=T)
FeaturePlot(MDS_NBM_NK, split.by = "Group", features=c("IL7R","SELL","TCF7"),min.cutoff=0.25, order=T,label=T)

densityplot(data=MDS_NBM_NK,reduction="umap",assay="RNA",features="CCR7",split.by = "Group")


DotPlot(subset(MDS_NBM_NK,subset=Group=="NBM"),features=c("IFNG","NCAM1","FCGR3A","PRF1","GZMB","IL7R","SELL","GZMK"),assay="RNA")#&coord_flip()&theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))
DotPlot(MDS_NBM_NK,split.by = "Group",features=c("IFNG","FCGR3A","PRF1","GZMB","GZMK","IL7R","CCR7"),cols=c("blue","blue"))#&coord_flip()&theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))

DotPlot(MDS_NBM_NK,split.by = "Group",features=c("PRF1","FCGR3A","FASLG"),cols=c("blue","blue"))#&coord_flip()&theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))


VlnPlot(MDS_NBM_NK,features=c("PRF1","FCGR3A"),pt.size=0,split.by = "Group",assay="RNA",cols=c("steelblue4","tomato4"))&theme(legend.position = "right")

VlnPlot(MDS_NBM_NK,features=c("PRF1","FCGR3A"),pt.size=0,split.by = "SF3B1_prediction",assay="RNA",ncol=1,cols=c("grey40","tomato4","steelblue4"))&theme(legend.position = "right")


VlnPlot(MDS_NBM_NK,features=c("NFKBIA","DUSP1","RELB","CD44","REL","CD69"),ncol=3,pt.size=0,split.by = "Group",assay="RNA",cols=c("steelblue4","tomato4"))

VlnPlot(MDS_NBM_NK,features=c("PRF1","FCGR3A"),pt.size=0,group.by = "Sample",assay="RNA",cols=c(rep("tomato4",5),rep("steelblue4",4)))


BiocManager::install("dittoSeq")
library(dittoSeq)
dittoBarPlot(MDS_NBM_NK, Idents(MDS_NBM_NK),group.by = "Group")
dittoBarPlot(MDS_NBM_NK, Idents(MDS_NBM_NK),group.by = "Sample")
NK_proportion<-MDS_NBM_NK@meta.data%>%group_by(Sample,Group, Immune_type)%>%summarize(n=n())%>%group_by(Group,Sample)%>%
  summarize(Immune_type=Immune_type, perc=n/sum(n))

NK_proportion$Group<-factor(NK_proportion$Group,levels=c("NBM","MDS"))
P<-NK_proportion%>%ggplot(aes(Group, perc,color=Group))+geom_boxplot()+geom_jitter(width=0.1)+theme_classic()+
  facet_wrap(.~Immune_type)+theme(strip.background = element_blank(),strip.text =element_text(size=12))

P+stat_compare_means(method = "t.test",size=3)+scale_color_manual(values=c("steelblue3","tomato4"))


MDS_NBM_NK_deseq<-run_pseudobulkDE_new(data=MDS_NBM_NK,
                                     contrasts='condition,MDS,NBM',design='~condition',
                                     sample_col="Sample", condition_col="Group",extr_col=NULL,do_prefiltering = TRUE)
MDS_NBM_NK_deseq["Gene"]<-row.names(MDS_NBM_NK_deseq)
MDS_NBM_NK_deseq<-MDS_NBM_NK_deseq[order(MDS_NBM_NK_deseq$padj),]

write_xlsx(MDS_NBM_NK_deseq,"single_cell_sequencing/MDS single cell/Summary/NK_MDS_vs_NBM.xlsx")


MDS_NBM_NK_deseq_sig<-MDS_NBM_NK_deseq%>%filter(padj<0.05)%>%filter(abs(log2FoldChange)>0.585)
MDS_NBM_NK_deseq_sig["regulation"]<-ifelse(MDS_NBM_NK_deseq_sig$log2FoldChange>0,"Up","Down")

MDS_NBM_NK_deseq%>%ggplot(aes(log2FoldChange,-log2(padj)))+geom_point(color="grey",size=0.15)+
  geom_text_repel(data=MDS_NBM_NK_deseq_sig,mapping=aes(log2FoldChange,-log2(padj),label=Gene,color=regulation),max.overlaps=60)+
  geom_point(data=MDS_NBM_NK_deseq_sig,mapping=aes(log2FoldChange,-log2(padj)),color="grey40",size=0.25)+theme_bw()+
  geom_hline(yintercept= -log2(0.05),color="grey")+
  geom_vline(xintercept= 0.585,color="grey")+geom_vline(xintercept= -0.585,color="grey")+
  scale_color_manual(values=c("steelblue4","tomato4"))


library(ggpubr)
NK_proportion<-MDS_NBM_NK@meta.data%>%group_by(Sample, Immune_type)%>%summarize(n=n())%>%group_by(Sample)%>%
  summarize(Immune_type=Immune_type, Fraction=n/sum(n))
NK_proportion["Group"]<-ifelse(grepl("MDS", NK_proportion$Sample),"MDS","NBM")
NK_proportion$Group<-factor(NK_proportion$Group,levels=c("NBM","MDS"))
NK_proportion$Immune_type<-factor(NK_proportion$Immune_type,levels=c("NK Naive","NK Cytotoxic"))

P<-NK_proportion%>%ggplot(aes(Group, Fraction,color=Group))+geom_boxplot()+geom_jitter(width=0.1)+theme_classic()+
  facet_wrap(.~Immune_type)+theme(strip.background = element_blank(),strip.text =element_text(size=12))

P+stat_compare_means(method = "t.test",size=3)+scale_color_manual(values=c("steelblue3","tomato4"))


MDS_NBM_NK_deseq$pvalue[which(MDS_NBM_NK_deseq$pvalue==0)]<- 3e-103
MDS_NBM_NK_deseq["rank"]<-  -sign(MDS_NBM_NK_deseq$log2FoldChange)*log10(MDS_NBM_NK_deseq$pvalue)
rank_MDS_NBM_NK_deseq<-MDS_NBM_NK_deseq$rank
names(rank_MDS_NBM_NK_deseq)<-row.names(MDS_NBM_NK_deseq)
rank_MDS_NBM_NK_deseq<-rank_MDS_NBM_NK_deseq[!is.na(rank_MDS_NBM_NK_deseq)]

MDS_NK_Hall <- fgsea(hallmakers_set, rank_MDS_NBM_NK_deseq, minSize=15, maxSize = 500, nperm=10000)
head(MDS_NK_Hall[order(padj), ], n=10)
MDS_NK_Hall_sig<-MDS_NK_Hall%>%filter(padj<0.05)%>%arrange(NES)%>%
  mutate(pathway=factor(pathway,levels=pathway))

MDS_NK_Hall_sig%>%ggplot(aes(NES,pathway,fill= -padj))+
  geom_col()+theme_classic()+scale_fill_gradient(high="red",low="blue")



MDS_NK_C2 <- fgsea(C2_set, rank_MDS_NBM_NK_deseq, minSize=15, maxSize = 500, nperm=10000)
head(MDS_NK_C2[order(padj), ], n=10)
MDS_NK_C2_sig<-MDS_NK_C2%>%filter(padj<0.25)%>%arrange(desc(NES))

MDS_NK_C5 <- fgsea(C5_set, rank_MDS_NBM_NK_deseq, minSize=15, maxSize = 500, nperm=10000)
head(MDS_NK_C5[order(padj), ], n=10)
MDS_NK_C5_sig<-MDS_NK_C5%>%filter(padj<0.25)%>%arrange(desc(NES))


plotEnrichment(C2_set[["KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY"]], rank_MDS_NBM_NK_deseq)+labs(title="KEGG_natural_killer_cell_mediated_cytotoxicity")
plotEnrichment(C5_set[["GOBP_NATURAL_KILLER_CELL_MEDIATED_IMMUNITY"]], rank_MDS_NBM_NK_deseq)+labs(title="GOBP_NATURAL_KILLER_CELL_MEDIATED_IMMUNITY")


DefaultAssay(MDS_NBM_NK)<-"RNA"
MDS_NBM_NK_score<-AddModuleScore(MDS_NBM_NK,features=hallmakers_set[1],name="NFkB_via_TNF")
FeaturePlot(MDS_NBM_NK_score, split.by = "Group", features=c("NFkB_via_TNF1"),min.cutoff=0.1, order=T,label=T)

# CD16- deseq

MDS_NBM_NK_CD16low_deseq<-run_pseudobulkDE_new(data=subset(MDS_NBM_NK,ident=c("NK CD16-_1","NK CD16-_2")),
                                            contrasts='condition,MDS,NBM',design='~condition',
                                            sample_col="Sample", condition_col="Group",extr_col=NULL,do_prefiltering = TRUE)
MDS_NBM_NK_CD16low_deseq["Gene"]<-row.names(MDS_NBM_NK_CD16low_deseq)
MDS_NBM_NK_CD16low_deseq<-MDS_NBM_NK_CD16low_deseq[order(MDS_NBM_NK_CD16low_deseq$padj),]

write_xlsx(MDS_NBM_NK_CD16low_deseq,"single_cell_sequencing/MDS single cell/Summary/NK_CD16low_MDS_vs_NBM.xlsx")

# CD16+ deseq

MDS_NBM_NK_CD16high_deseq<-run_pseudobulkDE_new(data=subset(MDS_NBM_NK,ident=c("NK CD16+")),
                                               contrasts='condition,MDS,NBM',design='~condition',
                                               sample_col="Sample", condition_col="Group",extr_col=NULL,do_prefiltering = TRUE)
MDS_NBM_NK_CD16high_deseq["Gene"]<-row.names(MDS_NBM_NK_CD16high_deseq)
MDS_NBM_NK_CD16high_deseq<-MDS_NBM_NK_CD16high_deseq[order(MDS_NBM_NK_CD16high_deseq$padj),]

write_xlsx(MDS_NBM_NK_CD16high_deseq,"single_cell_sequencing/MDS single cell/Summary/NK_CD16high_MDS_vs_NBM.xlsx")

# MT vs others

MDS_NK_MT_deseq<-run_pseudobulkDE_new(data=MDS_NBM_NK,
                                                contrasts='condition,MT,Unknown',design='~condition + Extra',
                                                sample_col="Sample", condition_col="SF3B1_prediction",extr_col="Sample",do_prefiltering = TRUE)
MDS_NK_MT_deseq["Gene"]<-row.names(MDS_NK_MT_deseq)
MDS_NK_MT_deseq<-MDS_NK_MT_deseq[order(MDS_NK_MT_deseq$padj),]

write_xlsx(MDS_NK_MT_deseq,"single_cell_sequencing/MDS single cell/Summary/MDS_NK_MT_deseq.xlsx")


###########################
##HSPC+Monocyte Clusitification 
################

summary(factor(MDS_NBM_all_mt$type))

MDS_NBM_HSPC_M<-subset(MDS_NBM_all_mt,ident=c("HSPC(CD34+)","MEPs","CD14+ Mono","Erythroid progenitors","pDCs","DC"))



MDS_NBM_HSPC_M <- ScaleData(MDS_NBM_HSPC_M);
DefaultAssay(MDS_NBM_HSPC_M)<-"integrated"

MDS_NBM_HSPC_M <- RunPCA(MDS_NBM_HSPC_M, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_HSPC_M)
nDims=15
MDS_NBM_HSPC_M <- FindNeighbors(MDS_NBM_HSPC_M, reduction = "pca", dims = 1:nDims)
MDS_NBM_HSPC_M <- FindClusters(MDS_NBM_HSPC_M, resolution=0.4);
MDS_NBM_HSPC_M <- RunUMAP(MDS_NBM_HSPC_M,dims = 1:nDims);
DimPlot(MDS_NBM_HSPC_M, reduction = "umap", label=T,pt.size = 0.25,split.by = "Sample")

MDS_NBM_HSPC_M<-MDS_NBM_HSPC_M
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

DefaultAssay(MDS_NBM_HSPC_M)<-"integrated"

MDS_NBM_HSPC_M <- CellCycleScoring(MDS_NBM_HSPC_M, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

MDS_NBM_HSPC_M$Proliferation_score <- MDS_NBM_HSPC_M$S.Score - MDS_NBM_HSPC_M$G2M.Score

DefaultAssay(MDS_NBM_HSPC_M)<-"integrated"


MDS_NBM_HSPC_M<-subset(MDS_NBM_HSPC_M,ident="4",invert=T)

FindVariableFeatures(MDS_NBM_HSPC_M)
MDS_NBM_HSPC_M<- ScaleData(MDS_NBM_HSPC_M,vars.to.regress = c("S.Score", "G2M.Score"))


MDS_NBM_HSPC_M <- RunPCA(MDS_NBM_HSPC_M, npcs = 30,verbose = FALSE);

ElbowPlot(MDS_NBM_HSPC_M,ndims = 30)
nDims<-20
MDS_NBM_HSPC_M <- FindNeighbors(MDS_NBM_HSPC_M, reduction = "pca", dims = 1:nDims)
MDS_NBM_HSPC_M <- FindClusters(MDS_NBM_HSPC_M, resolution=0.4);
MDS_NBM_HSPC_M <- RunUMAP(MDS_NBM_HSPC_M,dims = 1:nDims);



MDS_NBM_HSPC_M$Group<-factor(MDS_NBM_HSPC_M$Group,levels=c("NBM","MDS"))

DimPlot(MDS_NBM_HSPC_M, reduction = "umap",repel=T ,label=F,pt.size = 0.25,split.by = "Group")

Marker<-FindMarkers(MDS_NBM_HSPC_M,ident.1="8")
DimPlot(MDS_NBM_HSPC_M, reduction = "umap", label=F,pt.size = 0.25,split.by = "Mutant_state")

DimPlot(MDS_NBM_HSPC_M, reduction = "umap", label=F,pt.size = 0.25,group.by = "Phase")

DotPlot(MDS_NBM_HSPC_M,features=c("CD55","ECM1","GDF15","HDGF","IGFBP7","IL17RA","IL18","IL6ST","IRS2","LEPR","NECTIN2","PDZK1IP1",
                                  "PLGRKT","PLXNA1","PVR","TF","TGFB1","TGFBR1","TMPO","TEML2"),split.by = "Group",cols=c("blue","blue"))&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))

DotPlot(MDS_NBM_HSPC_M,features=c("ECM1","GDF15","PVR","PLXNA1"),split.by = "Group",cols=c("blue","blue"))&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))

DotPlot(MDS_NBM_HSPC_M,features=c("PTPRC","TFRC","CD34"),split.by = "Group",cols=rep("blue",9))&
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))


FeaturePlot(MDS_NBM_HSPC_M, features=c("FCER1A","CST3","HLA-DRB5"),reduction = "umap", order = T ,label=F,min.cutoff=0.25,pt.size = 0.25,split.by = "Group")
                    #DC markers
FeaturePlot(MDS_NBM_HSPC_M, features=c("IL7R","CD79A","CD19"),reduction = "umap", order = T ,min.cutoff=0.25,pt.size = 0.25,split.by = "Group")

FeaturePlot(MDS_NBM_HSPC_M, features=c("KIT"),reduction = "umap", order = T ,label=F,min.cutoff=0.25,pt.size = 0.25,split.by = "Group")


FeaturePlot(MDS_NBM_HSPC_M, features=c("PTPRC","TFRC","CD34"),reduction = "umap", order = T ,label=F,min.cutoff=0.25,pt.size = 0.25,split.by = "Group")


HSPC_ref_2<-read.csv("~/single_cell_sequencing/Reference dataset/Cell_lineage_specific_gene.csv")
gene_ID<-read.csv("~/single_cell_sequencing/Reference dataset/Ens_ID.csv")
HSPC_ref_2<-HSPC_ref_2[c(1,2)]%>%dplyr::rename(Ens=Ensembl_gene_id)%>%left_join(gene_ID)


DefaultAssay(MDS_NBM_HSPC_M)="RNA"
MDS_NBM_HSPC_M<-AddModuleScore(MDS_NBM_HSPC_M,features=list(HSPC_ref_2$name[HSPC_ref_2$Model=="HSC"]),name="HSC",ctrl=10)
MDS_NBM_HSPC_M<-AddModuleScore(MDS_NBM_HSPC_M,features=list(HSPC_ref_2$name[HSPC_ref_2$Model=="GMP"]),name="GMP",ctrl=10)
MDS_NBM_HSPC_M<-AddModuleScore(MDS_NBM_HSPC_M,features=list(HSPC_ref_2$name[HSPC_ref_2$Model=="CLP"]),name="CLP",ctrl=10)
MDS_NBM_HSPC_M<-AddModuleScore(MDS_NBM_HSPC_M,features=list(HSPC_ref_2$name[HSPC_ref_2$Model=="MEP"]),name="MEP",ctrl=10)


FeaturePlot(MDS_NBM_HSPC_M, reduction = "umap", features=c("TLR4"), label=F,pt.size = 0.5,sort.cell = T,min.cutoff =0,split.by = "Group")
FeaturePlot(MDS_NBM_HSPC_M, reduction = "umap", features=c("CD34","KIT"), label=F,pt.size = 0.5,sort.cell = T,min.cutoff =0)

FeaturePlot(MDS_NBM_HSPC_M, reduction = "umap", features=c("HSC1","MPP1","CMP1"), label=F,pt.size = 0.5,sort.cell = T,min.cutoff =0,split.by = "Group")
FeaturePlot(MDS_NBM_HSPC_M, reduction = "umap", features=c("HSC1","GMP1","CLP1","MEP1"), 
            split.by = "Group",label=F,pt.size = 0.2,order=T,min.cutoff =0.05,cols=c("grey80","steelblue4"))&theme(axis.title=element_blank())

DimPlot(MDS_NBM_HSPC_M, reduction = "umap", label=T,pt.size = 0.25,split.by = "Group")&NoLegend()

MDS_NBM_HSPC_M$type.clustify

Idents(MDS_NBM_HSPC_M)<-"seurat_clusters"

DefaultAssay(MDS_NBM_HSPC_M)<-"RNA"
MDS_NBM_HSPC_M<-FindVariableFeatures(MDS_NBM_HSPC_M)

Idents(MDS_NBM_HSPC_M)<-"seurat_clusters"

DimPlot(MDS_NBM_HSPC_M, reduction = "umap", split.by="Group" ,label=T,pt.size = 0.25)
MDS_NBM_HSPC_M[["HSPC_type"]]<-as.character(MDS_NBM_HSPC_M$type)
MDS_NBM_HSPC_M$HSPC_type[MDS_NBM_HSPC_M$seurat_clusters %in% c("6")]<-"HSC/MPPs"
MDS_NBM_HSPC_M$HSPC_type[MDS_NBM_HSPC_M$seurat_clusters %in% c("10","11")]<-"CLPs"
MDS_NBM_HSPC_M$HSPC_type[MDS_NBM_HSPC_M$seurat_clusters %in% c("4","12")]<-"MEPs"
MDS_NBM_HSPC_M$HSPC_type[MDS_NBM_HSPC_M$seurat_clusters %in% c("2")]<-"Erythroid progenitors"
MDS_NBM_HSPC_M$HSPC_type[MDS_NBM_HSPC_M$seurat_clusters %in% c("5")]<-"LMPPs"
MDS_NBM_HSPC_M$HSPC_type[MDS_NBM_HSPC_M$seurat_clusters %in% c("8")]<-"GMPs"
MDS_NBM_HSPC_M$HSPC_type[MDS_NBM_HSPC_M$seurat_clusters %in% c("3")]<-"Monoblasts"

MDS_NBM_HSPC_M$HSPC_type[MDS_NBM_HSPC_M$seurat_clusters %in% c("0","1")]<-"CD14+ Mono"

MDS_NBM_HSPC_M@meta.data$HSPC_type[MDS_NBM_HSPC_M@meta.data$seurat_clusters %in% c("7","9")]<-"DCs"


MDS_NBM_HSPC_M$HSPC_type<-factor(MDS_NBM_HSPC_M$HSPC_type, levels=c("HSC/MPPs","LMPPs","CLPs","MEPs","Erythroid progenitors","GMPs","Monoblasts","CD14+ Mono","DCs"))
Idents(MDS_NBM_HSPC_M)<-"HSPC_type"
DimPlot(MDS_NBM_HSPC_M, reduction = "umap", split.by = "Group",ncol=2,label=T,pt.size = 0.25,repel = T,
        cols=c("grey","tan4","darkgreen","tomato2","tomato4","steelblue","steelblue3","steelblue4","grey80","tomato3"))



NBM_markers<-FindAllMarkers(subset(MDS_NBM_HSPC_M,cells=names(MDS_NBM_HSPC_M$Group[MDS_NBM_HSPC_M$Group=="NBM"])),assay="RNA",only.pos = T) 
NBM_markers_top<-NBM_markers%>%filter(p_val_adj<0.05)%>%group_by(cluster)%>%top_n(n=30,wt=avg_log2FC)

FeaturePlot(MDS_NBM_HSPC_M, features=c("PTX3"),reduction = "umap", order = T ,label=F,min.cutoff=0.25,pt.size = 0.25,split.by = "Group")
FeaturePlot(MDS_NBM_HSPC_M, features=c("MS4A1","CD79A","IL7R"),reduction = "umap", order = T ,label=F,min.cutoff=0.25,pt.size = 0.25,split.by = "Group")


DefaultAssay(MDS_NBM_HSPC_M)<-"RNA"
HSPC_marker<-DotPlot(MDS_NBM_HSPC_M,features = unique(NBM_markers_top$gene),split.by = "Group")$data
HSPC_marker$HSPC_type<-str_split(HSPC_marker$id,pattern="_",simplify = T)[,1]
HSPC_marker$Group<-str_split(HSPC_marker$id,pattern="_",simplify = T)[,2]
HSPC_marker$HSPC_type<-factor(HSPC_marker$HSPC_type,levels=c(levels(MDS_NBM_HSPC_M)))
HSPC_marker$gene<-factor(HSPC_marker$features.plot,levels=rev(unique(NBM_markers_top$gene)))

HSPC_marker%>%filter(Group=="NBM")%>%ggplot(aes(HSPC_type,gene,fill=avg.exp.scaled))+
  geom_raster()+scale_fill_viridis_c()+theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))+theme(axis.text.y=element_text(size=3))

HSPC_marker%>%filter(Group=="MDS")%>%ggplot(aes(HSPC_type,gene,fill=avg.exp.scaled))+
  geom_raster()+scale_fill_viridis_c()+theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))+theme(axis.text.y=element_text(size=3))

Idents(MDS_NBM_HSPC_M)<-"HSPC_type"

VlnPlot(MDS_NBM_HSPC_M, features=c("NFKBIA","IRF1","IFITM3","GADD45B","DDIT3","BCL2L2"),
        split.by = "Group",assay="RNA",pt.size=0,cols=c("steelblue4","tomato4"),ncol=3)&theme(axis.title = element_blank())

VlnPlot(MDS_NBM_HSPC_M, features=c("DDX6"),
        split.by = "Group",assay="RNA",pt.size=0,cols=c("steelblue4","tomato4"))&theme(axis.title = element_blank())


DotPlot(MDS_NBM_HSPC_M, features=c("NLRP1"), split.by = "Group",cols=rep("blue",9))+coord_flip()&
  theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))

DotPlot(MDS_NBM_HSPC_M, features=c("IL7R","CD19","CD79A"), split.by = "Sample",cols=rep("blue",9))+coord_flip()&
  theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))

DotPlot(subset(MDS_NBM_HSPC_M), features=c("TNF","IL1B","IFNG"), split.by = "Mutant_state",cols=rep("blue",4))+coord_flip()&
  theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))


VlnPlot(subset(MDS_NBM_HSPC_M,cells=names(MDS_NBM_HSPC_M$type[MDS_NBM_HSPC_M$Mutant_state %in% c("NBM","Mutant")])), features=c("NFKBIA"),split.by = "Sample",assay="RNA",pt.size=0)
FeaturePlot(MDS_NBM_HSPC_M,features=c("DDIT4"),order=T,split.by="Group",min.cutoff = 0.25)


HSPC_proportion<-MDS_NBM_HSPC_M@meta.data%>%group_by(Sample,HSPC_type)%>%summarize(n=n())%>%group_by(Sample)%>%
  filter(HSPC_type %in% c("HSC/MPPs","LMPPs","CLPs","MEPs","GMP/Monoblasts"))%>%
  summarize(HSPC_type=HSPC_type,Perc=n/sum(n))
HSPC_proportion["Group"]<-factor(ifelse(grepl("NBM",HSPC_proportion$Sample),"NBM","MDS"),levels=c("NBM","MDS"))

HSPC_proportion%>%ggplot(aes(Sample,Perc,fill=HSPC_type))+
  geom_bar(stat="identity")+theme(axis.text.x=element_text(angle=45, vjust=0.8,hjust=0.8))+theme_classic()

group_compare<-HSPC_proportion%>%ggplot(aes(Group,Perc,color=Group))+
  geom_boxplot(outlier.size = -1)+geom_jitter(width=0.25)+theme_classic()+
  facet_wrap(.~HSPC_type,ncol=10)+theme(strip.background = element_blank(),strip.placement = "outside",strip.text = element_text(size=12,angle=0))+labs(x="")
group_compare+ stat_compare_means(method = "t.test")+scale_color_manual(values=c("steelblue4","tomato4"))

Mutant_proportion%>%filter(HSPC_type=="HSC/MPPs")


Mutant_proportion<-MDS_NBM_HSPC_M@meta.data%>%group_by(Sample, Mutant_state,HSPC_type)%>%summarize(n=n())%>%group_by(HSPC_type,Sample)%>%
  summarize(Mutant_state=Mutant_state,Perc=n/sum(n))%>%filter(Mutant_state=="Mutant")

Mutant_proportion%>%ggplot(aes(type,Perc,fill=type))+
  geom_col()+facet_wrap(.~Sample, ncol=1,scales = "free_y")+theme_classic()+theme(axis.text.x=element_text(angle=45, vjust=0.8,hjust=0.8))




DotPlot(MDS_NBM_HSPC_M,features=c("IL15"),cols=rep("blue",2),split.by = "Group",assay="RNA")



saveRDS(MDS_NBM_HSPC_M,"~/single_cell_sequencing/MDS single cell/MDS_NBM_HSPC_M.RDS")
MDS_NBM_HSPC_M<-readRDS("scRNAseq_dataset/RDSfile/chenMDS/MDS single cell/MDS_NBM_HSPC_M.RDS")

MDS_NBM_HSPC_M$Group<-factor(MDS_NBM_HSPC_M$Group,levels=c("NBM","MDS"))
##################
#pseudobulkDE
##########

contrasts='condition,MDS,NBM'
design='~ condition'

MDS_NBM_HSPC<-subset(MDS_NBM_HSPC_M,idents=c("CD14+ Mono","DCs","Erythroid progenitors"),invert=T)

MDS_NBM_HSPC_deseq<-run_pseudobulkDE_new(data=MDS_NBM_HSPC,
                                    contrasts='condition,MDS,NBM',design='~ condition',
                                    sample_col="Sample", condition_col="Group",cluster_col="type",extr_col=NULL,do_prefiltering = TRUE)
MDS_NBM_HSPC_deseq["Gene"]<-row.names(MDS_NBM_HSPC_deseq)

MDS_NBM_HSC<-subset(MDS_NBM_HSPC_M,ident="HSC/MPPs")
MDS_NBM_LMPP<-subset(MDS_NBM_HSPC_M,ident="LMPPs")
MDS_NBM_GMP<-subset(MDS_NBM_HSPC_M,ident="GMP/Promonocytes")
MDS_NBM_Mono<-subset(MDS_NBM_HSPC_M,ident="CD14+ Mono")




MDS_NBM_H_deseq<-run_pseudobulkDE_new(data=MDS_NBM_HSC,
                                           contrasts='condition,MDS,NBM',design='~ condition',
                                           sample_col="Sample", condition_col="Group",cluster_col="type",extr_col=NULL,do_prefiltering = TRUE)
MDS_NBM_H_deseq["Gene"]<-row.names(MDS_NBM_H_deseq)
MDS_NBM_H_deseq["rank"]<- -sign(MDS_NBM_H_deseq$log2FoldChange)*log2(MDS_NBM_H_deseq$pvalue)

MDS_NBM_H_deseq_sig<-MDS_NBM_H_deseq[which(MDS_NBM_H_deseq$padj<0.05),]
MDS_NBM_H_deseq_top<-MDS_NBM_H_deseq_sig%>%mutate(gene_change=case_when(.$log2FoldChange>0~"Up", TRUE~"Down"))

MDS_NBM_H_deseq%>%ggplot(aes(log2FoldChange,-log10(padj)))+geom_point(size=0.1,col="grey")+
  geom_point(data=MDS_NBM_H_deseq_sig,mapping=aes(log2FoldChange,-log10(padj)),col="black",size=0.5)+
  geom_vline(xintercept= c(0.5,-0.5),col="grey",type=2)+geom_hline(yintercept = -log10(0.05),col="grey",type=2)+
  geom_text_repel(MDS_NBM_H_deseq_top,mapping=aes(log2FoldChange,-log10(padj),label=Gene),max.overlaps = 20)+
  theme_classic()


MDS_NBM_LMPP_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_LMPP@meta.data,expr_mat=MDS_NBM_LMPP@assays$RNA@counts,
                                  contrasts='condition,MDS,NBM',design='~condition',
                                  sample_col="Sample", condition_col="Group",cluster_col="HSPC_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)
MDS_NBM_LMPP_deseq["Gene"]<-row.names(MDS_NBM_LMPP_deseq)

MDS_NBM_GMP_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_GMP@meta.data,expr_mat=MDS_NBM_GMP@assays$RNA@counts,
                                  contrasts='condition,MDS,NBM',design='~condition',
                                  sample_col="Sample", condition_col="Group",cluster_col="type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)

MDS_NBM_Mono_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_Mono@meta.data,expr_mat=MDS_NBM_Mono@assays$RNA@counts,
                                  contrasts='condition,MDS,NBM',design='~condition',
                                  sample_col="Sample", condition_col="Group",cluster_col="type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)
MDS_NBM_Mono_deseq["Gene"]<-row.names(MDS_NBM_Mono_deseq)


M2_signature<-read.table("~/single_cell_sequencing/Gene_signatures_Monocytes_scRNAseq.txt",sep = "\t")
M2_signature<-M2_signature[,-c(2)]

M2_signature<-split(M2_signature,f=M2_signature$V1)

M2_signature<-lapply(M2_signature,function(x) as.character(x))
M2_signature<-lapply(M2_signature,function(x) x[x!=""])
M2_signature<-lapply(M2_signature,function(x) x[-1])



MDS_NBM_Mono_deseq$rank<- -sign(MDS_NBM_Mono_deseq$log2FoldChange)* log10(MDS_NBM_Mono_deseq$pvalue)
Mono_NBM_MDS_rank<-MDS_NBM_Mono_deseq$rank
names(Mono_NBM_MDS_rank)<-MDS_NBM_Mono_deseq$Gene
Mono_NBM_MDS_rank<-Mono_NBM_MDS_rank[!is.na(Mono_NBM_MDS_rank)]
MDS_NBM_Mono_M2 <- fgsea(M2_signature, Mono_NBM_MDS_rank, minSize=15, maxSize = 500, nperm=10000)
MDS_NBM_Mono_M2_sig<-MDS_NBM_Mono_M2%>%filter(padj<0.25)%>%arrange(desc(NES))

MT_NBM_Mono_deseq<-run_pseudobulkDE(meta_data=MT_NBM_Mono@meta.data,expr_mat=MT_NBM_Mono@assays$RNA@counts,
                                     contrasts='condition,MDS,NBM',design='~condition',
                                     sample_col="Sample", condition_col="Group",cluster_col="HSPC_type",extr_col=NA,do_prefiltering = TRUE,use_celltype = F)

MT_WT_Mono_deseq<-run_pseudobulkDE(meta_data=MT_WT_Mono@meta.data,expr_mat=MT_WT_Mono@assays$RNA@counts,
                                   contrasts='condition,WT,Mutant',design='~condition',
                                   sample_col="Sample", condition_col="Mutant_state",cluster_col="type",extr_col=NA,do_prefiltering = TRUE,use_celltype = TRUE)

#GSEA

#HSPC
MDS_NBM_HSPC_deseq$pvalue[MDS_NBM_HSPC_deseq$pvalue==0]<- 1e-103
rank_HSPC<-  -sign(MDS_NBM_HSPC_deseq$log2FoldChange)*log10(MDS_NBM_HSPC_deseq$pvalue)
names(rank_HSPC)<-row.names(MDS_NBM_HSPC_deseq)
rank_HSPC<-rank_HSPC[!is.na(rank_HSPC)]

MDS_HSPC_Hall <- fgsea(hallmakers_set, rank_HSPC, minSize=15, maxSize = 500, nperm=1000)
head(MDS_HSPC_Hall[order(padj), ], n=10)
MDS_HSPC_Hall_sig<-MDS_HSPC_Hall%>%filter(padj<0.05)%>%arrange(NES)
MDS_HSPC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSPC_Hall_sig$NES<0,"NBM","MDS")
MDS_HSPC_Hall_sig[1]$leadingEdge[[1]]
MDS_HSPC_Hall_sig%>%mutate(pathway=factor(pathway, levels=MDS_HSPC_Hall_sig$pathway))%>%group_by(Enrichment)%>%
  ggplot(aes(NES,pathway))+geom_col()+theme_classic()


#HSC
MDS_NBM_H_deseq$pvalue[MDS_NBM_H_deseq$pvalue==0]<- 1e-103
rank_HSC<-  -sign(MDS_NBM_H_deseq$log2FoldChange)*log10(MDS_NBM_H_deseq$pvalue)
names(rank_HSC)<-row.names(MDS_NBM_H_deseq)
rank_HSC<-rank_HSC[!is.na(rank_HSC)]

MDS_HSC_Hall <- fgsea(hallmakers_set, rank_HSC, minSize=15, maxSize = 500, nperm=10000)
head(MDS_HSC_Hall[order(padj), ], n=10)
MDS_HSC_Hall_sig<-MDS_HSC_Hall%>%filter(padj<0.05)%>%arrange(NES)
MDS_HSC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSC_Hall_sig$NES<0,"NBM","MDS")
MDS_HSC_Hall_sig[1]$leadingEdge[[1]]
MDS_HSC_Hall_sig%>%mutate(pathway=factor(pathway, levels=MDS_HSC_Hall_sig$pathway))%>%group_by(Enrichment)%>%
  ggplot(aes(NES,pathway))+geom_col()+theme_classic()


MDS_HSC_C2 <- fgsea(C2_set, rank_HSC, minSize=15, maxSize = 500, nperm=10000)
MDS_HSC_C2_sig<-MDS_HSC_C2%>%filter(padj<0.25)%>%arrange(NES)

plotEnrichment(C2_set[["GRAHAM_NORMAL_QUIESCENT_VS_NORMAL_DIVIDING_UP"]], rank_HSC,)+labs(title="Graham_quiescent")
plotEnrichment(C2_set[["IVANOVA_HEMATOPOIESIS_STEM_CELL_LONG_TERM"]], rank_HSC,)+labs(title="IVANOVA")




NFKB_marker_gene<-MDS_HSC_Hall_sig[1]$leadingEdge[[1]]
Apoptosis_marker_gene<-MDS_HSC_Hall_sig[2]$leadingEdge[[1]]
NFkB_Apoptosis_markers<-unique(NFKB_marker_gene,Apoptosis_marker_gene)


MDS_NBM_H_deseq$pvalue[MDS_NBM_H_deseq$pvalue==0]<- 1e-103
rank_HSC<-  -sign(MDS_NBM_H_deseq$log2FoldChange)*log10(MDS_NBM_H_deseq$pvalue)
names(rank_HSC)<-row.names(MDS_NBM_H_deseq)
rank_HSC<-rank_HSC[!is.na(rank_HSC)]

MDS_HSC_Hall <- fgsea(hallmakers_set, rank_HSC, minSize=15, maxSize = 500, nperm=10000)
head(MDS_HSC_Hall[order(padj), ], n=10)
MDS_HSC_Hall_sig<-MDS_HSC_Hall%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_HSC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSC_Hall_sig$NES<0,"NBM","MDS")
MDS_HSC_Hall_sig[1]$leadingEdge[[1]]
MDS_HSC_Hall_sig%>%mutate(pathway=factor(pathway, levels=MDS_HSC_Hall_sig$pathway))%>%group_by(Enrichment)%>%
  top_n(10,wt=NES)%>%ggplot(aes(NES,pathway,size= -log2(padj)))+geom_point()+theme_classic()

NFKB_marker_gene<-MDS_HSC_Hall_sig[30]$leadingEdge[[1]]
Apoptosis_marker_gene<-MDS_HSC_Hall_sig[29]$leadingEdge[[1]]
TP53_marker_gene<-MDS_HSC_Hall_sig[27]$leadingEdge[[1]]

Gene_list<-unique(c(NFKB_marker_gene,Apoptosis_marker_gene,TP53_marker_gene))

plotEnrichment(hallmakers_set[["HALLMARK_TNFA_SIGNALING_VIA_NFKB"]], rank_HSC,)+labs(title="HALLMARK_TNFA_SIGNALING_VIA_NFKB")
plotEnrichment(hallmakers_set[["HALLMARK_APOPTOSIS"]], rank_HSC,)+labs(title="HALLMARK_APOPTOSIS")

MDS_HSC_C2 <- fgsea(C2_set, rank_HSC, minSize=15, maxSize = 500, nperm=10000)
head(MDS_HSC_C2[order(padj), ], n=10)
MDS_HSC_C2_sig<-MDS_HSC_C2%>%filter(padj<0.25)%>%arrange(desc(NES))

MDS_HSC_Hall_sig$leadingEdge[10]
MDS_HSC_Hall_sig$leadingEdge[6]

rank_LMPP<-  -sign(MDS_NBM_LMPP_deseq$log2FoldChange)*log10(MDS_NBM_LMPP_deseq$pvalue)
names(rank_LMPP)<-row.names(MDS_NBM_LMPP_deseq)
rank_LMPP<-rank_LMPP[!is.na(rank_LMPP)]
MT_NBM_LMPP_Hall <- fgsea(hallmakers_set, rank_LMPP, minSize=15, maxSize = 500, nperm=10000)
MT_NBM_LMPP_Hall_sig<-MT_NBM_Mono_Hall%>%filter(padj<0.25)%>%arrange(desc(NES))


rank_LMPP<-  -sign(MDS_NBM_Mono_deseq$log2FoldChange)*log10(MDS_NBM_Mono_deseq$pvalue)
names(rank_Mono)<-row.names(MDS_NBM_Mono_deseq)
rank_Mono<-rank_Mono[!is.na(rank_Mono)]
MT_NBM_Mono_Hall <- fgsea(hallmakers_set, rank_Mono, minSize=15, maxSize = 500, nperm=10000)
MT_NBM_Mono_Hall_sig<-MT_NBM_Mono_Hall%>%filter(padj<0.25)%>%arrange(desc(NES))




DefaultAssay(MDS_NBM_HSPC_M)<-"RNA"
MDS_NBM_H_M_score<-AddModuleScore(MDS_NBM_HSPC_M,features=hallmakers_set[1],name="NFkB_via_TNF")
MDS_NBM_H_M_score<-AddModuleScore(MDS_NBM_H_M_score,features=hallmakers_set[10],name="Apoptosis")
MDS_NBM_H_M_score<-AddModuleScore(MDS_NBM_H_M_score,features=hallmakers_set[37],name="TP53_signaling")
MDS_NBM_H_M_score<-AddModuleScore(MDS_NBM_H_M_score,features=hallmakers_set["HALLMARK_INTERFERON_GAMMA_RESPONSE"],name="Interferon_gamma_responses")
MDS_NBM_H_M_score<-AddModuleScore(MDS_NBM_H_M_score,features=hallmakers_set["HALLMARK_G2M_CHECKPOINT"],name="G2M_checkpoint")

MDS_NBM_H_M_score$Group<-factor(MDS_NBM_H_M_score$Group,levels=c("NBM","MDS"))
VlnPlot(MDS_NBM_H_M_score,features=c("Apoptosis1"), split.by = "Sample", pt.size = 0, 
        cols=c(rep("tomato4",5),rep("steelblue4",4)))
VlnPlot(MDS_NBM_H_M_score,features=c("NFKBIA","GADD45A","PMAIP1","BID","IRF1","IFITM3"), split.by = "Sample", pt.size = 0,
  cols=c(rep("tomato4",4),rep("steelblue4",4)),ncol=3)
VlnPlot(MDS_NBM_H_M_score,features=c("HLF"), split.by = "Sample", pt.size = 0)

MDS_NBM_HSPC_M_pseud@meta.data%>%ggplot(aes(Myelo_pseudo,GMP1,color=Group))+geom_smooth(se=F)+theme_classic()+
  scale_color_manual(values=c(rep("steelblue4",1),rep("tomato4",1)))


MDS_NBM_HSPC_M_pseud@meta.data%>%mutate(gene=MDS_NBM_HSPC_M_pseud@assays$RNA@counts["CD38",])%>%ggplot(aes(Myelo_pseudo,gene,color=Group))+geom_smooth(se=F)+theme_classic()+
  scale_color_manual(values=c(rep("steelblue4",1),rep("tomato4",1)))

#MT vs WT
Idents(MDS_NBM_HSPC)<-"HSPC_type"

MDS_NBM_HSPC$SF3B1[is.na(MDS_NBM_HSPC$SF3B1)]<-"Unknown"
MDS_NBM_HSPC$TET2[is.na(MDS_NBM_HSPC$TET2)]<-"Unknown"
MDS_NBM_HSPC$DNMT31[is.na(MDS_NBM_HSPC$DNMT31)]<-"Unknown"

MDS_NBM_HSPC[["Mutation"]]<-"Unknown"
MDS_NBM_HSPC[["Mutation"]]<-ifelse(MDS_NBM_HSPC$SF3B1=="Mutant","Mutant",MDS_NBM_HSPC$Mutation)
MDS_NBM_HSPC[["Mutation"]]<-ifelse(MDS_NBM_HSPC$TET2=="Mutant","Mutant",MDS_NBM_HSPC$Mutation)
MDS_NBM_HSPC[["Mutation"]]<-ifelse(MDS_NBM_HSPC$DNMT31=="Mutant","Mutant",MDS_NBM_HSPC$Mutation)


MDS_HSPC<-subset(MDS_NBM_HSPC,cells=names(MDS_NBM_HSPC$Mutant_state[MDS_NBM_HSPC$Mutant_state%in% c("NBM")]),invert=T)


MDS_HSPC@meta.data%>%group_by(Sample,Mutant_state,HSPC_type)%>%summarize(n=n())%>%
  group_by(Sample,Mutant_state)%>%summarize(HSPC_type=HSPC_type, per=n/sum(n))%>%
  ggplot(aes(Mutant_state,per,color=Mutant_state))+geom_boxplot(outlier.size = -1)+geom_jitter(width=0.25)+
  theme_classic()+facet_wrap(.~HSPC_type)

#MDS_HSPC<-subset(MDS_HSPC,cells=names(MDS_HSPC$Sample[MDS_HSPC$Sample%in% c("MDS_SF3B1")]),invert=T)
Idents(MDS_HSPC)<-"Sample"
VlnPlot(MDS_HSPC,features=c("NFKBIA","GADD45A","PMAIP1","BID","IRF1","IFITM3"), split.by = "Mutant_state", pt.size = 0,
        ,ncol=3,assay="RNA")&theme(legend.position = "right",axis.title=element_blank())

VlnPlot(MDS_HSPC,features=c("NFIL3","KLF2","BCL2A1","CFLAR","GADD45A"), split.by = "Mutant_state", pt.size = 0,
        ,ncol=3,assay="RNA")&theme(legend.position = "right",axis.title=element_blank())

MDS_HSPC_deseq<-run_pseudobulkDE(meta_data=MDS_HSPC@meta.data,expr_mat=MDS_HSPC@assays$RNA@counts,
                                    contrasts='condition,Mutant,Unknown',design='~condition+Extra',
                                    sample_col="Sample", condition_col="Mutation",cluster_col="type",extr_col="Sample",do_prefiltering = TRUE,use_celltype =F)
MDS_HSPC_deseq["Gene"]<-row.names(MDS_HSPC_deseq)


MDS_HSPC_deseq$pvalue[MDS_HSPC_deseq$pvalue==0]<- 1e-103
rank_MDS_HSPC<-  -sign(MDS_HSPC_deseq$log2FoldChange)*log10(MDS_HSPC_deseq$pvalue)
names(rank_MDS_HSPC)<-row.names(MDS_HSPC_deseq)
rank_MDS_HSPC<-rank_MDS_HSPC[!is.na(rank_MDS_HSPC)]

MDS_HSPC_Hall <- fgsea(hallmakers_set, rank_MDS_HSPC, minSize=15, maxSize = 500, nperm=10000)
head(MDS_HSPC_Hall[order(padj), ], n=10)
MDS_HSPC_Hall_sig<-MDS_HSPC_Hall%>%filter(padj<0.25)%>%arrange(NES)
MDS_HSPC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSPC_Hall_sig$NES<0,"Bulk","Mutant")



MDS_HSPC_Hall_sig%>%mutate(pathway=factor(pathway, levels=MDS_HSPC_Hall_sig$pathway))%>%group_by(Enrichment)%>%
  top_n(10,wt=abs(NES))%>%ggplot(aes(NES,pathway))+geom_col()+theme_classic()

MDS_HSPC_Hall_sig[1]$leadingEdge[[1]]

MDS_H_Mpro_C2 <- fgsea(C2_set, rank_H_Mpro, minSize=15, maxSize = 500, nperm=10000)
head(MDS_H_Mpro_C2[order(padj), ], n=10)
MDS_H_Mpro_C2_sig<-MDS_H_Mpro_C2%>%filter(padj<0.05)%>%arrange(NES)
MDS_H_Mpro_C2_sig[["Enrichment"]]<-ifelse(MDS_H_Mpro_C2_sig$NES<0,"Bulk","Mutant")

plotEnrichment(hallmakers_set[["HALLMARK_TNFA_SIGNALING_VIA_NFKB"]], rank_MDS_HSPC,)+labs(title="HALLMARK_TNFA_SIGNALING_VIA_NFKB")
plotEnrichment(C2_set[["WP_APOPTOSIS"]], rank_MDS_HSPC,)+labs(title="WP_APOPTOSIS")


################
#GSEA score for whole HSPC+M
#############
DefaultAssay(MDS_NBM_HSPC_M)<-"RNA"


MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M,features=hallmakers_set[1],name="NFkB_via_TNF")
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score,features=hallmakers_set[10],name="Apoptosis")
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score,features=hallmakers_set[37],name="TP53_signaling")
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score,features=hallmakers_set["HALLMARK_INTERFERON_GAMMA_RESPONSE"],name="IFN_response")
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score,features=hallmakers_set["HALLMARK_HEME_METABOLISM"],name="HEME_metabolism")


VlnPlot(MDS_NBM_HSPC_M_score,features=c("NFkB_via_TNF1", "Apoptosis1","TP53_signaling1"), split.by = "Sample", pt.size = 0, 
        cols=c(rep("tomato4",5),rep("steelblue4",4)))

VlnPlot(MDS_NBM_HSPC_M_score,features=c("IFN_response1", "HEME_metabolism1"), split.by = "Sample", pt.size = 0, 
        cols=c(rep("tomato4",5),rep("steelblue4",4)))

geneset="NFkB_via_TNF1"
Heatmap_geneset<-function(data=MDS_NBM_HSPC_M_score, geneset="NFkB_via_TNF1"){
  data<-MDS_NBM_HSPC_M_score@meta.data[c("Group","Sample",geneset,"HSPC_type")]
  names(data)<-c("Group","Sample","geneset","HSPC_type")
  data<-data%>%group_by(Sample,HSPC_type)%>%summarise(meanScore=mean(geneset))
  data["Group"]<-ifelse(grepl("MDS",data$Sample),"MDS","NBM")
  data$Sample<-factor(data$Sample,levels=c("NBM1","NBM2","NBM3","NBM4","MDS_10209","MDS_124","MDS_337","MDS_345"))
  data<-data%>%group_by(HSPC_type)%>%summarise(Sample=Sample,ScaledScore=scale(meanScore))
  data<-data%>%filter(!HSPC_type%in%c("MDS_Progs","CD16+ Mono"))
  data%>%ggplot(aes(Sample,HSPC_type,fill=ScaledScore))+geom_raster()+
    scale_fill_gradient(low = "blue", high = "red")+theme_classic()+labs(title=geneset)+
    theme(axis.line = element_blank(), axis.text.x=element_text(angle=45, hjust=1,vjust=1))
  }
Heatmap_geneset(data=MDS_NBM_HSPC_M_score, geneset="IFN_response1")+theme(axis.title=element_blank())

VlnPlot(MDS_NBM_HSPC_M_score,features=c("NFkB_via_TNF1","Apoptosis1","TP53_signaling1"), split.by = "Group", pt.size = 0,
        cols=c("steelblue4","tomato4"),ncol=3)&theme(axis.title =element_blank())

VlnPlot(MDS_NBM_HSPC_M_score,features=c("NFKBIA","IRF1","IFITM3","GADD45A","PMAIP1","BID"), split.by = "Group", pt.size = 0,
        cols=c("steelblue4","tomato4"),ncol=3)&theme(axis.title =element_blank())



#################################
#HSC
############
MDS_NBM_HSC<-subset(MDS_NBM_HSPC_M,ident=c("HSC/MPPs"))


DefaultAssay(MDS_NBM_HSC)<-"integrated"
DimPlot(MDS_NBM_HSC, reduction = "umap", label=F,pt.size = 0.25,split.by = "Sample",ncol=4)#&theme(axis.title = element_blank())

MDS_NBM_HSC<-subset(MDS_NBM_HSC,ident=c("3"),invert=T)


MDS_NBM_HSC <- ScaleData(MDS_NBM_HSC,vars.to.regress = c("S.Score", "G2M.Score"));
FindVariableFeatures(MDS_NBM_HSC)
MDS_NBM_HSC <- RunPCA(MDS_NBM_HSC, npcs = 30,verbose = FALSE);

ElbowPlot(MDS_NBM_HSC)
nDims=20
MDS_NBM_HSC <- FindNeighbors(MDS_NBM_HSC, reduction = "pca", dims = 1:nDims)
MDS_NBM_HSC <- FindClusters(MDS_NBM_HSC, resolution = 0.3);
MDS_NBM_HSC <- RunUMAP(MDS_NBM_HSC,dims = 1:nDims);
MDS_NBM_HSC$Group<-factor(MDS_NBM_HSC$Group,levels=c("NBM","MDS"))
MDS_NBM_HSC$Sample<-factor(MDS_NBM_HSC$Sample,levels=c("NBM1","NBM2","NBM3","NBM4","MDS_10209","MDS_124","MDS_337","MDS_345","MDS_SF3B1"))
DimPlot(MDS_NBM_HSC, reduction = "umap", label=F,pt.size = 0.25,split.by = "Group",ncol=2)#&theme(axis.title = element_blank())


Idents(MDS_NBM_HSC)<-"seurat_clusters"

  
MDS_NBM_HSC[["HSC_cluster"]]<-MDS_NBM_HSC$seurat_clusters
MDS_NBM_HSC$HSC_cluster<-ifelse(MDS_NBM_HSC$seurat_clusters=="0","Low-output HSC",MDS_NBM_HSC$seurat_clusters)
MDS_NBM_HSC$HSC_cluster<-ifelse(MDS_NBM_HSC$seurat_clusters=="1","High-output HSC2",MDS_NBM_HSC$HSC_cluster)
MDS_NBM_HSC$HSC_cluster<-ifelse(MDS_NBM_HSC$seurat_clusters %in% c("2"),"High-output HSC1",MDS_NBM_HSC$HSC_cluster)

#MDS_NBM_HSC$HSC_cluster<-ifelse(MDS_NBM_HSC$seurat_clusters=="3","HSC3",MDS_NBM_HSC$HSC_cluster)


MDS_NBM_HSC$HSC_cluster<-factor(MDS_NBM_HSC$HSC_cluster,levels=c("Low-output HSC","High-output HSC1","High-output HSC2","High-output HSC3"))

Idents(MDS_NBM_HSC)<-"HSC_cluster"

#MDS_NBM_HSC<-subset(MDS_NBM_HSC,ident="HSC3",invert=T)
DimPlot(MDS_NBM_HSC, reduction = "umap", label=F,pt.size = 0.25,split.by  = "Mutation",ncol=2)#&theme(axis.title = element_blank())
DimPlot(MDS_NBM_HSC, reduction = "umap", label=F,pt.size = 0.25,split.by  = "Group",ncol=2,cols=c("grey40","darkgreen","tomato4"))#&theme(axis.title = element_blank())
DimPlot(subset(MDS_NBM_HSC,subset=Group=="NBM"), reduction = "umap", label=F,cols=c("grey40","darkgreen","tomato4"))

MDS_NBM_HSC$UMAP_1<-MDS_NBM_HSC@reductions$umap@cell.embeddings[,1]
MDS_NBM_HSC$UMAP_2<-MDS_NBM_HSC@reductions$umap@cell.embeddings[,2]


saveRDS(MDS_NBM_HSC,"~/single_cell_sequencing/MDS single cell/MDS_NBM_HSC.RDS")

MDS_NBM_HSC<-readRDS("scRNAseq_dataset/RDSfile/chenMDS/MDS single cell//MDS_NBM_HSC.RDS")


DimPlot(MDS_NBM_HSC, reduction = "umap", label=F,pt.size = 0.25, cols=c("grey40","darkgreen","tomato4"))
FeaturePlot(MDS_NBM_HSC,features=c("CD34"))

MDS_NBM_HSC$Mutation<-ifelse(MDS_NBM_HSC$SF3B1=="Mutant","Mutant","Unknown")
MDS_NBM_HSC$Mutation<-ifelse(MDS_NBM_HSC$TET2=="Mutant","Mutant",MDS_NBM_HSC$Mutation)


DimPlot(subset(MDS_NBM_HSC,subset=Group=="MDS"), reduction = "umap", cells.highlight = names(MDS_NBM_HSC$Mutant_state[MDS_NBM_HSC$Mutant_state=="Mutant"]),
          cols.highlight = "tomato4",label=F)#&theme(axis.title = element_blank())


DimPlot(subset(MDS_NBM_HSC,cells=names(MDS_NBM_HSC$Mutant_state[!is.na(MDS_NBM_HSC$Mutant_state)])), reduction = "umap", pt.size=1,label=F,split.by = "Sample")#&theme(axis.title = element_blank())

Phase<-MDS_NBM_HSC@meta.data%>%group_by(Sample,Group,Phase,HSC_cluster)%>%summarise(n=n())%>%group_by(Sample,Group,HSC_cluster)%>%
 summarise(Phase=Phase,fraction=n/sum(n))

FeaturePlot(MDS_NBM_HSC,features=c("HLF","HES1","THY1"),min.cutoff = 0.25,split.by = "Group",sort.cell = T)

FeaturePlot(MDS_NBM_HSC,features=c("CD74","NR4A2","CDK6","MKI67"),min.cutoff = 0.25,split.by = "Group",sort.cell = T)
FeaturePlot(MDS_NBM_HSC,features=c("PROM1","CD244"),min.cutoff = 0,split.by = "Group",sort.cell = T)



FeaturePlot(subset(MDS_NBM_HSC, cells=c(names(MDS_NBM_HSC$Sample)[c(MDS_NBM_HSC$Group=="MDS")], 
                                                                  sample(names(MDS_NBM_HSC$Sample)[c(MDS_NBM_HSC$Group=="NBM")],908,replace=F))),
            features=c("CDKN1A","GADD45A","GADD45B"),min.cutoff = 0.25,split.by = "Group",sort.cell = T)

VlnPlot(MDS_NBM_HSC,split.by = "Sample",cols=c(rep("tomato4",4),rep("steelblue4",4)),features=c("Proliferation_score"))&theme(axis.title=element_blank())

DotPlot(MDS_NBM_HSC,split.by = "Group",cols=rep("tomato4",10),assay="RNA",features=c("HLF","MEIS1","HES1"))&theme(axis.title=element_blank())
DotPlot(MDS_NBM_HSC,split.by = "Group",cols=rep("tomato4",10),assay="RNA",features=c("AVP","CD74","NR4A2","CDK6","IGFBP7"))&theme(axis.title=element_blank())
DotPlot(MDS_NBM_HSC,split.by = "Group",cols=rep("tomato4",10),assay="RNA",features=c("HLA-DRA","CRHBP"))&theme(axis.title=element_blank())
VlnPlot(MDS_NBM_HSC,split.by = "Group",cols=c(rep("steelblue4",1),rep("tomato4",1)),features=c("ETS2"))&theme(axis.title=element_blank())

Idents(MDS_NBM_HSC)<-"HSC_cluster"


HSC_sub_proportion<-MDS_NBM_HSC@meta.data%>%group_by(HSC_cluster, Sample)%>%summarize(n=n())%>%group_by(Sample)%>%
  summarize(HSC_cluster=HSC_cluster, Percentage=n/sum(n)*100)
HSC_sub_proportion["Group"]<-ifelse(grepl("NBM",HSC_sub_proportion$Sample), "NBM", "MDS")
HSC_sub_proportion$Group<-factor(HSC_sub_proportion$Group,levels=c("NBM","MDS"))


HSC_sub_proportion$HSC_cluster<-factor(HSC_sub_proportion$HSC_cluster,levels=c("Low-output HSC","High-output HSC1","High-output HSC2"))

HSC_sub_proportion%>%ggplot(aes(Sample,Percentage, fill=HSC_cluster))+geom_bar(stat="identity")+theme_classic()+
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1))+scale_fill_manual(values=c("grey40","darkgreen","tomato4"))

Plot<-HSC_sub_proportion%>%ggplot(aes(Group,Percentage, col=Group))+geom_boxplot(outlier.size= -1, width=0.5)+geom_jitter(width=0.2)+
  facet_wrap(.~HSC_cluster,ncol=4)+theme_classic()+scale_color_manual(values=c("steelblue4","tomato4"))+
  theme(strip.background = element_blank(),strip.text = element_text(size=12))
Plot+stat_compare_means(method = "t.test")

HSC_markers<-FindAllMarkers(subset(MDS_NBM_HSC,cells=names(MDS_NBM_HSC$Group[MDS_NBM_HSC$Group=="NBM"])),assay="RNA", only.pos = T)




Idents(MDS_NBM_HSC)<-"HSC_cluster"
DotPlot(MDS_NBM_HSC,split.by = "Group",cols=rep("tomato4",10),features=c("CDKN1A","GADD45A","GADD45B"))

FeaturePlot(MDS_NBM_HSC, reduction = "umap", 
            features=c("HLF","ADGRG6","PCDH9"), label=F,pt.size = 0.25,sort.cell = T,min.cutoff =0,split.by = "Group") #HSC markers from Danilo et al, NC, 2019
FeaturePlot(MDS_NBM_HSC, reduction = "umap", 
            features=c("CRHBP","DUSP1"), label=F,pt.size = 0.25,order = T,min.cutoff =0,split.by = "Group") #HSC markers from Danilo et al, NC, 2019

FeaturePlot(MDS_NBM_HSC, reduction = "umap", 
            features=c("CDK6","HBB"), label=F,pt.size = 0.25,sort.cell = T,min.cutoff =0.25,split.by = "Group") #HSC markers from Danilo et al, NC, 2019




FeaturePlot(MDS_NBM_HSC, reduction = "umap", 
            features=c("CSF3R"), label=F,pt.size = 0.25,sort.cell = T,min.cutoff =0.25,split.by = "Group") #HSC markers from Danilo et al, NC, 2019


FeaturePlot(MDS_NBM_HSC, reduction = "umap", 
            features=c("AVP","HLF"), label=F,pt.size = 0.25,sort.cell = T,min.cutoff =0,split.by = "Group")

VlnPlot(MDS_NBM_HSC, features=c("LSP1"),split.by = "Sample",pt.size = 0,assay="RNA")


MDS_NBM_H_deseq<-run_pseudobulkDE_new(data=MDS_NBM_HSC,
                                    contrasts='condition,MDS,NBM',design='~condition',
                                    sample_col="Sample", condition_col="Group",cluster_col="HSPC_type",extr_col=NULL,do_prefiltering = TRUE)


MDS_NBM_H_deseq["Gene"]<-row.names(MDS_NBM_H_deseq)
MDS_NBM_H_deseq[MDS_NBM_H_deseq$Gene %in% c("NFKBIA","RELB","TXNIP","RIPK2","GADD45B","ATF3"),]

MDS_NBM_H_deseq["rank"]<- -sign(MDS_NBM_H_deseq$log2FoldChange)*log2(MDS_NBM_H_deseq$pvalue)

write_xlsx(MDS_NBM_H_deseq,"single_cell_sequencing/MDS single cell/HSC_MDS_vs_NBM.xlsx")

MDS_NBM_H_deseq_sig<-MDS_NBM_H_deseq[which(MDS_NBM_H_deseq$padj<0.05 & abs(MDS_NBM_H_deseq$log2FoldChange)>0.585),]
MDS_NBM_H_deseq_top<-MDS_NBM_H_deseq_sig%>%mutate(gene_change=case_when(.$log2FoldChange>0~"Up", TRUE~"Down"))

MDS_NBM_H_deseq%>%ggplot(aes(log2FoldChange,-log10(padj)))+geom_point(size=0.1,col="grey")+
  geom_point(data=MDS_NBM_H_deseq_sig,mapping=aes(log2FoldChange,-log10(padj)),col="grey40",size=0.5)+
  geom_vline(xintercept= c(0.5,-0.5),col="grey",type=2)+geom_hline(yintercept = -log10(0.05),col="grey",type=2)+
  geom_text_repel(MDS_NBM_H_deseq_top,mapping=aes(log2FoldChange,-log10(padj),label=Gene,col=gene_change),max.overlaps = 30)+
  theme_bw()+scale_color_manual(values=c("steelblue4","tomato4"))

rank_HSC<-MDS_NBM_H_deseq$rank
names(rank_HSC)<-MDS_NBM_H_deseq$Gene
rank_HSC<-rank_HSC[is.finite(rank_HSC)]
MDS_HSC_Hall <- fgsea(hallmakers_set, rank_HSC, minSize=15, maxSize = 500, nperm=10000)
head(MDS_HSC_Hall[order(padj), ], n=10)
MDS_HSC_Hall_sig<-MDS_HSC_Hall%>%filter(padj<0.05)%>%arrange(NES)
MDS_HSC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSC_Hall_sig$NES<0,"NBM","MDS")
MDS_HSC_Hall_sig[MDS_HSC_Hall_sig$pathway=="HALLMARK_APOPTOSIS"]$leadingEdge
MDS_HSC_Hall_sig%>%mutate(pathway=factor(pathway, levels=MDS_HSC_Hall_sig$pathway))%>%group_by(Enrichment)%>%
  top_n(15,wt=NES)%>%ggplot(aes(NES,pathway,fill= padj))+geom_col()+scale_fill_gradient(low="red",high="blue")+theme_classic()



MDS_HSC_C2 <- fgsea(C2_set, rank_HSC, minSize=15, maxSize = 500, nperm=10000)
head(MDS_HSC_C2[order(padj), ], n=10)
MDS_HSC_C2_sig<-MDS_HSC_C2%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_HSC_C2_sig[["Enrichment"]]<-ifelse(MDS_HSC_C2_sig$NES<0,"NBM","MDS")

MDS_HSC_C5 <- fgsea(C5_set, rank_HSC, minSize=15, maxSize = 500, nperm=10000)
head(MDS_HSC_C5[order(padj), ], n=10)
MDS_HSC_C5_sig<-MDS_HSC_C5%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_HSC_C5_sig[["Enrichment"]]<-ifelse(MDS_HSC_C5_sig$NES<0,"NBM","MDS")


MDS_HSC_output <- fgsea(HSC_output_set, rank_HSC, minSize=15, maxSize = 500, nperm=1000)


plotEnrichment(C2_set[["GOLDRATH_HOMEOSTATIC_PROLIFERATION"]], rank_HSC)+labs(title="GOLDRATH_hema_proliferation")
plotEnrichment(HSC_output_set[["high_output"]], rank_HSC)+labs(title="High-output HSC")

plotEnrichment(C2_set[["GRAHAM_NORMAL_QUIESCENT_VS_NORMAL_DIVIDING_UP"]], rank_HSC)+labs(title="QUIESCENT_VS_NORMAL_DIVIDING_UP")


DefaultAssay(MDS_NBM_HSC)<-"RNA"
Idents(MDS_NBM_HSC_score)<-"HSC_cluster"
MDS_NBM_HSC_score<-AddModuleScore(MDS_NBM_HSC,features=hallmakers_set["HALLMARK_TNFA_SIGNALING_VIA_NFKB"], name="NFkB_signaling")
MDS_NBM_HSC_score<-AddModuleScore(MDS_NBM_HSC_score,features=hallmakers_set["HALLMARK_P53_PATHWAY"], name="P53_pathway")

MDS_NBM_HSC_score<-AddModuleScore(MDS_NBM_HSC_score,features=hallmakers_set["HALLMARK_APOPTOSIS"], name="Apoptosis")
MDS_NBM_HSC_score<-AddModuleScore(MDS_NBM_HSC_score,features=hallmakers_set["HALLMARK_INTERFERON_GAMMA_RESPONSE"], name="Interferon_signaling")
MDS_NBM_HSC_score<-AddModuleScore(MDS_NBM_HSC_score,features=hallmakers_set["HALLMARK_G2M_CHECKPOINT"], name="G2M_checkpoint")

MDS_NBM_HSC_score<-AddModuleScore(MDS_NBM_HSC_score,features=HSC_output_set["low_output"], name="low_output")
MDS_NBM_HSC_score<-AddModuleScore(MDS_NBM_HSC_score,features=HSC_output_set["high_output"], name="high_output")

VlnPlot(MDS_NBM_HSC_score,features=c("low_output1","high_output1"),split.by = "Group",pt.size=0.1,cols=c("steelblue4","tomato3"))

VlnPlot(MDS_NBM_HSC_score,features=c("NFkB_signaling1","Apoptosis1"),split.by = "Group",pt.size=0)

VlnPlot(MDS_NBM_HSC_score,features=c("NFkB_signaling1"),split.by = "Group",pt.size=0)
VlnPlot(MDS_NBM_HSC_score,features=c("high_output1","G2M_checkpoint1"),split.by = "Group",pt.size=0)
VlnPlot(MDS_NBM_HSC_score,features=c("Interferon_signaling1"),split.by = "Group",pt.size=0)

VlnPlot(MDS_NBM_HSC_score,features=c("NFkB_signaling1","Apoptosis1"),
        split.by = "Group",cols=c(rep("steelblue4",1),rep("tomato4",1)),pt.size=0.1)

VlnPlot(MDS_NBM_HSC_score,features=c("NFkB_signaling1","Apoptosis1","Interferon_signaling1"),group.by = "Sample",pt.size=0)

FeaturePlot(subset(MDS_NBM_HSC_score,subset=Group=="NBM"), reduction = "umap",
            features=c("low_output1"), order=T,min.cutoff =0.4) &theme(axis.title = element_blank())

FeaturePlot(subset(MDS_NBM_HSC_score,subset=Group=="NBM"), reduction = "umap",
            features=c("high_output1"), order=T,min.cutoff =0.25)&theme(axis.title = element_blank())



MDS_NBM_HSC_score$UMAP_1<-MDS_NBM_HSC_score@reductions$umap@cell.embeddings[,"UMAP_1"]
MDS_NBM_HSC_score$umap_2<-MDS_NBM_HSC_score@reductions$umap@cell.embeddings[,"UMAP_2"]


plot_density(MDS_NBM_HSC_score, c("low_output1"))



MDS_NBM_HSC_barcode<-MDS_NBM_HSC@meta.data
MDS_NBM_HSPC_M_barcode<-MDS_NBM_HSPC_M@meta.data
MDS_NBM_HSPC_M_barcode$HSPC_type<-as.character(MDS_NBM_HSPC_M_barcode$HSPC_type)
for (i in 1:nrow(MDS_NBM_HSPC_M_barcode)){
  MDS_NBM_HSPC_M_barcode$HSPC_type[i]<-if (row.names(MDS_NBM_HSPC_M_barcode)[i] %in% row.names(MDS_NBM_HSC_barcode)){
    as.character(MDS_NBM_HSC_barcode$HSC_cluster[row.names(MDS_NBM_HSC_barcode) == row.names(MDS_NBM_HSPC_M_barcode)[i]])
  }else{MDS_NBM_HSPC_M_barcode$HSPC_type[i]}
}


MDS_NBM_HSPC_M[["HSC_cluster"]]<-MDS_NBM_HSPC_M_barcode$HSPC_type
MDS_NBM_HSPC_M$HSC_cluster[MDS_NBM_HSPC_M$HSC_cluster=="HSC/MPPs"]<-"LMPPs"
Idents(MDS_NBM_HSPC_M)<-"HSC_cluster"
#MDS_NBM_HSC<-subset(MDS_NBM_HSC,ident="HSC3",invert=T)
DimPlot(MDS_NBM_HSPC_M, reduction = "umap", label=T,pt.size = 0.25,split.by = "Group",ncol=4)#&theme(axis.title = element_blank())





RaCH_list<-list.files("scRNAseq data transfer Lanpeng Chen/LRMDS/Mutant cell barcodes/RaCHseq",full.names = T)
RaCH_name<-list.files("scRNAseq data transfer Lanpeng Chen/LRMDS/Mutant cell barcodes/RaCHseq")
names(RaCH_list)<-RaCH_name

RaCH_list<-lapply(RaCH_list, function(x){
  x=data.frame(read.table(x))})

for (i in 1:length(RaCH_list)){
  RaCH_list[[i]]["bar_code"]<- tolower(paste0(sub("_pos.txt", replacement="",RaCH_name)[i],"_",RaCH_list[[i]]$V1,"_1"))
  RaCH_list[[i]]<-RaCH_list[[i]]["bar_code"]
}


RaCH_df<-unlist(RaCH_list)

MDS_NBM_HSC[["RaCH"]]<-"Unknown"


MDS_NBM_HSC$RaCH<-ifelse(MDS_NBM_HSC$bar_code %in% RaCH_df, "SF3B1m", "Unknown")

DimPlot(subset(MDS_NBM_HSC,subset=Group=="MDS"), label=T,split.by = "Group",reduction = "umap", cells.highlight = names(MDS_NBM_all_mt$RaCH[MDS_NBM_all_mt$RaCH=="SF3B1m"]), cols.highlight = "tomato4",repel = T)



#HSC NBM

NBM_HSC<-subset(MDS_NBM_HSC,cells=names(MDS_NBM_HSC$Group[MDS_NBM_HSC$Group=="NBM"]))

DimPlot(NBM_HSC, reduction = "umap", label=F,pt.size = 0.25,split.by = "Group",ncol=4)#&theme(axis.title = element_blank())

HSC_markers<-FindAllMarkers(NBM_HSC, only.pos = T, assay="RNA")
HSC_markers<-HSC_markers%>%filter(p_val_adj<0.05)
View(HSC_markers)


HSC_heatmap<-DotPlot(NBM_HSC, split.by = "Group",features=c("H2AFY","CDK6","MKI67","DNMT1","PLAC8","NKG7","STMN1","FLT3","NR4A1","NR4A2","NR4A3","TNFAIP3","HLA-DRA","HLA-DRB1","CD74","HLF","HES1","CRHBP","SOCS2","MLLT3"), assay="RNA")$data

HSC_heatmap%>%dplyr::rename("HSC_cluster"=id)%>%ggplot(aes(HSC_cluster,features.plot,fill=avg.exp.scaled))+
  geom_raster()+scale_fill_gradient(low="blue",high="red")+theme_classic()+labs(y="")+theme(axis.text.x=element_text(angle = 45,hjust=1,vjust=1))


NBM_HSC[["Cluster_STHSC1"]]<-ifelse(NBM_HSC$HSC_cluster=="High-output HSC1","STHSC1","Others")

NBM_STHSC_1_deseq<-run_pseudobulkDE(meta_data=NBM_HSC@meta.data,expr_mat=NBM_HSC@assays$RNA@counts,
                                  contrasts='condition,STHSC1,Others',design='~condition+Extra',
                                  sample_col="Sample", condition_col="Cluster_STHSC1",cluster_col="HSPC_type",extr_col="Sample",do_prefiltering = TRUE,use_celltype = F)


NBM_STHSC_1_deseq["Gene"]<-row.names(NBM_STHSC_1_deseq)
NBM_STHSC_1_deseq["rank"]<- -sign(NBM_STHSC_1_deseq$log2FoldChange)*log2(NBM_STHSC_1_deseq$pvalue)
#################
#Monocyte
##############
Idents(MDS_NBM_HSPC_M)<-"HSPC_type"
MDS_NBM_Monocyte<-subset(MDS_NBM_HSPC_M,ident=c("CD14+ Mono"))

DefaultAssay(MDS_NBM_Monocyte)<-"integrated"

MDS_NBM_Monocyte <- ScaleData(MDS_NBM_Monocyte);

MDS_NBM_Monocyte <- RunPCA(MDS_NBM_Monocyte, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_Monocyte)
nDims=10
MDS_NBM_Monocyte <- FindNeighbors(MDS_NBM_Monocyte)
MDS_NBM_Monocyte <- FindClusters(MDS_NBM_Monocyte, resolution = 0.3);
MDS_NBM_Monocyte <- RunUMAP(MDS_NBM_Monocyte,dims = 1:nDims);
DimPlot(MDS_NBM_Monocyte, reduction = "umap", label=F,pt.size = 0.25,split.by = "Sample")

Mono_markers<-FindAllMarkers(MDS_NBM_Monocyte,only.pos = T,assay="RNA")
Mono_markers_sig<-Mono_markers%>%filter(p_val_adj<0.05)

FeaturePlot(MDS_NBM_Monocyte,features=c("CD163","CD68","MRC1"),min.cutoff = 0.25,split.by = "Group",sort.cell = T) #M2 macropae
FeaturePlot(MDS_NBM_Monocyte,features=c("TNF"),min.cutoff = 0.25,split.by = "Group",sort.cell = T) #M1 macropae

DotPlot(MDS_NBM_Monocyte,group.by = "Sample",features=c("TLR4"))

MDS_NBM_Monocyte_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_Monocyte@meta.data,expr_mat=MDS_NBM_Monocyte@assays$RNA@counts,
                                  contrasts='condition,MDS,NBM',design='~condition',
                                  sample_col="Sample", condition_col="Group",cluster_col="type_re",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)

MDS_NBM_Monocyte_deseq["Gene"]<-row.names(MDS_NBM_Monocyte_deseq)
MDS_NBM_Monocyte_deseq["rank"]<- -sign(MDS_NBM_Monocyte_deseq$log2FoldChange)*log2(MDS_NBM_Monocyte_deseq$pvalue)

rank_Mono<-MDS_NBM_Monocyte_deseq$rank
names(rank_Mono)<-MDS_NBM_Monocyte_deseq$Gene
rank_Mono<-rank_Mono[is.finite(rank_Mono)]
MDS_Mono_Hall <- fgsea(hallmakers_set, rank_Mono, minSize=15, maxSize = 500, nperm=10000)
head(MDS_Mono_Hall[order(padj), ], n=10)
MDS_HSC_Hall_sig<-MDS_HSC_Hall%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_HSC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSC_Hall_sig$NES<0,"NBM","MDS")

MDS_Mono_C2 <- fgsea(C2_set, rank_Mono, minSize=15, maxSize = 500, nperm=10000)
head(MDS_Mono_C2[order(padj), ], n=10)
MDS_Mono_C2_sig<-MDS_Mono_C2%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_Mono_C2_sig[["Enrichment"]]<-ifelse(MDS_Mono_C2_sig$NES<0,"NBM","MDS")


MDS_Mono_C5 <- fgsea(C5_set, rank_Mono, minSize=15, maxSize = 500, nperm=10000)
head(MDS_Mono_C5[order(padj), ], n=10)
MDS_Mono_C5_sig<-MDS_Mono_C5%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_Mono_C5_sig[["Enrichment"]]<-ifelse(MDS_Mono_C5_sig$NES<0,"NBM","MDS")
MDS_Mono_C5_sig[56]$leadingEdge

M2_signature<-read.table("~/single_cell_sequencing/Gene_signatures_Monocytes_scRNAseq.txt",sep = "\t")
M2_signature<-M2_signature[,-c(2)]

M2_signature<-split(M2_signature,f=M2_signature$V1)

M2_signature<-lapply(M2_signature,function(x) as.character(x))
M2_signature<-lapply(M2_signature,function(x) x[x!=""])
M2_signature<-lapply(M2_signature,function(x) x[-1])


MDS_Mono_M2 <- fgsea(M2_signature, rank_Mono, minSize=15, maxSize = 500, nperm=10000)



##########
#DCs
##############
MDS_NBM_DC<-subset(MDS_NBM_HSPC_M,ident=c("DCs"))

DefaultAssay(MDS_NBM_DC)<-"integrated"

MDS_NBM_DC <- ScaleData(MDS_NBM_DC);

MDS_NBM_DC <- RunPCA(MDS_NBM_DC, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_DC)
nDims=6
MDS_NBM_DC <- FindNeighbors(MDS_NBM_DC)
MDS_NBM_DC <- FindClusters(MDS_NBM_DC, resolution = 0.3);
MDS_NBM_DC <- RunUMAP(MDS_NBM_DC,dims = 1:nDims);
DimPlot(MDS_NBM_DC, reduction = "umap", label=F,pt.size = 0.25,split.by = "Sample")
FeaturePlot(MDS_NBM_DC,features=c("IRAK4"),min.cutoff = 0.25,split.by = "Sample",sort.cell = T)
DotPlot(MDS_NBM_DC,split.by = "Sample",cols=rep("blue",10),features=c("CDKN1A","GADD45A","GADD45B"))


MDS_NBM_DC_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_DC@meta.data,expr_mat=MDS_NBM_DC@assays$RNA@counts,
                                  contrasts='condition,MDS,NBM',design='~condition',
                                  sample_col="Sample", condition_col="Group",cluster_col="HSPC_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)

MDS_NBM_DC_deseq["Gene"]<-row.names(MDS_NBM_DC_deseq)
MDS_NBM_DC_deseq["rank"]<- -sign(MDS_NBM_DC_deseq$log2FoldChange)*log2(MDS_NBM_DC_deseq$pvalue)

MDS_NBM_DC_deseq_sig<-MDS_NBM_DC_deseq[which(MDS_NBM_DC_deseq$padj<0.05),]
MDS_NBM_DC_deseq_top<-MDS_NBM_DC_deseq%>%mutate(gene_change=case_when(.$log2FoldChange>0~"Up", TRUE~"Down"))

MDS_NBM_H_deseq%>%ggplot(aes(log2FoldChange,-log10(padj)))+geom_point(size=0.1,col="grey")+
  geom_point(data=MDS_NBM_H_deseq_sig,mapping=aes(log2FoldChange,-log10(padj)),col="black",size=0.5)+
  geom_vline(xintercept= c(0.5,-0.5),col="grey",type=2)+geom_hline(yintercept = -log10(0.05),col="grey",type=2)+
  geom_text_repel(MDS_NBM_H_deseq_top,mapping=aes(log2FoldChange,-log10(padj),label=Gene),max.overlaps = 20)+
  theme_classic()


rank_HSC<-MDS_NBM_H_deseq$rank
names(rank_HSC)<-MDS_NBM_H_deseq$Gene
rank_HSC<-rank_HSC[is.finite(rank_HSC)]
MDS_HSC_Hall <- fgsea(hallmakers_set, rank_HSC, minSize=15, maxSize = 500, nperm=10000)
head(MDS_HSC_Hall[order(padj), ], n=10)
MDS_HSC_Hall_sig<-MDS_HSC_Hall%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_HSC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSC_Hall_sig$NES<0,"NBM","MDS")
MDS_HSC_Hall_sig[1]$leadingEdge
MDS_HSC_Hall_sig%>%mutate(pathway=factor(pathway, levels=MDS_HSC_Hall_sig$pathway))%>%group_by(Enrichment)%>%
  top_n(10,wt=NES)%>%ggplot(aes(NES,pathway,size= -log10(padj)))+geom_point()+theme_classic()

######
#HSPCs
################
MDS_NBM_HSPC<-subset(MDS_NBM_HSPC_M,ident=c("HSC/MPPs","LMPPs","CLPs","MEPs","GMPs"))


DefaultAssay(MDS_NBM_HSPC)<-"integrated"

FindVariableFeatures(MDS_NBM_HSPC)
MDS_NBM_HSPC <- ScaleData(MDS_NBM_HSPC,vars.to.regress = c("S.Score", "G2M.Score"));

MDS_NBM_HSPC <- RunPCA(MDS_NBM_HSPC, npcs = 30,verbose = FALSE);


ElbowPlot(MDS_NBM_HSPC)
nDims=15
MDS_NBM_HSPC <- FindNeighbors(MDS_NBM_HSPC, reduction = "pca", dims = 1:nDims)
MDS_NBM_HSPC <- FindClusters(MDS_NBM_HSPC, resolution = 0.4);
MDS_NBM_HSPC <- RunUMAP(MDS_NBM_HSPC,dims = 1:nDims);

Idents(MDS_NBM_HSPC)<-"HSPC_type"
Idents(MDS_NBM_HSPC)<-"Mutation"
Idents(MDS_NBM_HSPC)<-"seurat_clusters"
DimPlot(MDS_NBM_HSPC, reduction = "umap", split.by = "Group",ncol=2,label=T,pt.size = 0.25,repel = T)
MDS_NBM_HSPC$UMAP_1<-MDS_NBM_HSPC@reductions$umap@cell.embeddings[,"umap_1"]
MDS_NBM_HSPC$UMAP_2<-MDS_NBM_HSPC@reductions$umap@cell.embeddings[,"umap_2"]
#MDS_NBM_HSPC<-subset(MDS_NBM_HSPC,ident=c("9","10"),invert=T)

Idents(MDS_NBM_HSPC)<-"HSPC_type"
Idents(MDS_NBM_HSPC)<-"seurat_clusters"

#MDS_NBM_HSPC<-subset(MDS_NBM_HSPC,ident="7",invert=T)
MDS_NBM_HSPC<-readRDS("scRNAseq_dataset/RDSfile/chenMDS/MDS single cell/MDS_NBM_HSPC.RDS")

MDS_NBM_HSPC$Mutation<-ifelse(MDS_NBM_HSPC$SF3B1=="Mutant","Mutant","Unknown")
MDS_NBM_HSPC$Mutation<-ifelse(MDS_NBM_HSPC$TET2=="Mutant","Mutant",MDS_NBM_HSC$Mutation)


MDS_NBM_HSPC$Group<-factor(MDS_NBM_HSPC$Group,levels=c("NBM","MDS"))

DimPlot(MDS_NBM_HSPC, reduction = "umap", split.by = "Group",ncol=2,label=T,pt.size = 0.25,repel = T)

DimPlot(MDS_NBM_HSPC, reduction = "umap", split.by = "Group",ncol=2,label=T,pt.size = 0.25,repel = T,
        cols=c("grey","tan4","darkgreen","tomato3","steelblue4","tomato3"))


densityplot(data=MDS_NBM_HSPC,feature="AVP",assay="RNA",split.by ="Group")

DimPlot(subset(MDS_NBM_HSPC,subset=Group=="MDS"), reduction = "umap", cells.highlight = names(MDS_NBM_HSPC$Mutation[MDS_NBM_HSPC$Mutation=="Mutant"]), cols.highlight = "tomato4",ncol=2,label=F,pt.size = 0.25,repel = T)

DimPlot(subset(MDS_NBM_HSPC,subset=Mutation=="Mutant"), reduction = "umap", ncol=2,label=F,pt.size = 0.25,repel = T)


DefaultAssay(MDS_NBM_HSPC)<-"RNA"
FeaturePlot(MDS_NBM_HSPC, features=c("PTX3"),reduction = "umap", order = T ,label=F,min.cutoff=0.25,cols=c("grey","blue"),split.by = "Group")&
  theme(axis.title=element_blank())


VlnPlot(MDS_NBM_HSPC,ncol=3,pt.size = 0,features=c("NFKBIA","RIPK2","RELB","GADD45B","TXNIP","ATF3"),split.by = "Group",
        cols=c("steelblue4","tomato4"))&theme(axis.title = element_blank())

DotPlot(MDS_NBM_HSPC,features=c("NFKBIA","NFKBIE","RELB","GADD45B","ATF3","TXNIP"),split.by = "Group",cols=rep("tomato4",10))
DotPlot(MDS_NBM_HSPC,features=c("RNF217"),split.by = "Group",cols=rep("tomato4",10))

VlnPlot(MDS_NBM_HSPC,ncol=3,pt.size = 0,features=c("RNF217"),split.by = "Group",
        cols=c("steelblue4","tomato4"))&theme(axis.title = element_blank())

HSC_proportion<-MDS_NBM_HSPC@meta.data%>%group_by(Group,Sample, HSPC_type)%>%summarize(n=n())%>%group_by(Sample)%>%
  summarize(Group=Group,HSPC_type=HSPC_type,Perc=n/sum(n))

HSC_proportion%>%ggplot(aes(Sample,Perc,fill=HSPC_type))+geom_bar(stat="identity")

HSC_proportion%>%ggplot(aes(Group,Perc,color=Group))+geom_boxplot(outlier.size = -1)+
  geom_jitter()+facet_wrap(.~HSPC_type,ncol=5)+theme_classic()+theme(strip.background = element_blank(),strip.text =element_text(size=12))+
  scale_color_manual(values=c("steelblue4","tomato4"))+stat_compare_means(method="t.test",label =  "p.format")

MDS_NBM_HSPC_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_HSPC@meta.data,expr_mat=MDS_NBM_HSPC@assays$RNA@counts,
                                  contrasts='condition,MDS,NBM',design='~condition',
                                  sample_col="Sample", condition_col="Group",cluster_col="HSPC_type",extr_col=NULL,do_prefiltering = TRUE,use_celltype = F)

MDS_NBM_HSPC_deseq["Gene"]<-row.names(MDS_NBM_HSPC_deseq)
MDS_NBM_HSPC_deseq["rank"]<- -sign(MDS_NBM_HSPC_deseq$log2FoldChange)*log2(MDS_NBM_HSPC_deseq$pvalue)

write.xlsx(MDS_NBM_HSPC_deseq,"single_cell_sequencing/MDS single cell/HSPC_MDS_vs_NBM.xlsx")

MDS_NBM_HSPC_deseq[MDS_NBM_HSPC_deseq$Gene=="BCL10",]

Gene_list<-c("NFKBIA","RIPK2","RELB","GADD45B","TXNIP","ATF3","IFIT2","PTX3","SAT1")

MDS_NBM_deseq_gene<-MDS_NBM_HSPC_deseq[MDS_NBM_HSPC_deseq$Gene %in% Gene_list,]
MDS_NBM_deseq_gene<-MDS_NBM_deseq_gene[1:9]
MDS_NBM_deseq_gene<-data.frame(t(MDS_NBM_deseq_gene))
MDS_NBM_deseq_gene<-MDS_NBM_deseq_gene[,Gene_list]
MDS_NBM_deseq_gene["Group"]<-ifelse(grepl("NBM",row.names(MDS_NBM_deseq_gene)),"NBM","MDS")
MDS_NBM_deseq_gene$Group<-factor(MDS_NBM_deseq_gene$Group,levels=c("NBM","MDS"))
MDS_NBM_deseq_gene<-melt(MDS_NBM_deseq_gene,value.name = "Expression",id.vars="Group",variable.names="Gene")
MDS_NBM_deseq_gene%>%ggplot(aes(Group,Expression,color=Group))+geom_boxplot(outlier.size = -1)+geom_jitter(width=0.2)+
  theme_bw()+facet_wrap(.~variable,scales="free")+scale_color_manual(values=c("steelblue4","tomato4"))+theme(strip.background = element_blank())


rank_MDS_NBM<-MDS_NBM_HSPC_deseq$rank
names(rank_MDS_NBM)<-MDS_NBM_HSPC_deseq$Gene
rank_MDS_NBM<-rank_MDS_NBM[is.finite(rank_MDS_NBM)]
MDS_MDS_HSPC_Hall <- fgsea(hallmakers_set, rank_MDS_NBM, minSize=15, maxSize = 500, nperm=1000)
head(MDS_MDS_HSPC_Hall[order(padj), ], n=10)
MDS_MDS_HSPC_Hall_sig<-MDS_MDS_HSPC_Hall%>%filter(padj<0.25)%>%arrange(NES)
MDS_MDS_HSPC_Hall_sig[["Enrichment"]]<-ifelse(MDS_MDS_HSPC_Hall_sig$NES<0,"NBM","MDS")
MDS_MDS_HSPC_Hall_sig[MDS_MDS_HSPC_Hall_sig$pathway=="HALLMARK_APOPTOSIS"]$leadingEdge
MDS_MDS_HSPC_Hall_sig[MDS_MDS_HSPC_Hall_sig$pathway=="HALLMARK_TNFA_SIGNALING_VIA_NFKB"]$leadingEdge

MDS_MDS_HSPC_Hall_sig%>%mutate(pathway=factor(pathway, levels=MDS_MDS_HSPC_Hall_sig$pathway))%>%group_by(Enrichment)%>%
  top_n(10,wt=NES)%>%ggplot(aes(NES,pathway,fill=padj))+geom_bar(stat="identity")+theme_classic()+scale_fill_gradient(low="red",high="blue")

DefaultAssay(MDS_NBM_HSPC)<-"RNA"

MDS_NBM_HSPC_score<-AddModuleScore(MDS_NBM_HSPC,features=hallmakers_set[1],name="NFkB_via_TNF")
MDS_NBM_HSPC_score<-AddModuleScore(MDS_NBM_HSPC_score,features=hallmakers_set[10],name="Apoptosis")
MDS_NBM_HSPC_score<-AddModuleScore(MDS_NBM_HSPC_score,features=hallmakers_set[37],name="TP53_signaling")
MDS_NBM_HSPC_score<-AddModuleScore(MDS_NBM_HSPC_score,features=hallmakers_set["HALLMARK_INTERFERON_GAMMA_RESPONSE"],name="IFN_response")

geneset="NFkB_via_TNF1"
Heatmap_geneset<-function(data=MDS_NBM_HSPC_score, geneset="NFkB_via_TNF1"){
  data<-data@meta.data[c("Group","Sample",geneset,"HSPC_type")]
  names(data)<-c("Group","Sample","geneset","HSPC_type")
  data<-data%>%group_by(Sample,HSPC_type)%>%summarise(meanScore=mean(geneset))
  data["Group"]<-ifelse(grepl("MDS",data$Sample),"MDS","NBM")
  data$Sample<-factor(data$Sample,levels=c("NBM1","NBM2","NBM3","NBM4","MDS_10209","MDS_124","MDS_337","MDS_345","MDS_SF3B1"))
  data<-data%>%group_by(HSPC_type)%>%summarise(Sample=Sample,ScaledScore=scale(meanScore))
  data<-data%>%filter(!HSPC_type%in%c("MDS_Progs","CD16+ Mono"))
  data%>%ggplot(aes(Sample,HSPC_type,fill=ScaledScore))+geom_raster()+
    scale_fill_gradient(low = "blue", high = "red")+theme_classic()+labs(title=geneset)+
    theme(axis.line = element_blank(), axis.text.x=element_text(angle=45, hjust=1,vjust=1))
}
Heatmap_geneset(data=MDS_NBM_HSPC_score, geneset="NFkB_via_TNF1")+theme(axis.title=element_blank())+NoLegend()+
  Heatmap_geneset(data=MDS_NBM_HSPC_score, geneset="Apoptosis1")+theme(axis.title=element_blank())+NoLegend()+
  Heatmap_geneset(data=MDS_NBM_HSPC_score, geneset="TP53_signaling1")+theme(axis.title=element_blank())+NoLegend()+
  Heatmap_geneset(data=MDS_NBM_HSPC_score, geneset="IFN_response1")


########
#LMPPs

MDS_NBM_LMPP<-subset(MDS_NBM_HSPC_M,ident=c("LMPPs"))

DefaultAssay(MDS_NBM_LMPP)<-"integrated"

MDS_NBM_LMPP <- ScaleData(MDS_NBM_LMPP,vars.to.regress = c("S.Score", "G2M.Score"));

MDS_NBM_LMPP <- RunPCA(MDS_NBM_LMPP, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_LMPP)
nDims=7
MDS_NBM_LMPP <- FindNeighbors(MDS_NBM_LMPP)
MDS_NBM_LMPP <- FindClusters(MDS_NBM_LMPP, resolution = 0.3);
MDS_NBM_LMPP <- RunUMAP(MDS_NBM_LMPP,dims = 1:nDims);
DimPlot(MDS_NBM_LMPP, reduction = "umap", label=F,pt.size = 0.25,split.by = "Group")
FeaturePlot(MDS_NBM_LMPP,features=c("LYZ","CD34","CD33"),min.cutoff = 0.25,split.by = "Group",sort.cell = T)
DotPlot(MDS_NBM_LMPP,group.by = "Sample",features=c("IGFBP7"))

MDS_NBM_LMPP_deseq<-run_pseudobulkDE_new(data=MDS_NBM_LMPP,
                                        contrasts='condition,MDS,NBM',design='~condition',
                                        sample_col="Sample", condition_col="Group",cluster_col="HSPC_type",extr_col=NULL,do_prefiltering = TRUE)


MDS_NBM_LMPP_deseq["Gene"]<-row.names(MDS_NBM_LMPP_deseq)
MDS_NBM_LMPP_deseq[MDS_NBM_LMPP_deseq$Gene %in% c("NFKBIA","RELB","TXNIP","RIPK2","GADD45B","ATF3"),]

MDS_NBM_LMPP_deseq["rank"]<- -sign(MDS_NBM_LMPP_deseq$log2FoldChange)*log2(MDS_NBM_LMPP_deseq$pvalue)

rank_LMPP<-MDS_NBM_LMPP_deseq$rank
names(rank_LMPP)<-MDS_NBM_LMPP_deseq$Gene
rank_LMPP<-rank_LMPP[is.finite(rank_LMPP)]
MDS_LMPP_Hall <- fgsea(hallmakers_set, rank_LMPP, minSize=15, maxSize = 500, nperm=10000)

head(MDS_LMPP_Hall[order(padj), ], n=10)
MDS_LMPP_Hall_sig<-MDS_LMPP_Hall%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_LMPP_Hall_sig[["Enrichment"]]<-ifelse(MDS_LMPP_Hall_sig$NES<0,"NBM","MDS")

MDS_LMPP_Hall_sig[MDS_LMPP_Hall_sig$pathway=="HALLMARK_APOPTOSIS"]$leadingEdge

MDS_LMPP_C2 <- fgsea(C2_set, rank_LMPP, minSize=15, maxSize = 500, nperm=10000)
head(MDS_LMPP_C2[order(padj), ], n=10)
MDS_LMPP_C2_sig<-MDS_LMPP_C2%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_LMPP_C2_sig[["Enrichment"]]<-ifelse(MDS_LMPP_C2_sig$NES<0,"NBM","MDS")


MDS_LMPP_C5 <- fgsea(C5_set, rank_LMPP, minSize=15, maxSize = 500, nperm=10000)
head(MDS_LMPP_C5[order(padj), ], n=10)
MDS_LMPP_C5_sig<-MDS_LMPP_C5%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_LMPP_C5_sig[["Enrichment"]]<-ifelse(MDS_LMPP_C5_sig$NES<0,"NBM","MDS")
#
#MEPs

MDS_NBM_MEP<-subset(MDS_NBM_HSPC,ident=c("MEPs"))

DefaultAssay(MDS_NBM_MEP)<-"integrated"

MDS_NBM_MEP <- ScaleData(MDS_NBM_MEP,vars.to.regress = c("S.Score", "G2M.Score"));

MDS_NBM_MEP <- RunPCA(MDS_NBM_MEP, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_MEP)
nDims=7
MDS_NBM_MEP <- FindNeighbors(MDS_NBM_MEP)
MDS_NBM_MEP <- FindClusters(MDS_NBM_MEP, resolution = 0.3);
MDS_NBM_MEP <- RunUMAP(MDS_NBM_MEP,dims = 1:nDims);
DimPlot(MDS_NBM_MEP, reduction = "umap", label=F,pt.size = 0.25,split.by = "Group")
FeaturePlot(MDS_NBM_MEP,features=c("LYZ","CD34","CD33"),min.cutoff = 0.25,split.by = "Group",sort.cell = T)
DotPlot(MDS_NBM_MEP,group.by = "Sample",features=c("IGFBP7"))



MDS_NBM_MEP_deseq<-run_pseudobulkDE_new(data=MDS_NBM_MEP,
                                     contrasts='condition,MDS,NBM',design='~condition',
                                     sample_col="Sample", condition_col="Group",cluster_col="HSPC_type",extr_col=NULL,do_prefiltering = TRUE)


MDS_NBM_MEP_deseq["Gene"]<-row.names(MDS_NBM_MEP_deseq)
MDS_NBM_MEP_deseq[MDS_NBM_MEP_deseq$Gene %in% c("NFKBIA","RELB","TXNIP","RIPK2","GADD45B","ATF3"),]


MDS_NBM_MEP_deseq["Gene"]<-row.names(MDS_NBM_MEP_deseq)
MDS_NBM_MEP_deseq["rank"]<- -sign(MDS_NBM_MEP_deseq$log2FoldChange)*log2(MDS_NBM_MEP_deseq$pvalue)

rank_MEP<-MDS_NBM_MEP_deseq$rank
names(rank_MEP)<-MDS_NBM_MEP_deseq$Gene
rank_MEP<-rank_MEP[is.finite(rank_MEP)]
MDS_MEP_Hall <- fgsea(hallmakers_set, rank_MEP, minSize=15, maxSize = 500, nperm=10000)

head(MDS_MEP_Hall[order(padj), ], n=10)
MDS_MEP_Hall_sig<-MDS_MEP_Hall%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_MEP_Hall_sig[["Enrichment"]]<-ifelse(MDS_MEP_Hall_sig$NES<0,"NBM","MDS")

MDS_MEP_Hall_sig[MDS_MEP_Hall_sig$pathway=="HALLMARK_APOPTOSIS"]$leadingEdge

MDS_MEP_C2 <- fgsea(C2_set, rank_MEP, minSize=15, maxSize = 500, nperm=10000)
head(MDS_MEP_C2[order(padj), ], n=10)
MDS_MEP_C2_sig<-MDS_MEP_C2%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_MEP_C2_sig[["Enrichment"]]<-ifelse(MDS_MEP_C2_sig$NES<0,"NBM","MDS")


MDS_MEP_C5 <- fgsea(C5_set, rank_MEP, minSize=15, maxSize = 500, nperm=10000)
head(MDS_MEP_C5[order(padj), ], n=10)
MDS_MEP_C5_sig<-MDS_MEP_C5%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_MEP_C5_sig[["Enrichment"]]<-ifelse(MDS_MEP_C5_sig$NES<0,"NBM","MDS")

#CLP


MDS_NBM_CLP<-subset(MDS_NBM_HSPC,ident=c("CLPs"))

DefaultAssay(MDS_NBM_CLP)<-"integrated"

MDS_NBM_CLP <- ScaleData(MDS_NBM_CLP,vars.to.regress = c("S.Score", "G2M.Score"));

MDS_NBM_CLP <- RunPCA(MDS_NBM_CLP, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_CLP)
nDims=7
MDS_NBM_CLP <- FindNeighbors(MDS_NBM_CLP)
MDS_NBM_CLP <- FindClusters(MDS_NBM_CLP, resolution = 0.3);
MDS_NBM_CLP <- RunUMAP(MDS_NBM_CLP,dims = 1:nDims);
DimPlot(MDS_NBM_CLP, reduction = "umap", label=F,pt.size = 0.25,split.by = "Group")
FeaturePlot(MDS_NBM_CLP,features=c("LYZ","CD34","CD33"),min.cutoff = 0.25,split.by = "Group",sort.cell = T)
DotPlot(MDS_NBM_CLP,group.by = "Sample",features=c("IGFBP7"))


MDS_NBM_CLP_deseq<-run_pseudobulkDE_new(data=MDS_NBM_CLP,
                                        contrasts='condition,MDS,NBM',design='~condition',
                                        sample_col="Sample", condition_col="Group",cluster_col="HSPC_type",extr_col=NULL,do_prefiltering = TRUE)


MDS_NBM_CLP_deseq["Gene"]<-row.names(MDS_NBM_CLP_deseq)
MDS_NBM_CLP_deseq[MDS_NBM_CLP_deseq$Gene %in% c("NFKBIA","RELB","TXNIP","RIPK2","GADD45B","ATF3"),]




MDS_NBM_CLP_deseq["Gene"]<-row.names(MDS_NBM_CLP_deseq)
MDS_NBM_CLP_deseq["rank"]<- -sign(MDS_NBM_CLP_deseq$log2FoldChange)*log2(MDS_NBM_CLP_deseq$pvalue)

rank_CLP<-MDS_NBM_CLP_deseq$rank
names(rank_CLP)<-MDS_NBM_CLP_deseq$Gene
rank_CLP<-rank_CLP[is.finite(rank_CLP)]
MDS_CLP_Hall <- fgsea(hallmakers_set, rank_CLP, minSize=15, maxSize = 500, nperm=10000)

head(MDS_CLP_Hall[order(padj), ], n=10)
MDS_CLP_Hall_sig<-MDS_CLP_Hall%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_CLP_Hall_sig[["Enrichment"]]<-ifelse(MDS_CLP_Hall_sig$NES<0,"NBM","MDS")

MDS_CLP_Hall_sig[MDS_CLP_Hall_sig$pathway=="HALLMARK_APOPTOSIS"]$leadingEdge

MDS_CLP_C2 <- fgsea(C2_set, rank_CLP, minSize=15, maxSize = 500, nperm=10000)
head(MDS_CLP_C2[order(padj), ], n=10)
MDS_CLP_C2_sig<-MDS_CLP_C2%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_CLP_C2_sig[["Enrichment"]]<-ifelse(MDS_CLP_C2_sig$NES<0,"NBM","MDS")


MDS_CLP_C5 <- fgsea(C5_set, rank_CLP, minSize=15, maxSize = 500, nperm=10000)
head(MDS_CLP_C5[order(padj), ], n=10)
MDS_CLP_C5_sig<-MDS_CLP_C5%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_CLP_C5_sig[["Enrichment"]]<-ifelse(MDS_CLP_C5_sig$NES<0,"NBM","MDS")

#GMP

MDS_NBM_GMP<-subset(MDS_NBM_HSPC,ident=c("GMPs"))

DefaultAssay(MDS_NBM_GMP)<-"integrated"

MDS_NBM_GMP <- ScaleData(MDS_NBM_GMP,vars.to.regress =  c("S.Score", "G2M.Score"));

MDS_NBM_GMP <- RunPCA(MDS_NBM_GMP, npcs = 30,verbose = FALSE);
ElbowPlot(MDS_NBM_GMP)
nDims=10
MDS_NBM_GMP <- FindNeighbors(MDS_NBM_GMP)
MDS_NBM_GMP <- FindClusters(MDS_NBM_GMP, resolution = 0.3);
MDS_NBM_GMP <- RunUMAP(MDS_NBM_GMP,dims = 1:nDims);
DimPlot(MDS_NBM_GMP, reduction = "umap", label=F,pt.size = 0.25,split.by = "Group")
FeaturePlot(MDS_NBM_GMP,features=c("C1QTNF4","MKI67","CDK6","IL3RA"),min.cutoff = 0.25,split.by = "Group",sort.cell = T)
FeaturePlot(MDS_NBM_GMP,features=c("ANXA1"),min.cutoff = 0.25,split.by = "Group",sort.cell = T)

DotPlot(MDS_NBM_GMP,group.by = "Sample",features=c("IGFBP7"))

MDS_NBM_GMP_deseq<-run_pseudobulkDE_new(data=MDS_NBM_GMP,
                                        contrasts='condition,MDS,NBM',design='~condition',
                                        sample_col="Sample", condition_col="Group",cluster_col="HSPC_type",extr_col=NULL,do_prefiltering = TRUE)


MDS_NBM_GMP_deseq["Gene"]<-row.names(MDS_NBM_GMP_deseq)
MDS_NBM_GMP_deseq[MDS_NBM_GMP_deseq$Gene %in% c("NFKBIA","RELB","TXNIP","RIPK2","GADD45B","ATF3"),]

MDS_NBM_GMP_deseq["Gene"]<-row.names(MDS_NBM_GMP_deseq)
MDS_NBM_GMP_deseq["rank"]<- -sign(MDS_NBM_GMP_deseq$log2FoldChange)*log2(MDS_NBM_GMP_deseq$pvalue)

rank_GMP<-MDS_NBM_GMP_deseq$rank
names(rank_GMP)<-MDS_NBM_GMP_deseq$Gene
rank_GMP<-rank_GMP[is.finite(rank_GMP)]
MDS_GMP_Hall <- fgsea(hallmakers_set, rank_GMP, minSize=15, maxSize = 500, nperm=10000)
head(MDS_GMP_Hall[order(padj), ], n=10)
MDS_GMP_Hall_sig<-MDS_GMP_Hall%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_GMP_Hall_sig[["Enrichment"]]<-ifelse(MDS_GMP_Hall_sig$NES<0,"NBM","MDS")

MDS_GMP_C2 <- fgsea(C2_set, rank_GMP, minSize=15, maxSize = 500, nperm=10000)
head(MDS_GMP_C2[order(padj), ], n=10)
MDS_GMP_C2_sig<-MDS_GMP_C2%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_GMP_C2_sig[["Enrichment"]]<-ifelse(MDS_GMP_C2_sig$NES<0,"NBM","MDS")


MDS_GMP_C5 <- fgsea(C5_set, rank_Mono, minSize=15, maxSize = 500, nperm=10000)
head(MDS_GMP_C5[order(padj), ], n=10)
MDS_GMP_C5_sig<-MDS_Mono_C5%>%filter(padj<0.25)%>%arrange(desc(NES))
MDS_GMP_C5_sig[["Enrichment"]]<-ifelse(MDS_GMP_C5_sig$NES<0,"NBM","MDS")


Gene_list<-c("NFKBIA","RIPL2","RELB","GADD45A","TXNIP","ATF3")

##########
#HSPC WT 
#####################
NBM_CD34_HSPC_Myeloid_meta<-list(c(NBM1_CD34_Myeloid,NBM2_CD34_Myeloid,
                                   NBM3_CD34_Myeloid,NBM4_CD34_Myeloid))

Idents(MDS_NBM_HSPC_M)<-"type"

MDS_NBM_HSPC_M_barcode<-MDS_NBM_HSPC_M@meta.data[c("orig.ident","Mutant_state","type")]
MDS_NBM_HSPC_M_barcode$orig.ident<-ifelse(MDS_NBM_HSPC_M_barcode$orig.ident=="NMB4_CD34_Myeloid",
                                          "NBM4_CD34_Myeloid",MDS_NBM_HSPC_M_barcode$orig.ident)
MDS_NBM_HSPC_M_barcode["bar_code"]<-paste0(MDS_NBM_HSPC_M_barcode$orig.ident,"_",
                                           str_split(row.names(MDS_NBM_HSPC_M_barcode),pattern="_",simplify = T)[,3])


NBM1_HSPC_M<-NBM1_CD34_Myeloid
NBM2_HSPC_M<-NBM2_CD34_Myeloid
NBM3_HSPC_M<-NBM3_CD34_Myeloid
NBM4_HSPC_M<-NBM4_CD34_Myeloid

NBM1_HSPC_M[["bar_code"]]<-paste0(NBM1_HSPC_M$orig.ident,"_",
                                names(NBM1_HSPC_M$orig.ident))
NBM2_HSPC_M[["bar_code"]]<-paste0(NBM2_HSPC_M$orig.ident,"_",
                                  names(NBM2_HSPC_M$orig.ident))
NBM3_HSPC_M[["bar_code"]]<-paste0(NBM3_HSPC_M$orig.ident,"_",
                                  names(NBM3_HSPC_M$orig.ident))
NBM4_HSPC_M[["bar_code"]]<-paste0(NBM4_HSPC_M$orig.ident,"_",
                                  names(NBM4_HSPC_M$orig.ident))

NBM1_HSPC_M@meta.data<-NBM1_HSPC_M@meta.data%>%left_join(MDS_NBM_HSPC_M_barcode)
NBM2_HSPC_M@meta.data<-NBM2_HSPC_M@meta.data%>%left_join(MDS_NBM_HSPC_M_barcode)
NBM3_HSPC_M@meta.data<-NBM3_HSPC_M@meta.data%>%left_join(MDS_NBM_HSPC_M_barcode)
NBM4_HSPC_M@meta.data<-NBM4_HSPC_M@meta.data%>%left_join(MDS_NBM_HSPC_M_barcode)

NBM1_HSPC_M<-subset(NBM1_HSPC_M,cells=names(NBM1_HSPC_M$type[!is.na(NBM1_HSPC_M$type)]))
NBM2_HSPC_M<-subset(NBM2_HSPC_M,cells=names(NBM2_HSPC_M$type[!is.na(NBM2_HSPC_M$type)]))
NBM3_HSPC_M<-subset(NBM3_HSPC_M,cells=names(NBM3_HSPC_M$type[!is.na(NBM3_HSPC_M$type)]))
NBM4_HSPC_M<-subset(NBM4_HSPC_M,cells=names(NBM4_HSPC_M$type[!is.na(NBM4_HSPC_M$type)]))

NBM_HSPC_M_list<-list(NBM1_HSPC_M,NBM2_HSPC_M,NBM3_HSPC_M,NBM4_HSPC_M)


features_NBM_HSPC_M <- SelectIntegrationFeatures(object.list = NBM_HSPC_M_list) #Integratation Ctrl
NBM_anchors <- FindIntegrationAnchors(object.list = NBM_HSPC_M_list, anchor.features = features_NBM_HSPC_M, dims=1:30)
NBM_HSPC_M <- IntegrateData(anchorset = NBM_anchors,dims=1:30)



saveRDS(NBM_HSPC_M,"~NBM_HSPC_M.RDS")
readRDS("~NBM_HSPC_M.RDS")

NBM_HSPC_M <- RunPCA(NBM_HSPC_M, npcs = 30,verbose = FALSE);
ElbowPlot(NBM_HSPC_M)
nDims=12
NBM_HSPC_M <- FindNeighbors(NBM_HSPC_M, reduction = "pca", dims = 1:nDims)
NBM_HSPC_M <- FindClusters(NBM_HSPC_M, resolution = 0.3);
NBM_HSPC_M <- RunUMAP(NBM_HSPC_M,dims = 1:nDims);
Idents(NBM_HSPC_M)<-"type"
DimPlot(NBM_HSPC_M, reduction = "umap", label=F,pt.size = 0.25,split.by = "Group")

NBM_HSPC_M<-NBM_HSPC_M
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

NBM_HSPC_M <- CellCycleScoring(NBM_HSPC_M, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

NBM_HSPC_M$Proliferation_score <- NBM_HSPC_M$S.Score - MDS_NBM_HSPC_M$G2M.Score

DefaultAssay(NBM_HSPC_M)<-"integrated"


NBM_HSPC_M<- ScaleData(NBM_HSPC_M,vars.to.regress = c("S.Score", "G2M.Score"))


NBM_HSPC_M <- RunPCA(NBM_HSPC_M, npcs = 30,verbose = FALSE);

ElbowPlot(NBM_HSPC_M,ndims = 30)
nDims<-15

NBM_HSPC_M <- FindNeighbors(NBM_HSPC_M, reduction = "pca", dims = 1:nDims)
NBM_HSPC_M <- FindClusters(NBM_HSPC_M, resolution = 0.3);
NBM_HSPC_M <- RunUMAP(NBM_HSPC_M,dims = 1:nDims);


Idents(NBM_HSPC_M)<-"seurat_clusters"

DimPlot(NBM_HSPC_M, reduction = "umap", label=T,pt.size = 0.25)


FeaturePlot(NBM_HSPC_M, features=c("CD34","CD38"),order=T,min.cutoff = 0.25)

#########################
#Destiny diffusion for HSPC
###############
library(destiny)
MDS_NBM_HSPC_pca<-MDS_NBM_HSPC@reductions$pca@cell.embeddings[,1:10]
cell_type<-MDS_NBM_HSPC$HSPC_type
dm_MDS_NBM_HSPC <- DiffusionMap(MDS_NBM_HSPC_pca)
dm_MDS_NBM_HSPC_df<-data.frame(DC1 = eigenvectors(dm_MDS_NBM_HSPC)[, 1]*300,
                              DC2 = eigenvectors(dm_MDS_NBM_HSPC)[, 2]*300,
                              Samples = cell_type)

dm_MDS_NBM_HSPC_df%>%ggplot(aes(x = DC1, y = DC2, colour = Samples)) +
  geom_point(size=0.25)  + 
  xlab("Diffusion component 1") + 
  ylab("Diffusion component 2") +
  theme_classic()+scale_color_manual(values=c("grey","grey20","darkgreen","tomato2","tomato4","steelblue","steelblue3","steelblue4","grey80","tomato3"))
dm_MDS_NBM_HSPC<-dm_MDS_NBM_HSPC[-3]


MDS_NBM_HSPC[["diffusion"]] <- CreateDimReducObject(embeddings = as.matrix(dm_MDS_NBM_HSPC), key = "DC", assay = DefaultAssay(MDS_NBM_HSPC))

Idents(MDS_NBM_HSPC)<-"HSPC_type"
DimPlot(MDS_NBM_HSPC, reduction = "diffusion", split.by = "Group",ncol=2,label=T,pt.size = 0.25,repel = T,
        cols=c("grey","grey20","darkgreen","tomato2","tomato4","steelblue","steelblue3","steelblue4","grey80","tomato3"))

FeaturePlot(MDS_NBM_HSPC_M, reduction = "diffusion", split.by = "Group",features="C1QTNF4",min.cutoff = 0.25,ncol=2,label=T,pt.size = 0.25,order = T)&NoLegend()

#Monocle 3
############
#Monocle3 for HSPC
########
DimPlot(MDS_NBM_HSPC_M)

DefaultAssay(MDS_NBM_HSPC_M)<-"integrated"


Idents(MDS_NBM_HSPC_M)<-"HSPC_type"
DimPlot(MDS_NBM_HSPC_M, reduction = "diffusion", split.by = "Group",label=F,pt.size = 0.25)

MDS_NBM_M_lineage<-subset(MDS_NBM_HSPC_M,ident=c("HSC/MPPs","LMPPs","GMPs","Monoblasts","CD14+ Mono"))
MDS_NBM_E_lineage<-subset(MDS_NBM_HSPC_M,ident=c("HSC/MPPs","MEPs","Erythroid progenitors"))
MDS_NBM_L_lineage<-subset(MDS_NBM_HSPC_M,ident=c("HSC/MPPs","CLPs"))

library(SeuratWrappers)
library(monocle3)
MDS_NBM_HSPC_M.CDS <-as.cell_data_set(MDS_NBM_M_lineage)
MDS_NBM_HSPC_M.CDS@int_colData@listData$reducedDims@listData$UMAP <- MDS_NBM_HSPC_M.CDS@int_colData@listData$reducedDims@listData$UMAP[,c(1,2)]
MDS_NBM_HSPC_M.CDS <- cluster_cells(cds = MDS_NBM_HSPC_M.CDS, reduction_method = "UMAP",resolution=0.00001)
MDS_NBM_HSPC_M.CDS <- learn_graph(MDS_NBM_HSPC_M.CDS, use_partition = F)
MDS_NBM_HSPC_M.CDS <- order_cells(MDS_NBM_HSPC_M.CDS, reduction_method = "UMAP")

plot_cells(
  cds = MDS_NBM_HSPC_M.CDS,
  color_cells_by = "pseudotime",
  label_branch_points = F,
  label_leaves = F,
  show_trajectory_graph = TRUE,label_roots = T)



MDS_NBM_HSPC_M_pseud <- AddMetaData(
  object = MDS_NBM_HSPC_M,
  metadata = MDS_NBM_HSPC_M.CDS@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "Myelo_pseudo"
)


MDS_NBM_HSPC_E.CDS <-as.cell_data_set(MDS_NBM_E_lineage)
#MDS_NBM_HSPC_E.CDS@int_colData@listData$reducedDims@listData$UMAP <- MDS_NBM_HSPC_E.CDS@int_colData@listData$reducedDims@listData$DIFFUSION[,c(1,2)]
MDS_NBM_HSPC_E.CDS <- cluster_cells(cds = MDS_NBM_HSPC_E.CDS, reduction_method = "UMAP",resolution=0.00001)
MDS_NBM_HSPC_E.CDS <- learn_graph(MDS_NBM_HSPC_E.CDS, use_partition = F)
MDS_NBM_HSPC_E.CDS <- order_cells(MDS_NBM_HSPC_E.CDS, reduction_method = "UMAP")

plot_cells(
  cds = MDS_NBM_HSPC_E.CDS,
  color_cells_by = "pseudotime",
  label_branch_points = F,
  label_leaves = F,
  show_trajectory_graph = TRUE,label_roots = T)



MDS_NBM_HSPC_M_pseud <- AddMetaData(
  object = MDS_NBM_HSPC_M_pseud,
  metadata = MDS_NBM_HSPC_E.CDS@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "Ery_pseudo"
)


MDS_NBM_HSPC_L.CDS <-as.cell_data_set(MDS_NBM_L_lineage)
#MDS_NBM_HSPC_E.CDS@int_colData@listData$reducedDims@listData$UMAP <- MDS_NBM_HSPC_E.CDS@int_colData@listData$reducedDims@listData$DIFFUSION[,c(1,2)]
MDS_NBM_HSPC_L.CDS <- cluster_cells(cds = MDS_NBM_HSPC_L.CDS, reduction_method = "UMAP",resolution=0.00001)
MDS_NBM_HSPC_L.CDS <- learn_graph(MDS_NBM_HSPC_L.CDS, use_partition = F)
MDS_NBM_HSPC_L.CDS <- order_cells(MDS_NBM_HSPC_L.CDS, reduction_method = "UMAP")

plot_cells(
  cds = MDS_NBM_HSPC_L.CDS,
  color_cells_by = "pseudotime",
  label_branch_points = F,
  label_leaves = F,
  show_trajectory_graph = TRUE,label_roots = T)



MDS_NBM_HSPC_M_pseud <- AddMetaData(
  object = MDS_NBM_HSPC_M_pseud,
  metadata = MDS_NBM_HSPC_L.CDS@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "Lympo_pseudo"
)

MDS_NBM_HSPC_M_pseud$Group<-factor(MDS_NBM_HSPC_M_pseud$Group,levels=c("NBM","MDS"))


FeaturePlot(MDS_NBM_HSPC_M_pseud, reduction = "umap", label=F ,pt.size = 0.1,features="Myelo_pseudo",split.by = "Group")& 
  scale_color_viridis_c()&theme(axis.title=element_blank())
FeaturePlot(MDS_NBM_HSPC_M_pseud, reduction = "umap", label=F ,pt.size = 0.1,features="Ery_pseudo",split.by = "Group")& 
  scale_color_viridis_c()&theme(axis.title=element_blank())
FeaturePlot(MDS_NBM_HSPC_M_pseud, reduction = "umap", label=F ,pt.size = 0.1,features="Lympo_pseudo",split.by = "Group")& 
  scale_color_viridis_c()&theme(axis.title=element_blank())




DimPlot(MDS_NBM_HSPC_M_pseud, reduction = "umap",split.by = "Sample",label=T,repel=T ,pt.size = 0.5)&NoLegend()

DefaultAssay(MDS_NBM_HSPC_M_pseud)<-"RNA"

MDS_NBM_HSPC_M_pseud<-AddModuleScore(MDS_NBM_HSPC_M_pseud,features=hallmakers_set[1],name="NFkB_via_TNF")
MDS_NBM_HSPC_M_pseud<-AddModuleScore(MDS_NBM_HSPC_M_pseud,features=hallmakers_set[10],name="Apoptosis")

FeatureScatter(MDS_NBM_HSPC_M_pseud,feature1="Myelo_pseudo",feature2="",group.by ="Apoptosis1",smooth=T)

MDS_NBM_HSPC_M_pseud@meta.data%>%ggplot(aes(Myelo_pseudo,Apoptosis1,color=Sample))+geom_smooth()+scale_color_manual(values=c(rep("tomato4",4),rep("steelblue4",4)))
MDS_NBM_HSPC_M_pseud@meta.data%>%ggplot(aes(Ery_pseudo,Apoptosis1,color=Group))+geom_smooth()+scale_color_manual(values=c(rep("steelblue4",1),rep("tomato4",1)))
MDS_NBM_HSPC_M_pseud@meta.data%>%ggplot(aes(Lympo_pseudo,Apoptosis1,color=Group))+geom_smooth(se=F)+theme_classic()+scale_color_manual(values=c(rep("steelblue4",1),rep("tomato4",1)))
MDS_NBM_HSPC_M_pseud@meta.data%>%ggplot(aes(Myelo_pseudo,Proliferation_score,color=Group))+geom_smooth(se=F)+theme_classic(+scale_color_manual(values=c(rep("steelblue4",1),rep("tomato4",1)))

RidgePlot(MDS_NBM_HSPC_M_pseud, features = "Myelo_pseudo",group.by="HSPC_type")
RidgePlot(MDS_NBM_HSPC_M_pseud, features = "Myelo_pseudo",group.by="Sample",cols=c(rep("tomato3",5),rep("steelblue3",4)))
RidgePlot(MDS_NBM_HSPC_M_pseud, features = "Myelo_pseudo",group.by="Phase")

RidgePlot(MDS_NBM_H_M_pseud, features = "pseudotime",group.by="Sample")&NoLegend()
RidgePlot(subset(MDS_NBM_H_M_pseud,cells=names(MDS_NBM_H_M_pseud$Mutant_state[MDS_NBM_H_M_pseud$Mutant_state %in% c("NBM","Mutant")])), 
          features = "pseudotime",group.by="Sample",cols=c(rep("tomato3",5),rep("steelblue3",4)))

MDS_NBM_H_M_pseud<-subset(MDS_NBM_H_M_pseud,
                          cells=names(MDS_NBM_H_M_pseud$Sample[MDS_NBM_H_M_pseud$Sample!="MDS_SF3B1"]))
saveRDS(MDS_NBM_HSPC_M_pseud,"~/single_cell_sequencing/MDS single cell/MDS_NBM_HSPC_M_pseud.RDS")
MDS_NBM_H_M_pseud<-readRDS("~/single_cell_sequencing/MDS single cell/MDS_NBM_H_M_pseud.RDS")

##Gene expression following pseudotime in heatmap

Gene_list<-c(NBM_H_M_pseud_sample_cor_df_neg$name,NBM_H_M_pseud_sample_cor_df_pos$name)
NBM_HSC_Mono_markers<-data.frame(read_xlsx("~/single_cell_sequencing/Markers/NBM_HSC_Mono_markers.xlsx")) #top100 genes in HSPCs in NBM
NBM_HSC_Mono_markers<-NBM_HSC_Mono_markers[!duplicated(NBM_HSC_Mono_markers$Gene),]
row.names(NBM_HSC_Mono_markers)<-NBM_HSC_Mono_markers$Gene
Gene_list<-unique(NBM_HSC_Mono_markers$Gene)
Gene_list<-MDS_HSC_Hall_sig[30]$leadingEdge[[1]]

MDS_NBM_M_sample<-subset(MDS_NBM_H_M_pseud, cells=
                           c(sample(names(MDS_NBM_H_M_pseud$Group[MDS_NBM_H_M_pseud$Group=="MDS"]),5000,replace = F),
                             sample(names(MDS_NBM_H_M_pseud$Group[MDS_NBM_H_M_pseud$Group=="NBM"]),5000,replace = F)))
MDS_NBM_M_matrix<-MDS_NBM_M_sample@assays$RNA
MDS_NBM_M_matrix<-MDS_NBM_M_matrix[Gene_list,]
MDS_NBM_M_matrix<-t(scale(t(MDS_NBM_M_matrix)))
MDS_NBM_M_matrix<-apply(MDS_NBM_M_matrix,c(1,2),function(x){
  a<-ifelse(x>2,2,x)
  a<-ifelse(a< -2, -2,a)
  return(a)
})

annotation_pseud_order<-MDS_NBM_M_sample@meta.data%>%arrange(pseudotime)
annotation_pseud_order_NBM<-annotation_pseud_order[annotation_pseud_order$Group=="NBM",]
annotation_pseud_order_MDS<-annotation_pseud_order[annotation_pseud_order$Group=="MDS",]


annotaton_col<-list(HSPC_type=c("HSC/MPPs"="grey80","LMPPs"="steelblue3","GMP/Promonocytes"="darkgreen","CD14+ Mono"="tomato4"),
                    Markers=c("HSC/MPPs"="grey80","LMPPs"="steelblue3","GMP/Promonocytes"="darkgreen","CD14+ Mono"="tomato4"))
NBM_heatmap<-pheatmap(MDS_NBM_M_matrix[,row.names(annotation_pseud_order_MDS)],annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_MDS[c("HSPC_type","pseudotime")],show_colnames=F,show_rownames=F,cluster_cols = F,cluster_rows = T,clustering_method = "ward.D2")
row_NBM <- row_order(NBM_heatmap)

pheatmap(MDS_NBM_M_matrix[,row.names(annotation_pseud_order_NBM)],annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_NBM[c("HSPC_type")],show_colnames=F,show_rownames=F,cluster_cols = F,cluster_rows = F, row_order=row_NBM)
#heatmap NBM
pheatmap(MDS_NBM_M_matrix[,row.names(annotation_pseud_order_MDS)],annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_MDS[c("HSPC_type")],show_colnames=F,show_rownames=F,cluster_cols = F,cluster_rows = F, row_order=row_NBM)
#heatmap MDS

annotation_pseud_order<-NBM_H_M_pseud_sample_metadata%>%arrange(pseudotime)
annotaton_col<-list(HSPC_type=c("HSC/MPPs"="grey80","LMPPs"="steelblue","GMP/Promonocytes"="darkgreen","CD14+ Mono"="tomato4"),
                    gene_cluster=c("HSC/MPP"="grey80", "LMPP/GMP"="darkgreen","Monocyte"="tomato4"))

NBM_heatmap<-pheatmap(NBM_H_M_pseud_sample_matrix_cor_scaled[,row.names(annotation_pseud_order)],annotation_colors = annotaton_col,annotation_col=annotation_pseud_order[c("HSPC_type","pseudotime")],show_colnames=F,show_rownames=T,cluster_cols = F,cluster_rows = T,clustering_method = "ward.D2")
row_NBM <- row_order(NBM_heatmap)


row_to_label<-which(row.names(NBM_heatmap@matrix) %in% c("HLF", "MEIS1", "ELANE", "MPO", "CD14","ITGAM"))
labrow<-row.names(NBM_heatmap@matrix)
labrow[-row_to_label]<-""


pheatmap(NBM_H_M_pseud_sample_matrix_cor_scaled[,row.names(annotation_pseud_order)],use_raster=T,#annotation_row=NBM_HSC_Mono_markers["gene_cluster"],
         annotation_colors = annotaton_col,annotation_col=annotation_pseud_order[c("HSPC_type","pseudotime")],show_colnames=F,
         labels_row=labrow,cluster_cols = F,cluster_rows = F,row_order=row_NBM,clustering_method = "ward.D2",fontsize_row=5)



MDS_H_M_pseud<-subset(MDS_NBM_H_M_pseud, cells=names(MDS_NBM_H_M_pseud$Group[MDS_NBM_H_M_pseud$Group=="MDS"]))
MDS_H_M_pseud_sample<-subset(MDS_H_M_pseud,cells=sample(names(MDS_H_M_pseud$orig.ident),5000,replace=F))

MDS_H_M_pseud_sample_matrix<-MDS_H_M_pseud_sample@assays$RNA@counts
MDS_H_M_pseud_sample_matrix_cor_scaled<-MDS_H_M_pseud_sample_matrix[Gene_list,]
MDS_H_M_pseud_sample_matrix_cor_scaled<-t(scale(t(MDS_H_M_pseud_sample_matrix_cor_scaled)))
MDS_H_M_pseud_sample_matrix_cor_scaled<-apply(MDS_H_M_pseud_sample_matrix_cor_scaled,c(1,2),function(x){
  a<-ifelse(x>3,3,x)
  a<-ifelse(a< -3, -3,a)
  return(a)
})

MDS_H_M_pseud_sample_metadata<-MDS_H_M_pseud_sample@meta.data


MDS_annotation_pseud_order<-MDS_H_M_pseud_sample_metadata%>%arrange(pseudotime)
pheatmap(MDS_H_M_pseud_sample_matrix_cor_scaled[,row.names(MDS_annotation_pseud_order)],
         annotation_row=NBM_HSC_Mono_markers["gene_cluster"],
         annotation_colors = annotaton_col,annotation_col=MDS_annotation_pseud_order[c("HSPC_type","pseudotime")],
         show_colnames=F,cluster_cols = F,cluster_rows = F,row_order=ro_NBM,
         labels_row=labrow,fontsize_row=5,
         clustering_method = "ward.D2")

#  score 

DefaultAssay(MDS_NBM_HSPC_M_pseud)<-"RNA"

MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_pseud,features=hallmakers_set[1],name="NFkB_via_TNF")
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score,features=hallmakers_set[10],name="Apoptosis")
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score,features=hallmakers_set[19],name="Interferon_gamma_response")
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score, features=list(NBM_HSC_Mono_markers$Gene[NBM_HSC_Mono_markers$gene_cluster=="HSC/MPP"]),name="HSC_score", ctrl=10)
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score, features=list(NBM_HSC_Mono_markers$Gene[NBM_HSC_Mono_markers$gene_cluster=="LMPP/GMP"]),name="LMPP_score",ctrl=10)
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score, features=list(NBM_HSC_Mono_markers$Gene[NBM_HSC_Mono_markers$gene_cluster=="Monocyte"]),name="Mono_score",ctrl=10)
MDS_NBM_HSPC_M_score<-AddModuleScore(MDS_NBM_HSPC_M_score, features=list(NBM_HSC_Mono_markers$Gene[NBM_HSC_Mono_markers$gene_cluster=="Monocyte"]),name="Mono_score",ctrl=10)


VlnPlot(MDS_NBM_HSPC_M_score,features=c("HSC_score1","Mono_score1"), split.by="Group",pt.size=0,ncol=1)&theme(legend.position = "right")


MDS_NBM_H_M_sample<-subset(MDS_NBM_HSPC_M_score,cells=sample(names(MDS_NBM_HSPC_M_score$Group),5000,replace=T))


MDS_NBM_H_M_sample@meta.data%>%ggplot(aes(MDS_NBM_HSPC_M_score,Mono_score1,color=Sample))+geom_smooth(se=F)+theme_classic()+scale_color_manual(values=c(rep("tomato3",5), c(rep("steelblue4",4))))


#Gene expression following pseudotime in smooth curve

Gene_test<-c("GADD45A")
group_by="Sample"
Gene_seudo<-function(data=MDS_NBM_H_M_pseud,group_by="Group" ,Gene_test=Gene_test){
  meta <-MDS_NBM_H_M_pseud@meta.data
  table<-data.frame(MDS_NBM_H_M_pseud@assays$RNA@counts)[Gene_test,]
  table<-data.frame(t(table))
  meta <-cbind(meta,table)
  meta <-melt(meta, measure.vars=Gene_test,variable.name="Gene",value.name = "Expression")
  meta%>%ggplot(aes(pseudotime,Expression,color = Sample))+geom_smooth(se=FALSE)+
    theme_classic()+facet_wrap(.~Gene, ncol=1, scales="free_y")+
    theme(strip.background = element_blank(),
          strip.text = element_text(size=15))+
    scale_color_manual(values=c(rep("steelblue3",4), rep("tomato3",4)))
}

Gene_seudo(data=MDS_NBM_H_M_pseud,Gene_test="GADD45A")


saveRDS(MDS_NBM_H_M_pseud,"~/single_cell_sequencing/MDS single cell/MDS_NBM_H_M_pseud.RDS")


############
#MEP
########################


library(destiny)
Idents(MDS_NBM_HSPC_M)<-"HSPC_type"
MDS_NBM_E<-subset(MDS_NBM_HSPC_M,ident=c("HSC/MPPs","MEPs","Erythroid progenitors"))

DefaultAssay(MDS_NBM_E)<-"integrated"
MDS_NBM_E<- ScaleData(MDS_NBM_E,vars.to.regress = c("S.Score", "G2M.Score"))
MDS_NBM_E <- RunPCA(MDS_NBM_E, npcs = 30,verbose = FALSE);

ElbowPlot(MDS_NBM_E,ndims = 30)
nDims<-20


MDS_NBM_E <- FindNeighbors(MDS_NBM_E, reduction = "pca", dims = 1:nDims)
MDS_NBM_E <- FindClusters(MDS_NBM_E, resolution = 0.4);
MDS_NBM_E <- RunUMAP(MDS_NBM_E,dims = 1:nDims);
Idents(MDS_NBM_E)<-"seurat_clusters"
Idents(MDS_NBM_E)<-"HSPC_type"

DimPlot(MDS_NBM_E, reduction = "umap", split.by = "Sample",label=T,pt.size = 0.25)
#MDS_NBM_E<-subset(MDS_NBM_E,ident=c("8","2"),invert=T)

FeaturePlot(MDS_NBM_E, features=c("PTPRC","ITGA2B","ITGA4","SLC4A1","EPB42"),order=T, min.cutoff = 0.25)

MDS_NBM_E$percentHB <- Seurat::PercentageFeatureSet(MDS_NBM_E, pattern ="^HB",assay="RNA");
FeaturePlot(MDS_NBM_E, features=c("percentHB"),order=T)&lims()
plot(MDS_NBM_E$percentHB)

MDS_NBM_E_pca<-MDS_NBM_E@reductions$pca@cell.embeddings[,1:15]
cell_type<-MDS_NBM_E$seurat_clusters
dm_MDS_NBM_E <- DiffusionMap(MDS_NBM_E_pca)
dm_MDS_NBM_E_DM<-data.frame(DC1 = eigenvectors(dm_MDS_NBM_E)[, 1]*300,
                              DC2 = eigenvectors(dm_MDS_NBM_E)[, 2]*300,
                              DC3 = eigenvectors(dm_MDS_NBM_E)[, 3]*300,
                              DC4 = eigenvectors(dm_MDS_NBM_E)[, 4]*300,
                              Samples = cell_type)

dm_MDS_NBM_E_DM%>%ggplot(aes(x = DC1, y = DC2, colour = Samples)) +
  geom_point()  + 
  xlab("Diffusion component 1") + 
  ylab("Diffusion component 2") +
  theme_classic()
dm_MDS_NBM_E_DM<-dm_MDS_NBM_E_DM[-5]


MDS_NBM_E[["diffusion"]] <- CreateDimReducObject(embeddings = as.matrix(dm_MDS_NBM_E_DM), key = "DC", assay = DefaultAssay(MDS_NBM_E))

Idents(MDS_NBM_E)<-"HSPC_type"
DimPlot(MDS_NBM_E, reduction = "diffusion", split.by = "Group",label=T,pt.size = 0.25)
FeaturePlot(MDS_NBM_E,reduction = "umap",features=c("ITGAM"),split.by ="Group",min.cutoff = 0.25,pt.size=0.25,order=T)
DotPlot(MDS_NBM_E,split.by = "Sample",cols=rep("tomato4",10),features=c("TFRC"))
########################
#Monocle 3

MDS_NBM_E.CDS <-as.cell_data_set(MDS_NBM_E)
MDS_NBM_E.CDS@int_colData@listData$reducedDims@listData$UMAP <- MDS_NBM_E.CDS@int_colData@listData$reducedDims$DIFFUSION[,c(1,2)]
MDS_NBM_E.CDS <- cluster_cells(cds = MDS_NBM_E.CDS, reduction_method = "UMAP",resolution=0.0005)
MDS_NBM_E.CDS <- learn_graph(MDS_NBM_E.CDS, use_partition = F)
MDS_NBM_E.CDS <- order_cells(MDS_NBM_E.CDS, reduction_method = "UMAP")

plot_cells(
  cds = MDS_NBM_E.CDS,
  color_cells_by = "pseudotime",
  label_branch_points = F,
  label_leaves = F,
  show_trajectory_graph = TRUE,label_roots = T
)

Myeloid_lineage <- choose_graph_segments(MDS_NBM_HSPC_M.CDS)

MDS_NBM_E <- AddMetaData(
  object = MDS_NBM_E,
  metadata = MDS_NBM_E.CDS@principal_graph_aux@listData$UMAP$pseudotime,
  col.name = "pseudotime"
)

FeaturePlot(MDS_NBM_E, reduction = "diffusion", label=F ,pt.size = 0.5,features="pseudotime",split.by = "Group")& 
  scale_color_viridis_c()
MDS_NBM_H_M_pseud$Group<-factor(MDS_NBM_H_M_pseud$Group,levels=c("NBM","MDS"))

DimPlot(MDS_NBM_E, reduction = "diffusion",label=T,repel=T ,pt.size = 0.1)&NoLegend()

DefaultAssay(MDS_NBM_H_M_pseud)<-"RNA"
FeatureScatter(MDS_NBM_H_M_pseud,feature1="pseudotime",feature2="HLF",group.by="Group")
RidgePlot(MDS_NBM_E, features = "pseudotime",group.by="HSPC_type")
RidgePlot(MDS_NBM_E, features = "pseudotime",group.by="Sample",cols=c(rep("tomato3",5),rep("steelblue3",4)))
RidgePlot(MDS_NBM_E, features = "pseudotime",group.by="Phase")

RidgePlot(MDS_NBM_H_M_pseud, features = "pseudotime",group.by="Sample")&NoLegend()
RidgePlot(subset(MDS_NBM_H_M_pseud,cells=names(MDS_NBM_H_M_pseud$Mutant_state[MDS_NBM_H_M_pseud$Mutant_state %in% c("NBM","Mutant")])), features = "pseudotime",group.by="Sample")&NoLegend()



#E_lineage_marker gene identify

run_pseudobulkDE <- function(meta_data,
                             expr_mat,
                             contrasts,
                             design,
                             sample_col,
                             condition_col,
                             cluster_col,
                             extr_col = NULL,
                             do_prefiltering = TRUE,
                             use_celltype = TRUE) {
  
  
  # Pull the necessary columns from the metadata table
  anno_raw <- meta_data[, c(sample_col, condition_col, cluster_col, extr_col)]
  
  # Pull names from the additional column(s)
  extr_name <- colnames(anno_raw)[4:ncol(anno_raw)]
  # Replace forbidden chrs and add a chr in columns if values start with int
  anno_raw <- apply(anno_raw, 2, function(x) {
    x <- ifelse(grepl('^[0-9]', x), paste0('x', x), x)
    gsub('-|_', '.', x)
  })
  
  
  # Create cell-level annoation dataframe with columns as factors
  anno_cell <- data.frame(
    samples          = anno_raw[, sample_col],
    clusters         = anno_raw[, cluster_col],
    condition        = anno_raw[, condition_col],
    Extra            = anno_raw[, extr_col],
    row.names        = rownames(meta_data),
    stringsAsFactors = TRUE
  )
  # Add column name(s) to the additional column(s)
  
  # Add factor column that is the conjunction of samples and conditions
  anno_cell$sc <- factor(paste0(anno_cell$samples, '.', anno_cell$condition))
  # split the contrast into a strings
  contrast <- strsplit(contrasts, ',')[[1]]
  
  # Turn the design parameter into a formula object
  design <- formula(design)
  
  # Create groups that will be used to aggregate the expr. in the sparse matrix
  if (use_celltype) {
    agg_g <- S4Vectors::DataFrame(anno_cell[, c('clusters', 'sc')])
  } else {
    agg_g <- S4Vectors::DataFrame(anno_cell[, 'sc'])
  }
  
  # Aggregate the RNA counts by summation; making a sample-level count-matrix
  texpr_mat <- Matrix::t(expr_mat)
  
  agg_m <- Matrix.utils::aggregate.Matrix(texpr_mat, groupings = agg_g,
                                          fun = 'sum')
  
  if (use_celltype) {
    # Create vector that contains the cluster names * No. samples in that cluster
    split_f <- sapply(strsplit(rownames(agg_m), '_'), `[`, 1)
    # Turn into a list and split the list into components for each cluster and
    # transform, so rows are genes and columns are samples and make rownames
    # as the sample IDs
    
    # Split the aggregated matrix, creating dataframes for each cluster
    split_m <- split.data.frame(agg_m, factor(split_f))
    
    # Create renaming function
    get_colnames <- function(x) {
      magrittr::set_colnames(Matrix::t(x),
                             stringr::str_extract(rownames(x),
                                                  '(?<=_)[:graph:]+'))
    }
    # Apply renaming function
    agg_m <- lapply(split_m, get_colnames)
    
    # Create the intial dataframe used by DESeq2 for determining the conditions
    anno_con <- unique(anno_cell[, -1])
    
  } else {
    # Simply transpose agg.m
    agg_m <- Matrix::t(agg_m)
    
    # Create the intial dataframe used by DESeq2 for determining the conditions
    anno_con <- unique(anno_cell[,-c(1:2)])
    
  }
  
  # Subset the condition annotation based on the condition being present in the contrast of interest
  anno_con <- anno_con[anno_con$condition %in% contrast, ]
  
  if (use_celltype) {
    # Create the rownames of the condition annotation by pasting samples, condition and clusters
    rownames(anno_con) <- paste0(anno_con$sc, '.', anno_con$clusters)
    # Find the unique clusters present in the metadata
    uniq_clusters <- unique(anno_con$clusters)
    # Add the corresponding cluster name to the column names in agg.matrix
    for(i in uniq_clusters) {
      colnames(agg_m[[i]]) <- paste0(colnames(agg_m[[i]]), '.', i)
    }
    
    # Turn the per-cluster aggregated gene counts into a single matrix
    counts_obj <- do.call(cbind, agg_m)
    
  } else {
    # Create the rownames of the condition annotation by pasting samples and condition
    rownames(anno_con) <- anno_con$sc
    
    # change the name of the agg.m object to adhere to the rest of the script
    counts_obj <- agg_m
  }
  
  # Subset the full gene-count matrix based on the samples in the condition annotation
  count_name <- colnames(counts_obj)
  anno_name <- rownames(anno_con)
  counts_obj <- counts_obj[, which(count_name %in% anno_name)]
  
  # Reorder the condition annotation based on the column names in counts_obj
  anno_con <- anno_con[match(colnames(counts_obj), anno_name), ]
  
  # Create a DESeq2 object based on the aggregated counts, anno_con and design
  dds <- DESeq2::DESeqDataSetFromMatrix(counts_obj, colData = anno_con,
                                        design = design)
  
  # Remove outliers, i.e. genes only expressed in a few samples
  if (do_prefiltering) {
    
    message('Info: perform pre-filtering')
    
    con_id <- anno_con[, contrast[1]] == contrast[2]
    
    fragments <- DESeq2::fpm(dds, robust = TRUE)
    
    keep_g1 <- rowSums(fragments[, con_id] >= 2) >= round(sum(con_id) / 2)
    keep_g2 <- rowSums(fragments[, !con_id] >= 2) >= round(sum(!con_id) / 2)
    
    keep_tot <- keep_g1 | keep_g2
    
    dds <- DESeq2::DESeq(dds[keep_tot, ], parallel = FALSE)
    
  } else {
    dds <- DESeq2::DESeq(dds, parallel = FALSE, minReplicatesForReplace = Inf)
  }
  
  # Compute normalized counts
  norm_counts <- as.matrix(DESeq2::counts(dds, normalized = TRUE))
  
  # Generate result table and compute shrunken log fold change
  res <- DESeq2::results(dds, contrast = contrast)
  res_lfc <- DESeq2::lfcShrink(dds, contrast = contrast, type = 'normal',
                               quiet = TRUE)
  res_lfc_ashr <- DESeq2::lfcShrink(dds, contrast = contrast, type = 'ashr',
                                    quiet = TRUE)
  
  # Perpare all the output in a list
  output_l <- list(
    dds          = dds,
    norm_counts  = norm_counts,
    res          = res,
    res_lfc      = res_lfc,
    res_lfc_ashr = res_lfc_ashr
  )
  # Return the DE analyis output list
  Output<-data.frame(cbind(norm_counts,res_lfc_ashr))
  return(Output)
}


NBM_E<-subset(MDS_NBM_E, cells=names(MDS_NBM_E$Group[MDS_NBM_E$Group=="NBM"]))
NBM_E[["HSC_group"]]<-ifelse(NBM_E$HSPC_type=="HSC/MPPs","HSC/MPPs","Others")
NBM_E[["MEP_group"]]<-ifelse(NBM_E$HSPC_type=="MEPs","MEPs","Others")
NBM_E[["Eprog_group"]]<-ifelse(NBM_E$HSPC_type=="Erythroid progenitors","Eprog","Others")



HSC_markers_deseq<-run_pseudobulkDE(meta_data=NBM_E@meta.data,expr_mat=NBM_E@assays$RNA@counts,contrasts="condition,HSC/MPPs,Others",
                                      design= "~condition+Extra", sample_col="Sample",condition_col="HSC_group",cluster_col="HSPC_type", extr_col ="Sample")
HSC_markers_deseq["Gene"]<-row.names(HSC_markers_deseq)
HSC_markers_deseq["Rank"]<- -sign(HSC_markers_deseq$log2FoldChange)*log10(HSC_markers_deseq$padj)
HSC_markers<-HSC_markers_deseq%>%top_n(n=50,wt=Rank)
HSC_markers<-HSC_markers[c("Gene","log2FoldChange","padj")]
HSC_markers["Markers"]<-"HSC/MPPs"

MEP_markers_deseq<-run_pseudobulkDE(meta_data=NBM_E@meta.data,expr_mat=NBM_E@assays$RNA@counts,contrasts="condition,MEPs,Others",
                                    design= "~condition+Extra", sample_col="Sample",condition_col="MEP_group",cluster_col="HSPC_type", extr_col ="Sample")
MEP_markers_deseq["Gene"]<-row.names(MEP_markers_deseq)
MEP_markers_deseq["Rank"]<- -sign(MEP_markers_deseq$log2FoldChange)*log10(MEP_markers_deseq$padj)
MEP_markers<-MEP_markers_deseq%>%top_n(n=50,wt=Rank)
MEP_markers<-MEP_markers[c("Gene","log2FoldChange","padj")]
MEP_markers["Markers"]<-"MEPs"

Eprog_markers_deseq<-run_pseudobulkDE(meta_data=NBM_E@meta.data,expr_mat=NBM_E@assays$RNA@counts,contrasts="condition,Eprog,Others",
                                    design= "~condition+Extra", sample_col="Sample",condition_col="Eprog_group",cluster_col="HSPC_type", extr_col ="Sample")
Eprog_markers_deseq["Gene"]<-row.names(Eprog_markers_deseq)
Eprog_markers_deseq["Rank"]<- -sign(Eprog_markers_deseq$log2FoldChange)*log10(Eprog_markers_deseq$padj)
Eprog_markers<-Eprog_markers_deseq%>%top_n(n=50,wt=Rank)
Eprog_markers<-Eprog_markers[c("Gene","log2FoldChange","padj")]
Eprog_markers["Markers"]<-"Eprogs"

E_lineage_marker<-rbind(HSC_markers,MEP_markers,Eprog_markers)

###############################
#Heatmap on lineage
##############################
MDS_NBM_E_sample<-subset(MDS_NBM_E, cells=
                           c(names(MDS_NBM_E$Group[MDS_NBM_E$Group=="MDS"]),
                             sample(names(MDS_NBM_E$Group[MDS_NBM_E$Group=="NBM"]),5000,replace = F)))
MDS_NBM_E_matrix<-MDS_NBM_E_sample@assays$RNA
MDS_NBM_E_matrix<-MDS_NBM_E_matrix[E_lineage_marker$Gene,]
MDS_NBM_E_matrix<-t(scale(t(MDS_NBM_E_matrix)))
MDS_NBM_E_matrix<-apply(MDS_NBM_E_matrix,c(1,2),function(x){
  a<-ifelse(x>3,3,x)
  a<-ifelse(a< -3, -3,a)
  return(a)
})

annotation_pseud_order<-MDS_NBM_E_sample@meta.data%>%arrange(pseudotime)
annotation_pseud_order_NBM<-annotation_pseud_order[annotation_pseud_order$Group=="NBM",]
annotation_pseud_order_MDS<-annotation_pseud_order[annotation_pseud_order$Group=="MDS",]


annotaton_col<-list(HSPC_type=c("HSC/MPPs"="grey80","MEPs"="tomato4","Late erythroid progenitors"="red3"),
                    Markers=c("HSC/MPPs"="grey80", "MEPs"="tomato4","Eprogs"="red3"))
pheatmap(MDS_NBM_E_matrix[,row.names(annotation_pseud_order_NBM)],annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_NBM[c("HSPC_type","pseudotime")],show_colnames=F,show_rownames=F,cluster_cols = F,cluster_rows = T,clustering_method = "ward.D")
NBM_heatmap<-pheatmap(MDS_NBM_E_matrix[,row.names(annotation_pseud_order_NBM)],annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_NBM[c("HSPC_type","pseudotime")],show_colnames=F,show_rownames=F,cluster_cols = F,cluster_rows = T,clustering_method = "ward.D")
row_NBM <- row_order(NBM_heatmap)


pheatmap(MDS_NBM_E_matrix[,row.names(annotation_pseud_order_NBM)],annotation_row=E_lineage_marker["Markers"],
         annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_NBM[c("HSPC_type","pseudotime")],show_colnames=F, show_rownames = F,
         labels_row=labrow,cluster_cols = F,useRaster=T,cluster_rows = F,row_order=row_NBM)
          #heatmap NBM


pheatmap(MDS_NBM_E_matrix[,row.names(annotation_pseud_order_MDS)],use_raster=T,annotation_row=E_lineage_marker["Markers"],
         annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_MDS[c("HSPC_type","pseudotime")],show_colnames=F, show_rownames = F,
         labels_row=labrow,cluster_cols = F,cluster_rows = F,row_order=row_NBM,
         clustering_method = "ward.D2")
          #heatmap MDS

########
# HSPC MT vs Residual
###########
MDS_HSPC<-subset(MDS_NBM_HSPC,subset=Group=="MDS")
Idents(MDS_HSPC)<-"Mutation"

DimPlot(MDS_HSPC,split.by = "Sample")

summary(factor(MDS_HSPC$Mutation))


MDS_HSPC_pseudo<- AggregateExpression(MDS_HSPC, assays = "RNA", return.seurat = T, group.by = c("Mutation","Sample"))
Idents(MDS_HSPC_pseudo)<-"Mutation"


MDS_HSPC_markers<-FindMarkers(MDS_HSPC_pseudo,assay="RNA",ident.1 = "Mutant", ident.2="Unknown")

MDS_HSPC_markers["Gene"]<-row.names(MDS_HSPC_markers)

MDS_HSPC<-subset(MDS_HSPC,subset=Sample=="MDS_337",invert=T)

MDS_HSPC[["Mutation"]]<-"Unknown"
MDS_HSPC[["Mutation"]]<-ifelse(MDS_HSPC$SF3B1=="Mutant","Mutant",MDS_HSPC$Mutation)
MDS_HSPC[["Mutation"]]<-ifelse(MDS_HSPC$TET2=="Mutant","Mutant",MDS_HSPC$Mutation)
#MDS_NBM_all_mt[["Mutation"]]<-ifelse(MDS_NBM_all_mt$DNMT31=="Mutant","Mutant",MDS_NBM_all_mt$Mutation)


MDS_HSPC_deseq<-run_pseudobulkDE(meta_data=MDS_HSPC@meta.data,expr_mat=MDS_HSPC@assays$RNA@counts,
                                contrasts='condition,Mutant,Unknown',design='~condition+Extra',
                                sample_col="Sample", condition_col="Mutation",cluster_col="HSPC_type",extr_col='Sample',do_prefiltering = TRUE,use_celltype = F)



MDS_HSPC_deseq["Gene"]<-row.names(MDS_HSPC_deseq)
MDS_HSPC_deseq["rank"]<- -sign(MDS_HSPC_deseq$log2FoldChange)*log2(MDS_HSPC_deseq$pvalue)
MDS_HSPC_deseq<-MDS_HSPC_deseq[order(MDS_HSPC_deseq$padj),]
#write_xlsx(MDS_HSPC_deseq,"single_cell_sequencing/MDS single cell/HSPC_MT_vs_NA.xlsx")

rank_MDS_HSPC<-MDS_HSPC_deseq$rank
names(rank_MDS_HSPC)<-MDS_HSPC_deseq$Gene
rank_MDS_HSPC<-rank_MDS_HSPC[is.finite(rank_MDS_HSPC)]
MDS_HSPC_Hall <- fgsea(hallmakers_set, rank_MDS_HSPC, minSize=15, maxSize = 500, nperm=1000)

head(MDS_HSPC_Hall[order(padj), ], n=10)
MDS_HSPC_Hall_sig<-MDS_HSPC_Hall%>%filter(padj<0.05)%>%arrange(NES)
MDS_HSPC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSPC_Hall_sig$NES<0,"Mutant","NA")
MDS_HSPC_Hall_sig[1]$leadingEdge
MDS_HSPC_Hall_sig[9]$leadingEdge
MDS_HSPC_Hall_sig%>%group_by(Enrichment)%>%top_n(n=10,wt=abs(NES))%>%
  mutate(pathway=factor(pathway, levels=MDS_HSPC_Hall_sig$pathway))%>%ggplot(aes(NES,pathway,fill= padj))+geom_col()+scale_fill_gradient(low="red",high="blue")+theme_classic()


gene<-"NAMPT"
MDS_HSPC_gene<-MDS_HSPC_deseq[MDS_HSPC_deseq$Gene==gene,]
MDS_HSPC_gene<-MDS_HSPC_gene[1:8]

MDS_HSPC_gene<-data.frame(Gene=as.numeric(MDS_HSPC_gene),Group=rep(c("mutant","unknown"),4),name=c(rep(c("MDS10209"),2),
                                                                                                   rep(c("MDS124"),2),
                                                                                                   rep(c("MDS345"),2),
                                                                                                   rep(c("MDSSF3B1"),2)),Fullname=names(MDS_HSPC_gene))

t.test(MDS_HSPC_gene$Gene[MDS_HSPC_gene$Group=="mutant"],MDS_HSPC_gene$Gene[MDS_HSPC_gene$Group=="unknown"],paired=T)

MDS_HSPC_gene%>%ggplot(aes(Group,Gene,color=Group))+geom_boxplot()+geom_point()+ggtitle(gene)+theme_bw()+scale_color_manual(values=c("tomato4","steelblue4"))+
  geom_line(data=MDS_HSPC_gene,mapping = aes(Group,Gene,group=name),color="grey")+NoLegend()



MDS_HSPC_gene%>%ggplot(aes(Group,Gene))+geom_point()+ggtitle(gene)+theme_bw()+stat_compare_means(paired=T,method="t.test")

ggpaired(MDS_HSPC_gene, x = "Group", y = "Gene",id="name",
         color = "name", line.color = "gray", line.size = 0.4,
         palette = "name")+
  stat_compare_means(paired = TRUE,method="t.test")


############
#MDS vs NBM in E lineage

######


HSC_markers_deseq<-run_pseudobulkDE(meta_data=NBM_E@meta.data,expr_mat=NBM_E@assays$RNA@counts,contrasts="condition,HSC/MPPs,Others",
                                    design= "~condition+Extra", sample_col="Sample",condition_col="HSC_group",cluster_col="HSPC_type", extr_col ="Sample")
HSC_markers_deseq["Gene"]<-row.names(HSC_markers_deseq)
HSC_markers_deseq["Rank"]<- -sign(HSC_markers_deseq$log2FoldChange)*log10(HSC_markers_deseq$padj)
HSC_markers<-HSC_markers_deseq%>%top_n(n=50,wt=Rank)
HSC_markers<-HSC_markers[c("Gene","log2FoldChange","padj")]





MDS_NBM_MEP<-subset(MDS_NBM_E, ident="MEPs")
MDS_NBM_EProg<-subset(MDS_NBM_E, ident="Late erythroid progenitors")

MDS_NBM_E_MEP_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_MEP@meta.data,expr_mat=MDS_NBM_MEP@assays$RNA@counts,contrasts="condition,MDS,NBM",
                                  design= "~condition", sample_col="Sample",condition_col="Group",cluster_col="HSPC_type")
MDS_NBM_E_MEP_deseq["Gene"]<-row.names(MDS_NBM_E_MEP_deseq)


MDS_NBM_EProg_deseq<-run_pseudobulkDE(meta_data=MDS_NBM_EProg@meta.data,expr_mat=MDS_NBM_EProg@assays$RNA@counts,contrasts="condition,MDS,NBM",
                                  design= "~condition", sample_col="Sample",condition_col="Group",cluster_col="HSPC_type")
MDS_NBM_EProg_deseq["Gene"]<-row.names(MDS_NBM_EProg_deseq)

MDS_NBM_EProg_deseq_sig<-MDS_NBM_EProg_deseq[which(MDS_NBM_EProg_deseq$padj<0.05),]
MDS_NBM_EProg_deseq_top<-MDS_NBM_EProg_deseq_sig%>%mutate(gene_change=case_when(.$log2FoldChange>0~"Up", TRUE~"Down"))%>%
  group_by(gene_change)%>%top_n(n=30,wt="rank")%>%ungroup()

MDS_NBM_EProg_deseq%>%ggplot(aes(log2FoldChange,-log10(padj)))+geom_point(size=0.1,col="grey")+
  geom_point(data=MDS_NBM_EProg_deseq_sig,mapping=aes(log2FoldChange,-log10(padj)),col="black",size=0.5)+
  geom_vline(xintercept= c(0.5,-0.5),col="grey",type=2)+geom_hline(yintercept = -log10(0.05),col="grey",type=2)+
  geom_text_repel(MDS_NBM_EProg_deseq_top,mapping=aes(log2FoldChange,-log10(padj),label=Gene),max.overlaps = 10)+
  theme_classic()

#GSEA

MDS_NBM_EProg_deseq$pvalue[MDS_NBM_EProg_deseq$pvalue==0]<- 1e-103
rank_Eprog<-  -sign(MDS_NBM_EProg_deseq$log2FoldChange)*log10(MDS_NBM_EProg_deseq$pvalue)
names(rank_Eprog)<-row.names(MDS_NBM_EProg_deseq)
rank_Eprog<-rank_Eprog[!is.na(rank_Eprog)]

MDS_Eprog_Hall <- fgsea(hallmakers_set, rank_Eprog, minSize=15, maxSize = 500, nperm=10000)
head(MDS_Eprog_Hall[order(padj), ], n=10)
MDS_Eprog_Hall_sig<-MDS_Eprog_Hall%>%filter(padj<0.05)%>%arrange(desc(NES))
MDS_Eprog_Hall_sig[["Enrichment"]]<-ifelse(MDS_Eprog_Hall_sig$NES<0,"NBM","MDS")
MDS_Eprog_Hall_sig[1]$leadingEdge
MDS_Eprog_Hall_sig%>%mutate(pathway=factor(pathway, levels=MDS_Eprog_Hall_sig$pathway))%>%group_by(Enrichment)%>%
  top_n(10,wt=NES)%>%ggplot(aes(NES,pathway,size= -log10(padj)))+geom_point()+theme_classic()


plotEnrichment(hallmakers_set[["HALLMARK_TNFA_SIGNALING_VIA_NFKB"]], rank_Eprog,)+labs(title="HALLMARK_TNFA_SIGNALING_VIA_NFKB")
plotEnrichment(hallmakers_set[["HALLMARK_APOPTOSIS"]], rank_HSC,)+labs(title="HALLMARK_APOPTOSIS")
plotEnrichment(hallmakers_set[["HALLMARK_INTERFERON_GAMMA_RESPONSE"]], rank_HSC,)+labs(title="HALLMARK_INTERFERON_GAMMA_RESPONSE")


MDS_Eprog_C2 <- fgsea(C2_set, rank_Eprog, minSize=15, maxSize = 500, nperm=10000)
head(MDS_Eprog_C2[order(padj), ], n=10)
MDS_Eprog_C2_sig<-MDS_Eprog_C2%>%filter(padj<0.25)%>%arrange(desc(NES))




MDS_Eprog_C5 <- fgsea(C5_set, rank_Eprog, minSize=15, maxSize = 500, nperm=10000)
head(MDS_Eprog_C5[order(padj), ], n=10)
MDS_Eprog_C5_sig<-MDS_Eprog_C5%>%filter(padj<0.25)%>%arrange(desc(NES))


#####GSEA genes in Heatmap

MDS_NBM_E_sample<-subset(MDS_NBM_E, cells=
                           c(names(MDS_NBM_E$Group[MDS_NBM_E$Group=="MDS"]),
                             sample(names(MDS_NBM_E$Group[MDS_NBM_E$Group=="NBM"]),2800,replace = F)))


MDS_NBM_E_matrix<-MDS_NBM_E_sample@assays$RNA
MDS_NBM_E_matrix<-MDS_NBM_E_matrix[MDS_Eprog_Hall_sig[1]$leadingEdge[[1]],]
MDS_NBM_E_matrix<-t(scale(t(MDS_NBM_E_matrix)))
MDS_NBM_E_matrix<-apply(MDS_NBM_E_matrix,c(1,2),function(x){
  a<-ifelse(x>2,2,x)
  a<-ifelse(a< -2, -2,a)
  return(a)
})
MDS_NBM_E_matrix<-MDS_NBM_E_matrix[apply(MDS_NBM_E_matrix,1,function(x) any(!is.na(x))),]

annotation_pseud_order<-MDS_NBM_E_sample@meta.data%>%arrange(pseudotime)
annotation_pseud_order_NBM<-annotation_pseud_order[annotation_pseud_order$Group=="NBM",]
annotation_pseud_order_MDS<-annotation_pseud_order[annotation_pseud_order$Group=="MDS",]


annotaton_col<-list(HSPC_type=c("HSC/MPPs"="grey80","MEPs"="tomato4","Erythroid progenitors"="red3"))
NBM_heatmap<-pheatmap(MDS_NBM_E_matrix[,row.names(annotation_pseud_order_MDS)],use_raster=T,annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_MDS[c("HSPC_type","pseudotime")],show_colnames=F,show_rownames=F,cluster_cols = F,cluster_rows = T,clustering_method = "ward.D")
row_NBM <- row_order(NBM_heatmap)


pheatmap(MDS_NBM_E_matrix[,row.names(annotation_pseud_order_NBM)],use_raster=T,
         annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_NBM[c("HSPC_type")],show_colnames=F, show_rownames = F,
         cluster_cols = F,cluster_rows = F,row_order=row_NBM)

  #heatmap NBM


pheatmap(MDS_NBM_E_matrix[,row.names(annotation_pseud_order_MDS)],use_raster=T,
         annotation_colors = annotaton_col,annotation_col=annotation_pseud_order_MDS[c("HSPC_type")],show_colnames=F, show_rownames = F,
         cluster_cols = F,cluster_rows = F,row_order=row_NBM,
         clustering_method = "ward.D2")
#heatmap MDS




#SF3B1mt predicted by splicing
############

MDS_NBM_all_mt<-readRDS("~/single_cell_sequencing/MDS single cell//MDS_NBM_all_clustify_mutant_WT.RDS")

MT_prediction<-read.table("~/single_cell_sequencing/MDS single cell/MRLC_classification.tsv",header=T)

MT_prediction$Sample<-gsub("_HM","",MT_prediction$sample)
MT_prediction$Sample<-gsub("_NI","",MT_prediction$Sample)

MT_prediction["type"]<-ifelse(grepl("NI",MT_prediction$sample),"niche_immune","")
MT_prediction["type"]<-ifelse(grepl("HM",MT_prediction$sample),"cd34_myeloid",MT_prediction$type)




MT_prediction["bar_code"]<-tolower(paste0(MT_prediction$Sample,"_",MT_prediction$type,"_",MT_prediction$barcode,"_1"))
MT_prediction["bar_code"]<-gsub("__","_",MT_prediction$bar_code)

head( MDS_NBM_all_mt$bar_code %in% MT_prediction$bar_code)
sum(MDS_NBM_all_mt$bar_code %in% MT_prediction$bar_code
)

MT_prediction<-MT_prediction[c("Prediction","bar_code")]
metadata_MDS<-MDS_NBM_all_mt@meta.data

metadata_MDS<-metadata_MDS%>%left_join(MT_prediction)
row.names(metadata_MDS)<-metadata_MDS$bar_code
metadata_MDS<-metadata_MDS[MDS_NBM_all_mt$bar_code,]
MDS_NBM_all_mt[["SF3B1_prediction"]]<-metadata_MDS$Prediction
MDS_NBM_all_mt[["SF3B1_prediction"]]<-ifelse(MDS_NBM_all_mt$SF3B1_prediction,"MT","Unknown")
MDS_NBM_all_mt[["SF3B1_prediction"]]<-ifelse(grepl("NBM",MDS_NBM_all_mt$Group),"NBM",MDS_NBM_all_mt$SF3B1_prediction)


DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"),split.by = "Group",label = T,repel=T,cells.highlight = names(MDS_NBM_all_mt$prediction)[MDS_NBM_all_mt$SF3B1_prediction=="MT"], sizes.highlight = 0.3,pt.size=0.1,raster=F)
DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"),split.by = "Group",label = T,cells.highlight = names(MDS_NBM_all_mt$prediction)[MDS_NBM_all_mt$Mutant_state=="Mutant"], sizes.highlight = 0.2,pt.size=0.3,raster=F)
summary(factor(MDS_NBM_all_mt$SF3B1_prediction))
summary(factor(MDS_NBM_all_mt$Mutant_state))


DimPlot(MDS_NBM_all_mt,split.by = "Group")



HSPC_NBM_MDS<-subset(MDS_NBM_all_mt,ident=c("HSC/MPPs","LMPPs","GMPs","MEPs","CLPs"))



HSPC_NBM_MDS <- ScaleData(HSPC_NBM_MDS);
DefaultAssay(HSPC_NBM_MDS)<-"integrated"

HSPC_NBM_MDS <- RunPCA(HSPC_NBM_MDS, npcs = 30,verbose = FALSE);
ElbowPlot(HSPC_NBM_MDS)
nDims=10
HSPC_NBM_MDS <- FindNeighbors(HSPC_NBM_MDS, reduction = "pca", dims = 1:nDims)
HSPC_NBM_MDS <- FindClusters(HSPC_NBM_MDS, resolution=0.4);
HSPC_NBM_MDS <- RunUMAP(HSPC_NBM_MDS,dims = 1:nDims);
HSPC_NBM_MDS$SF3B1_prediction[is.na(HSPC_NBM_MDS$SF3B1_prediction)]<-"Unknown"
HSPC_NBM_MDS$SF3B1_prediction<-factor(HSPC_NBM_MDS$SF3B1_prediction,levels=c("NBM","MT","Unknown"))


Idents(HSPC_NBM_MDS)<-"type_re"
Idents(HSPC_NBM_MDS)<-"seurat_clusters"

DimPlot(HSPC_NBM_MDS, reduction = "umap", label=T,pt.size = 0.25,split.by = "Group")

DimPlot(HSPC_NBM_MDS,split.by = "Sample",cells.highlight = names(MDS_NBM_all_mt$prediction)[MDS_NBM_all_mt$SF3B1_prediction=="MT"],pt.size=0.1,raster=F)
FeaturePlot(HSPC_NBM_MDS,split.by = "Group",features="CD34",min.cutoff = 0.25,order=T,pt.size=0.1)
DotPlot(HSPC_NBM_MDS,features=c("MAP3K8"),group.by = "SF3B1_prediction",assay="RNA")

VlnPlot(HSPC_NBM_MDS,features=c("MAP3K8"),group.by = "SF3B1_prediction",assay="RNA",pt.size=0)
VlnPlot(HSPC_NBM_MDS,features=c("BTG1"),group.by = "SF3B1_prediction",assay="RNA",pt.size=0)



#HSPC_NBM_MDS<-subset(HSPC_NBM_MDS,ident=c("9","2","12","14"),inver=T)

summary(factor(HSPC_NBM_MDS$SF3B1_prediction))

HSPC_MDS<-subset(HSPC_NBM_MDS,subset=Group=="MDS")
HSPC_MDS<-subset(HSPC_MDS,subset=Sample=="MDS_337",invert=T)

DimPlot(HSPC_MDS, reduction = "umap", label=T,pt.size = 0.25,split.by = "Sample")



DimPlot(subset(HSPC_MDS,subset=Group=="MDS"),split.by = "Group",label = T,repel=T,cells.highlight = names(MDS_NBM_all_mt$prediction)[MDS_NBM_all_mt$SF3B1_prediction=="MT"], sizes.highlight = 0.3,pt.size=0.2,raster=F)
DimPlot(subset(HSPC_MDS,subset=Group=="MDS"),split.by = "Group",label = T,cells.highlight = names(MDS_NBM_all_mt$prediction)[MDS_NBM_all_mt$Mutant_state=="Mutant"], sizes.highlight = 0.3,pt.size=0.2,raster=F)
summary(factor(HSPC_MDS$SF3B1_prediction))
summary(factor(HSPC_MDS$Mutant_state))

MDS_HSPC_deseq<-run_pseudobulkDE_new(data=HSPC_MDS,
                                     contrasts='condition,MT,Unknown',design='~condition+Extra',
                                     sample_col="Sample", condition_col="SF3B1_prediction",cluster_col="type_re",extr_col='Sample',do_prefiltering = TRUE)



MDS_HSPC_deseq["Gene"]<-row.names(MDS_HSPC_deseq)
MDS_HSPC_deseq["rank"]<- -sign(
  MDS_HSPC_deseq$log2FoldChange)*log2(MDS_HSPC_deseq$pvalue)
MDS_HSPC_deseq<-MDS_HSPC_deseq[order(MDS_HSPC_deseq$padj),]
MDS_HSPC_deseq_sig<-MDS_HSPC_deseq[MDS_HSPC_deseq$padj<0.05,]


NBM_Mutant_HSPC_deseq<-run_pseudobulkDE_new(data=HSPC_NBM_MDS,
                                            contrasts='condition,MT,NBM',design='~condition',
                                            sample_col="Sample", condition_col="SF3B1_prediction",cluster_col="type_re",extr_col=NULL,do_prefiltering = TRUE)



NBM_Mutant_HSPC_deseq["Gene"]<-row.names(NBM_Mutant_HSPC_deseq)
MDS_HSPC_deseq["rank"]<- -sign(
  MDS_HSPC_deseq$log2FoldChange)*log2(MDS_HSPC_deseq$pvalue)
MDS_HSPC_deseq<-MDS_HSPC_deseq[order(MDS_HSPC_deseq$padj),]
MDS_HSPC_deseq_sig<-MDS_HSPC_deseq[MDS_HSPC_deseq$padj<0.05,]

write_xlsx(MDS_HSPC_deseq,"MDS_HSPCs_MT_vs_NA_Predict.xlsx")

MDS_HSPC_deseq%>%ggplot(aes(-log10(padj),log2FoldChange))+
  geom_point(alpha=0.2,color="grey80")+geom_vline(xintercept= -log10(0.05),linetype=3,cols="grey")+
  geom_point(data=MDS_HSPC_deseq_sig,mapping=aes(-log10(padj),log2FoldChange),alpha=0.2)+
  geom_text_repel(data=MDS_HSPC_deseq_sig,mapping=aes(-log10(padj),log2FoldChange,label=Gene),size=3,max.overlaps=30)+
  theme_bw()



#write_xlsx(MDS_HSPC_deseq,"single_cell_sequencing/MDS single cell/HSPC_MT_vs_NA.xlsx")

rank_MDS_HSPC<-MDS_HSPC_deseq$rank
names(rank_MDS_HSPC)<-MDS_HSPC_deseq$Gene
rank_MDS_HSPC<-rank_MDS_HSPC[is.finite(rank_MDS_HSPC)]
MDS_HSPC_Hall <- fgsea(hallmakers_set, rank_MDS_HSPC, minSize=15, maxSize = 500, nperm=1000)

head(MDS_HSPC_Hall[order(padj), ], n=10)
MDS_HSPC_Hall_sig<-MDS_HSPC_Hall%>%filter(padj<0.05)%>%arrange(NES)
MDS_HSPC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSPC_Hall_sig$NES<0,"Mutant","NA")
MDS_HSPC_Hall_sig[1]$leadingEdge
MDS_HSPC_Hall_sig[9]$leadingEdge
MDS_HSPC_Hall_sig%>%group_by(Enrichment)%>%top_n(n=10,wt=abs(NES))%>%
  mutate(pathway=factor(pathway, levels=MDS_HSPC_Hall_sig$pathway))%>%ggplot(aes(NES,pathway,fill= padj))+geom_col()+scale_fill_gradient(low="red",high="blue")+theme_classic()




MDS_CD34_niche_cor<-read.csv("GDCdata/MDS-GSE58831/Cor_iBMSC.csv")
MDS_HSPC_deseq_sig<-MDS_HSPC_deseq[MDS_HSPC_deseq$pvalue<0.01,]

MDS_CD34_niche_cor_sig<-MDS_CD34_niche_cor[MDS_CD34_niche_cor$cor_padj<0.05,]
#MDS_CD34_niche_cor_sig<-MDS_CD34_niche_cor_sig[MDS_CD34_niche_cor_sig$Cor_rho>0,]
intersect(MDS_CD34_niche_cor_sig$name,MDS_HSPC_deseq_sig$Gene)


#Overlap 
Cor_iBMSC<-read.csv("single_cell_sequencing/MDS single cell/Cor_iBMSC.csv")[-1]
Cor_iBMSC_sig<-Cor_iBMSC[Cor_iBMSC$cor_padj<0.05,]

Cor_iBMSC_MT_vs_Ctrl<-MDS_HSPC_deseq[MDS_HSPC_deseq$Gene %in% Cor_iBMSC_sig$name,]
Cor_iBMSC_MT_vs_Ctrl<-Cor_iBMSC_MT_vs_Ctrl%>%rename("name"=Gene)

Cor_iBMSC_MT_vs_Ctrl<-Cor_iBMSC_MT_vs_Ctrl%>%inner_join(Cor_iBMSC_sig)
Cor_iBMSC_MT_vs_Ctrl<-Cor_iBMSC_MT_vs_Ctrl%>%arrange(cor_padj)

write.xlsx(Cor_iBMSC_MT_vs_Ctrl,"single_cell_sequencing/MDS single cell/Cor_iBMSC_MT_vs_Ctrl.xlsx")


##########
#HSPC deconvolution cibersort
###############
#https://cibersortx.stanford.edu/

NBM_HSPC<-subset(MDS_NBM_HSPC,subset=Group=="LRMDS")

NBM_HSPC_matrix<-as.matrix(NBM_HSPC@assays$RNA@counts)
NBM_HSPC_title<-matrix(as.character(NBM_HSPC$HSPC_type),nrow=1)
rownames(NBM_HSPC_title)<-"gene"
colnames(NBM_HSPC_title)<-colnames(NBM_HSPC_matrix)
NBM_HSPC_matrix<-rbind(NBM_HSPC_title,NBM_HSPC_matrix)


write.table(NBM_HSPC_matrix,"~/NBM_HSPC_matrix.txt",sep="\t",quote=F,col.names = F)


LRMDS_all<-subset(MDS_NBM_HSPC,subset=Group=="LRMDS")

NBM_HSPC_matrix<-as.matrix(NBM_HSPC@assays$RNA@counts)
NBM_HSPC_title<-matrix(as.character(NBM_HSPC$HSPC_type),nrow=1)
rownames(NBM_HSPC_title)<-"gene"
colnames(NBM_HSPC_title)<-colnames(NBM_HSPC_matrix)
NBM_HSPC_matrix<-rbind(NBM_HSPC_title,NBM_HSPC_matrix)


write.table(NBM_HSPC_matrix,"~/NBM_HSPC_matrix.txt",sep="\t",quote=F,col.names = F)




MDS_NBM_all_mt[["type_all"]]<-as.character(MDS_NBM_all_mt$type_re)

MDS_NBM_all_type_all<-MDS_NBM_all_mt$type_all

MDS_NBM_all_immune_all<-MDS_NBM_Immune$Immune_type
MDS_NBM_all_type_all<-MDS_NBM_all_type_all[!names(MDS_NBM_all_type_all) %in% names(MDS_NBM_all_immune_all)]
MDS_NBM_all_type_all<-c(MDS_NBM_all_type_all,MDS_NBM_all_immune_all)
MDS_NBM_all_mt[["type_all"]]<-MDS_NBM_all_type_all[names(MDS_NBM_all_mt$type_re)]
Idents(MDS_NBM_all_mt)<-"type_all"
DimPlot(subset(MDS_NBM_all_mt,subset=Group=="LRMDS"), reduction = "umap", label=T,pt.size = 0.25,split.by = "Group",ncol=1)


LRMDS_all<-subset(MDS_NBM_all_mt,subset=Group=="LRMDS")
Idents(LRMDS_all)<-"type_all"
LRMDS_all<-subset(LRMDS_all,subset=type_all=="NK",invert=T)
LRMDS_all<-subset(LRMDS_all,subset=type_all=="BMSCs",invert=T)

DimPlot(LRMDS_all, reduction = "umap", label=T,pt.size = 0.25,split.by = "Group",ncol=1)

LRMDS_all<-subset(LRMDS_all,cells=sample(names(LRMDS_all$orig.ident),20000,replace = F))
LRMDS_all_matrix<-as.matrix(LRMDS_all@assays$RNA@counts)
LRMDS_all_title<-matrix(as.character(LRMDS_all$type_all),nrow=1)
rownames(LRMDS_all_title)<-"gene"
colnames(LRMDS_all_title)<-colnames(LRMDS_all_matrix)
LRMDS_all_matrix<-rbind(LRMDS_all_title,LRMDS_all_matrix)

write.table(LRMDS_all_matrix,"single_cell_sequencing/CIBERSORT/MDS/LRMDS_scRNAseq_all_matrix.txt",sep="\t",quote=F,col.names = F)





################
#Survival analysis
#############
#take data from CD34 R

#Survival analysis
#############


#Ligand receptor interaction
#################

#HSPC niche inflamed high vs.low
##################

HSPC_niche_inflamed<-data.frame(read_xlsx("single_cell_sequencing/MDS single cell/HSPC niche_inflammation_deseq/HSPC_iBMSC_high_vs_low.xlsx"))
MDS_CD34<-read.csv("MDS survival analysis/LRMDS_vs_control_HSC_DEseq.csv")[-1]


HSPC_niche_inflamed%>%filter(name=="BIRC3")

HSPC_niche_inflamed_sig<-HSPC_niche_inflamed%>%filter(padj<0.05)
HSPC_niche_inflamed_sig<-HSPC_niche_inflamed_sig[c(2,80:86)]

HSPC_niche_inflamed%>%ggplot(aes(-log10(padj),log2FoldChange))+geom_hline(yintercept= -0.5,linetype=3,color="grey")+
  geom_hline(yintercept=0.5,linetype=3,color="grey")+geom_vline(xintercept= -log10(0.05),linetype=3,color="grey")+
  geom_point(alpha=0.2)+
  geom_text_repel(data=HSPC_niche_inflamed_sig,mapping=aes(-log10(padj),log2FoldChange,label=name),size=3,max.overlaps=30)+
  theme_classic()



HSPC_niche_inflamed<-HSPC_niche_inflamed[!duplicated(HSPC_niche_inflamed$name),]
HSPC_niche_inflamed<-HSPC_niche_inflamed[!is.na(HSPC_niche_inflamed$name),]
HSPC_niche_inflamed_sig<-HSPC_niche_inflamed[HSPC_niche_inflamed$padj<0.05,]

HSPC_niche_inflamed_matrix<-data.frame(HSPC_niche_inflamed[4:41])
row.names(HSPC_niche_inflamed_matrix)<-HSPC_niche_inflamed$name

Var_gene<-apply(HSPC_niche_inflamed_matrix,c(1),var)
names(Var_gene)<-row.names(HSPC_niche_inflamed_matrix)
Var_gene_top<-head(Var_gene,n=1000)

Annotation_HSPC_niche<-read.csv("MDS survival analysis/MDS_Niche_signature.csv",header = T)[-1]
row.names(Annotation_HSPC_niche)<-Annotation_HSPC_niche$name
Annotation_HSPC_niche<-data.frame(t(Annotation_HSPC_niche[-56]))
Annotation_HSPC_niche["iBMSC_score"]<-ifelse(Annotation_HSPC_niche$BMSC2_up_sig_gene>
                                         quantile(Annotation_HSPC_niche$BMSC2_up_sig_gene,0.5),"High","Low")


row.names(Annotation_HSPC_niche)<-str_split(row.names(Annotation_HSPC_niche),"_",simplify = T)[,1]
#row.names(Annotation_HSPC_niche)<-Annotation_HSPC_niche$name
#Annotation_HSPC_niche$name<-str_split(Annotation_HSPC_niche$name,"_",simplify = T)[,1]


HSPC_niche_inflamed_matrix<-t(scale(t(HSPC_niche_inflamed_matrix)))
HSPC_niche_inflamed_matrix_heatmap<-apply(HSPC_niche_inflamed_matrix,c(1,2),function(x){
  a<-ifelse(x>3,3,x)
  a<-ifelse(x< -3, -3, a)
})

colnames(HSPC_niche_inflamed_matrix_heatmap)<-str_split(colnames(HSPC_niche_inflamed_matrix_heatmap),"_",simplify = T)[,1]

annotation_color<-list(niche_inflame=c("High"="tomato4","Low"="steelblue4"),iBMSC_score=colorRampPalette(c("white","grey","red"))(20))


pheatmap(as.matrix(HSPC_niche_inflamed_matrix_heatmap)[names(Var_gene_top),],annotation_col = Annotation_HSPC_niche[2],
         annotation_colors = annotation_color,
         clustering_method = "ward.D2",show_colnames = F,show_rownames = F)


HSPC_niche_inflamed_matrix_cor<-cor(as.matrix(HSPC_niche_inflamed_matrix_heatmap)[names(Var_gene_top),])

pheatmap(HSPC_niche_inflamed_matrix_cor,color=colorRampPalette(c("blue","black","yellow2"))(30),annotation_col = Annotation_HSPC_niche[3],
         annotation_colors = annotation_color,
         clustering_method = "complete",show_colnames =T,show_rownames =T,annotation_names_row = F)

corrplot(HSPC_niche_inflamed_matrix_cor)

heatmap_MDS<-pheatmap(HSPC_niche_inflamed_matrix_cor,annotation_col = Annotation_HSPC_niche[2],
                      color = colorRampPalette(c("blue","black","yellow3"))(30),
                      annotation_colors = annotation_color,
                      clustering_method = "complete")



PCA_HSPC_niche_inflamed<-data.frame(prcomp(t(as.matrix(HSPC_niche_inflamed_matrix)[names(Var_gene_top),]), scale= T)$x)
row.names(PCA_HSPC_niche_inflamed)<-str_split(row.names(PCA_HSPC_niche_inflamed),"_",simplify = T)[,1]

PCA_HSPC_niche_inflamed["patients"]<-row.names(PCA_HSPC_niche_inflamed)
Annotation_HSPC_niche["patients"]<-row.names(Annotation_HSPC_niche)


PCA_HSPC_niche_annotation<-PCA_HSPC_niche_inflamed%>%inner_join(Annotation_HSPC_niche)

PCA_HSPC_niche_annotation%>%ggplot(aes(PC1,PC2,color=iBMSC_score))+geom_point()+theme_classic()+scale_color_manual(values=c("tomato4","steelblue4"))




rank_HSPC<- -sign(HSPC_niche_inflamed$log2FoldChange)*log10(HSPC_niche_inflamed$pvalue)
names(rank_HSPC)<-HSPC_niche_inflamed$name
rank_HSPC<-rank_HSPC[!is.na(rank_HSPC)]


HSPC_NicheInflammation_Hall <- fgsea(hallmakers_set, rank_HSPC, minSize=15, maxSize = 500, nperm=10000)

HSPC_NicheInflammation_Hall_sig<-HSPC_NicheInflammation_Hall%>%filter(padj<0.05)%>%arrange(NES)
HSPC_NicheInflammation_Hall_sig[["Enrichment"]]<-ifelse(HSPC_NicheInflammation_Hall_sig$NES>0,"High","Low")
HSPC_NicheInflammation_Hall_sig[HSPC_NicheInflammation_Hall_sig$pathway=="HALLMARK_TNFA_SIGNALING_VIA_NFKB"]$leadingEdge
HSPC_NicheInflammation_Hall_sig%>%mutate(pathway=factor(pathway, levels=HSPC_NicheInflammation_Hall_sig$pathway))%>%group_by(Enrichment)%>%
  top_n(15,wt=NES)%>%ggplot(aes(NES,pathway,fill= padj))+geom_col()+scale_fill_gradient(low="red",high="blue")+theme_classic()

HSPC_NicheInflammation_Hall_sig[order(NES,decreasing=T),]

HSPC_NicheInflammation_C2 <- fgsea(C2_set, rank_HSPC, minSize=15, maxSize = 500, nperm=10000)

HSPC_NicheInflammation_C2_sig<-HSPC_NicheInflammation_C2%>%filter(padj<0.25)%>%arrange(desc(NES))
HSPC_NicheInflammation_C2_sig[["Enrichment"]]<-ifelse(HSPC_NicheInflammation_C2_sig$NES>0,"High","Low")

HSPC_NicheInflammation_C5 <- fgsea(C5_set, rank_HSPC, minSize=15, maxSize = 500, nperm=10000)
HSPC_NicheInflammation_C5_sig<-HSPC_NicheInflammation_C5%>%filter(padj<0.05)%>%arrange(desc(NES))
HSPC_NicheInflammation_C5_sig[["Enrichment"]]<-ifelse(HSPC_NicheInflammation_C5_sig$NES>0,"High","Low")



##########
#HSPC MDS vs. NBM bulk
#############

#take data from CD34 R

MDS_CD34<-read.csv("MDS survival analysis/LRMDS_vs_control_HSC_DEseq.csv")[-1]
MDS_patient<-read.csv("MDS survival analysis/Overall_survival.csv")[-1]

MDS_patient<-read_xlsx("scRNAseq_dataset/RDSfile/chenMDS/MDS single cell/")

MDS_patient<-MDS_patient%>%mutate(name=paste("X",.$name,sep=""))
MDS_patient<-MDS_patient%>%mutate(pfsi_event=case_when(.$pfsi..04.12.2020=="PD"~1, TRUE~0))

MDS_CD34[MDS_CD34$name=="MKI67",]

MDS_CD34_analysis<-MDS_CD34%>%select(name,log2FoldChange,padj)%>%mutate(log_padj=-log(padj,10))

MDS_CD34_analysis_sig<-MDS_CD34_analysis%>%filter(padj<=0.05)%>%mutate(change=case_when(.$log2FoldChange>0~"up",.$log2FoldChange<0~"down"))%>%
  mutate(change=factor(change,levels=c("up","down")))%>%filter(log2FoldChange<8)

MDS_CD34_analysis_sig_repel<-MDS_CD34_analysis_sig%>%
  filter(name%in%c("NFKBIA","GADD45A","BTG3","NFKB2",
                   "NFKBIE","GADD45B","ATF3","RIPK2","RELB","PTX3","BHLHE40","NFKB1","NFKB2","DDIT3","BCL10","SAT1","CDKN1A","PMAIP1"))


MDS_CD34_analysis%>%filter(log2FoldChange<8)%>%ggplot(aes(log2FoldChange,log_padj))+
  geom_point(size=1,col="grey70")+geom_vline(aes(xintercept = 0),col="grey70")+
  geom_hline(aes(yintercept = 0),col="grey70")+scale_color_manual(values=c("tomato4","steelblue4"))+
  labs(x="log2_FoldChange", y="-log10_padj",size=4)+
  geom_point(data=MDS_CD34_analysis_sig,mapping=aes(log2FoldChange,log_padj,col=change),size=1,alpha=0.1)+
  theme(legend.title=element_blank())+theme_bw()+
  geom_text_repel(data=MDS_CD34_analysis_sig_repel,mapping=aes(log2FoldChange,log_padj,label=name),size=3,max.overlaps=30)

MDS_CD34$pvalue[MDS_CD34$pvalue==0]<- 1e-103
rank_CD34<-  -sign(MDS_CD34$log2FoldChange)*log10(MDS_CD34$pvalue)
names(rank_CD34)<-MDS_CD34$name
rank_CD34<-rank_CD34[!is.na(rank_CD34)]

MDS_CD34_sig<-MDS_CD34%>%filter(padj<0.05)
MDS_CD34_sig<-MDS_CD34_sig[c(1,55:61)]
HSPC_niche_inflamed_sig
Intersection<-MDS_CD34_sig[which(MDS_CD34_sig$name %in% HSPC_niche_inflamed_sig$name),]


MDS_CD34_Hall <- fgsea(hallmakers_set, rank_CD34, minSize=15, maxSize = 500, nperm=10000)
head(MDS_CD34_Hall[order(padj,decreasing=T), ], n=10)
MDS_CD34_Hall<-MDS_CD34_Hall%>%filter(padj<0.25)%>%arrange(NES)
MDS_CD34_Hall[["Enrichment"]]<-ifelse(MDS_CD34_Hall$NES<0,"NBM","MDS")
MDS_CD34_Hall[MDS_CD34_Hall$pathway=="HALLMARK_TNFA_SIGNALING_VIA_NFKB"]$leadingEdge
MDS_CD34_Hall[MDS_CD34_Hall$pathway=="HALLMARK_APOPTOSIS"]$leadingEdge
MDS_CD34_Hall[MDS_CD34_Hall$pathway=="HALLMARK_G2M_CHECKPOINT"]$leadingEdge
MDS_CD34_Hall[MDS_CD34_Hall$pathway=="HALLMARK_E2F_TARGETS"]$leadingEdge


MDS_CD34_Hall%>%mutate(pathway=factor(pathway, levels=MDS_CD34_Hall$pathway))%>%group_by(Enrichment)%>%top_n(n=10,wt=abs(NES))%>%
  ggplot(aes(NES,pathway,fill= padj))+geom_col()+theme_classic()+scale_fill_gradient(low="red",high="blue")

NFKB_marker_gene<-MDS_CD34_Hall[1]$leadingEdge[[1]]
Apoptosis_marker_gene<-MDS_CD34_Hall[5]$leadingEdge[[1]]
UV_marker_gene<-MDS_CD34_Hall[2]$leadingEdge[[1]]
P53_marker_gene<-MDS_CD34_Hall[4]$leadingEdge[[1]]

Apoptosis_marker_gene<-Reduce(intersect, list(Apoptosis_marker_gene,P53_marker_gene))

NFkB_Apoptosis_markers<-unique(NFKB_marker_gene,Apoptosis_marker_gene)


MDS_CD34<-MDS_CD34[!duplicated(MDS_CD34$name),]

MDS_CD34_deg<-MDS_CD34[c(1,55:61)]
MDS_CD34_deg_seq<-MDS_CD34_deg[MDS_CD34_deg$padj<0.05,]



MDS_CD34<-MDS_CD34[!is.na(MDS_CD34$name),]

MDS_CD34_matrix<-data.frame(MDS_CD34[3:54])
row.names(MDS_CD34_matrix)<-MDS_CD34$name



Annotation_HSPC_niche<-Annotation_HSPC_niche%>%full_join(MDS_patient)
MDS_all_patient_ID<-data.frame(name=colnames(MDS_CD34_matrix),sample="MDS")
MDS_all_patient_ID$sample<-ifelse(grepl("normal",MDS_all_patient_ID$name),"NBM","MDS")
Annotation_HSPC_niche<-Annotation_HSPC_niche%>%full_join(MDS_all_patient_ID)
Annotation_HSPC_niche$niche_inflammation<-ifelse(grepl("normal",Annotation_HSPC_niche$name),
                                                 "NBM",Annotation_HSPC_niche$niche_inflammation)
Annotation_HSPC_niche<-Annotation_HSPC_niche[!is.na(Annotation_HSPC_niche$niche_inflammation),]
row.names(Annotation_HSPC_niche)<-Annotation_HSPC_niche$name


MDS_CD34_matrix_scale<-t(scale(t(MDS_CD34_matrix)))
MDS_CD34_matrix_scale<-apply(MDS_CD34_matrix_scale,c(1,2),function(x){
  a<-ifelse(x>3,3,x)
  a<-ifelse(a< -3, -3, a)
})

MDS_CD34_matrix_scale<-MDS_CD34_matrix_scale[,Annotation_HSPC_niche$name]

MDS_CD34_matrix_scale_cor<-cor(MDS_CD34_matrix_scale[which(row.names(MDS_CD34_matrix_scale)%in%names(Var_gene_top)),
                                                     Annotation_HSPC_niche$name])


pheatmap(MDS_CD34_matrix_scale_cor,annotation_col = Annotation_HSPC_niche[2],
         color = colorRampPalette(c("blue","black","yellow3"))(30),
         annotation_colors = annotation_color,
         clustering_method = "complete")



annotation_color<-list(niche_inflammation=c("High"="tomato4","Low"="steelblue4","NBM"="grey80"))

pheatmap(MDS_CD34_matrix_scale[which(row.names(MDS_CD34_matrix_scale)%in%names(Var_gene_top)),Annotation_HSPC_niche$name],
         color = colorRampPalette(c("blue","black","yellow3"))(30),
         annotation_col = Annotation_HSPC_niche[2],
         annotation_colors = annotation_color,show_colnames = F,show_rownames = F,
         clustering_method = "ward.D2")


Gene_test<-"CD44"
MDS_CD34_matrix_gene<-MDS_CD34_matrix[Gene_test,]
MDS_CD34_matrix_gene<-data.frame(name=names(MDS_CD34_matrix),gene=as.numeric(MDS_CD34_matrix_gene))

MDS_CD34_matrix_gene<-MDS_CD34_matrix_gene%>%full_join(Annotation_HSPC_niche[c(1,2)])
MDS_CD34_matrix_gene$niche_inflame[grepl("normal",MDS_CD34_matrix_gene$name)]<-"NBM"
MDS_CD34_matrix_gene<-MDS_CD34_matrix_gene[!is.na(MDS_CD34_matrix_gene$niche_inflame),]

MDS_CD34_matrix_gene%>%rename("Group"=niche_inflame)%>%ggplot(aes(Group,gene,colors=Group))+geom_boxplot(outlier.size = -1)+
  geom_jitter()+theme_classic()


MDS_CD34_GSVA<-read.csv("MDS survival analysis/CD34_hall_ssGSEA.CSV")
row.names(MDS_CD34_GSVA)<-MDS_CD34_GSVA$X
MDS_CD34_GSVA<-MDS_CD34_GSVA[names(MDS_CD34_GSVA)!="X"]

Pathwahy<-"HALLMARK_TNFA_SIGNALING_VIA_NFKB"
MDS_CD34_GSVA_path<-MDS_CD34_GSVA[Pathwahy,]
MDS_CD34_GSVA_path<-data.frame(name=names(MDS_CD34_GSVA),path=as.numeric(MDS_CD34_GSVA_path))

MDS_CD34_GSVA_path<-MDS_CD34_GSVA_path%>%full_join(Annotation_HSPC_niche[c(1,2)])
MDS_CD34_GSVA_path$niche_inflammation[grepl("NBM",MDS_CD34_GSVA_path$name)]<-"NBM"
MDS_CD34_GSVA_path<-MDS_CD34_GSVA_path[!is.na(MDS_CD34_GSVA_path$niche_inflammation),]

MDS_CD34_GSVA_path%>%rename("Group"=niche_inflammation)%>%ggplot(aes(Group,path,colors=Group))+geom_boxplot(outlier.size = -1)+
  geom_jitter()+theme_classic()


#####
#GO_analysis unsing intersection genes between MDS vs. NBM and MDS inflamehigh vs inflamelow

Intersection_up<-Intersection[Intersection$log2FoldChange>0,]

write.csv("MDS survival analysis/CD34_hall_ssGSEA.CSV")


gostres <- gost(query = Intersection$name, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE, highlight = TRUE)

gostplot(gostres, interactive = F)


#Event and survival free based on single genes



Gene_name<-"DDIT4"
MDS_CD34_genes<-MDS_CD34%>%filter(name==Gene_name)
row.names(MDS_CD34_genes)<-MDS_CD34_genes$name
MDS_CD34_genes_t<-data.frame(t(MDS_CD34_genes[-c(1,2,55:61)]))
names(MDS_CD34_genes_t)<-"gene_expression"
MDS_CD34_genes_t["name"]<-row.names(MDS_CD34_genes_t)

CI95<-mean(MDS_CD34_genes_t$gene_expression[46:52])+qnorm(0.975)*sd(MDS_CD34_genes_t$gene_expression[46:52])

MDS_patient_gene<-MDS_patient%>%inner_join(MDS_CD34_genes_t,by="name")

MDS_patient_gene<-MDS_patient_gene%>%mutate(Gene_group=case_when(.$gene_expression < median(.$gene_expression)~"Low",TRUE~"High"))

surv_MDS<-Surv(event=MDS_patient_gene$osi,time=MDS_patient_gene$os)
Survfit_cluster<-surv_fit(surv_MDS~Gene_group,data=MDS_patient_gene)
ggsurvplot(Survfit_cluster,data=MDS_patient_gene,pval=T)


coxph(Surv(event=MDS_patient_gene$osi,time=MDS_patient_gene$os)~gene_expression,data=MDS_patient_gene)



MDS_CD34_niche_cor<-read.csv("GDCdata/MDS-GSE58831/Cor_iBMSC.csv")
MDS_HSPC_deseq_sig<-MDS_HSPC_deseq[MDS_HSPC_deseq$pvalue<0.01,]

MDS_CD34_niche_cor_sig<-MDS_CD34_niche_cor[MDS_CD34_niche_cor$cor_padj<0.05,]
#MDS_CD34_niche_cor_sig<-MDS_CD34_niche_cor_sig[MDS_CD34_niche_cor_sig$Cor_rho>0,]
intersect(MDS_CD34_niche_cor_sig$name,MDS_HSPC_deseq_sig$Gene)


#Overlap 
Cor_iBMSC<-read.csv("single_cell_sequencing/MDS single cell/Cor_iBMSC.csv")[-1]
Cor_iBMSC_sig<-Cor_iBMSC[Cor_iBMSC$cor_padj<0.05,]

Cor_iBMSC_MT_vs_Ctrl<-MDS_HSPC_deseq[MDS_HSPC_deseq$Gene %in% Cor_iBMSC_sig$name,]
Cor_iBMSC_MT_vs_Ctrl<-Cor_iBMSC_MT_vs_Ctrl%>%rename("name"=Gene)

Cor_iBMSC_MT_vs_Ctrl<-Cor_iBMSC_MT_vs_Ctrl%>%inner_join(Cor_iBMSC_sig)
Cor_iBMSC_MT_vs_Ctrl<-Cor_iBMSC_MT_vs_Ctrl%>%arrange(cor_padj)

write.xlsx(Cor_iBMSC_MT_vs_Ctrl,"single_cell_sequencing/MDS single cell/Cor_iBMSC_MT_vs_Ctrl.xlsx")

#####################
#Deconvolution
#
iBMSC_score<-read.csv("scRNAseq_dataset/RDSfile/chenMDS/MDS single cell/BMSC2_markergene.xlsx")[-1]





iBMSC_score<-data.frame(name=names(iBMSC_score)[-56],iBMSC_score=as.numeric(iBMSC_score[1,-56]))

MDS_HSPC_deconvolution<-read_xlsx("single_cell_sequencing/CIBERSORT/MDS/CIBERSORTx_HSPC_deconvolution.xlsx")
MDS_HSPC_deconvolution$name<-paste0("X",MDS_HSPC_deconvolution$name)
MDS_HSPC_deconvolution%>%ggplot(aes(Group,MEPs))+geom_boxplot()+geom_jitter()


MDS_HSPC_deconvolution_iBMSC_score<-MDS_HSPC_deconvolution%>%inner_join(iBMSC_score)

MDS_HSPC_deconvolution_iBMSC_score%>%ggplot(aes(iBMSC_score,GMPs))+geom_point()+geom_smooth(method="lm")
cor.test(MDS_HSPC_deconvolution_iBMSC_score$GMPs,MDS_HSPC_deconvolution_iBMSC_score$iBMSC_score)



MDS_HSPC_deconvolution_iBMSC_score<-MDS_HSPC_deconvolution_iBMSC_score%>%arrange(desc(MEPs))
MDS_HSPC_deconvolution_iBMSC_score<-MDS_HSPC_deconvolution_iBMSC_score%>%arrange(iBMSC_score)
MDS_HSPC_deconvolution_iBMSC_score$name<-factor(MDS_HSPC_deconvolution_iBMSC_score$name,levels=MDS_HSPC_deconvolution_iBMSC_score$name)
MDS_HSPC_deconvolution_melt<-MDS_HSPC_deconvolution_iBMSC_score%>%select(name,Group,MEPs,LMPPs,CLPs,GMPs,HSCs,iBMSC_score)%>%melt(id.vars=c("name","Group","iBMSC_score"),variable.name="HSPC_type",value="fraction")
MDS_HSPC_deconvolution_melt$HSPC_type<-factor(MDS_HSPC_deconvolution_melt$HSPC_type,levels=c("HSCs","CLPs","GMPs","LMPPs","MEPs"))
MDS_HSPC_deconvolution_melt$name<-factor(MDS_HSPC_deconvolution_melt$name,levels=MDS_HSPC_deconvolution_iBMSC_score$name)
MDS_HSPC_deconvolution_melt%>%ggplot(aes(name,value,fill=HSPC_type))+geom_col()+scale_fill_manual(values=rev(c("tomato2","steelblue4","steelblue2","yellow4","grey60")))


MDS_HSPC_deconvolution_melt$HSPC_type<-factor(MDS_HSPC_deconvolution_melt$HSPC_type,levels=c("HSCs","CLPs","MEPs","LMPPs","GMPs"))
MDS_HSPC_deconvolution_melt$name<-factor(MDS_HSPC_deconvolution_melt$name,levels=MDS_HSPC_deconvolution_iBMSC_score$name)
MDS_HSPC_deconvolution_melt%>%ggplot(aes(name,value,fill=HSPC_type))+geom_col()+scale_fill_manual(values=rev(c("steelblue4","steelblue2","tomato2","yellow4","grey60")))+
  theme_bw()



############
#Extra Mutation analysis splicing and RaCHseq
############

MDS_NBM_all_mt<-readRDS("~/single_cell_sequencing/MDS single cell//MDS_NBM_all_clustify_mutant_WT.RDS")

MT_prediction<-read.table("~/single_cell_sequencing/MDS single cell/MRLC_classification.tsv",header=T)

MT_prediction$Sample<-gsub("_HM","",MT_prediction$sample)
MT_prediction$Sample<-gsub("_NI","",MT_prediction$Sample)

MT_prediction["type"]<-ifelse(grepl("NI",MT_prediction$sample),"niche_immune","")
MT_prediction["type"]<-ifelse(grepl("HM",MT_prediction$sample),"cd34_myeloid",MT_prediction$type)




MT_prediction["bar_code"]<-tolower(paste0(MT_prediction$Sample,"_",MT_prediction$type,"_",MT_prediction$barcode,"_1"))
MT_prediction["bar_code"]<-gsub("__","_",MT_prediction$bar_code)

Mutant_cells<-MT_prediction[MT_prediction$Prediction,]

MT_prediction<-MT_prediction[c("Prediction","bar_code")]
metadata_MDS<-MDS_NBM_all_mt@meta.data

metadata_MDS<-metadata_MDS%>%left_join(MT_prediction)
row.names(metadata_MDS)<-metadata_MDS$bar_code
metadata_MDS<-metadata_MDS[MDS_NBM_all_mt$bar_code,]
MDS_NBM_all_mt[["SF3B1_prediction"]]<-metadata_MDS$Prediction
MDS_NBM_all_mt[["SF3B1_prediction"]]<-ifelse(MDS_NBM_all_mt$SF3B1_prediction,"MT","Unknown")
MDS_NBM_all_mt[["SF3B1_prediction"]]<-ifelse(grepl("NBM",MDS_NBM_all_mt$Group),"NBM",MDS_NBM_all_mt$SF3B1_prediction)


DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"),split.by = "Group",label = T,repel=T,cells.highlight = names(MDS_NBM_all_mt$prediction)[MDS_NBM_all_mt$SF3B1_prediction=="MT"], sizes.highlight = 0.3,pt.size=0.1,raster=F)
DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"),split.by = "Group",label = T,cells.highlight = names(MDS_NBM_all_mt$prediction)[MDS_NBM_all_mt$Mutant_state=="Mutant"], cols.highlight = "tomato4",sizes.highlight = 0.2,pt.size=0.3,raster=F)

summary(factor(MDS_NBM_all_mt$SF3B1_prediction))
summary(factor(MDS_NBM_all_mt$Mutant_state))


DimPlot(MDS_NBM_all_mt,split.by = "Group")





MDS_NBM_HSPC<-readRDS("single_cell_sequencing/MDS single cell/MDS_NBM_HSPC.RDS")

MDS_NBM_HSPC[["SF3B1_prediction"]]<-ifelse(MDS_NBM_HSPC$bar_code%in%Mutant_cells$bar_code, "SF3B1m","Unknown")
DimPlot(subset(MDS_NBM_HSPC,subset=Group=="MDS"),label = T,cells.highlight = names(MDS_NBM_HSPC$SF3B1_prediction)[MDS_NBM_HSPC$SF3B1_prediction=="SF3B1m"], cols.highlight = "tomato4",sizes.highlight = 0.4,pt.size=0.2,raster=F)

Plot<-DimPlot(subset(MDS_NBM_HSPC,subset=Group=="MDS"),label = T,cells.highlight = names(MDS_NBM_HSPC$SF3B1_prediction)[MDS_NBM_HSPC$SF3B1_prediction=="SF3B1m"], cols.highlight = "tomato4",sizes.highlight = 0.4,pt.size=0.2,raster=F)

pdf("LRMDS_HSPC_mt.pdf", width=5, height=4)
print(Plot)
dev.off()

#RaCHseq
###########################
RaCH_list<-list.files("scRNAseq data transfer Lanpeng Chen/LRMDS/Mutant cell barcodes/RaCHseq",full.names = T)
RaCH_name<-list.files("scRNAseq data transfer Lanpeng Chen/LRMDS/Mutant cell barcodes/RaCHseq")
names(RaCH_list)<-RaCH_name

RaCH_list<-lapply(RaCH_list, function(x){
  x=data.frame(read.table(x))})

for (i in 1:length(RaCH_list)){
  RaCH_list[[i]]["bar_code"]<- tolower(paste0(sub("_pos.txt", replacement="",RaCH_name)[i],"_",RaCH_list[[i]]$V1,"_1"))
  RaCH_list[[i]]<-RaCH_list[[i]]["bar_code"]
}


RaCH_df<-unlist(RaCH_list)

MDS_NBM_all_mt[["RaCH"]]<-"Unknown"


MDS_NBM_all_mt$RaCH<-ifelse(MDS_NBM_all_mt$bar_code %in% RaCH_df, "SF3B1m", "Unknown")


DimPlot(subset(MDS_NBM_all_mt,subset=Group=="NBM"),
        cells.highlight = names(MDS_NBM_all_mt$RaCH[MDS_NBM_all_mt$RaCH=="SF3B1m"]),raster = F)

DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"), label=T,split.by = "Group",reduction = "umap", cells.highlight = names(MDS_NBM_all_mt$RaCH[MDS_NBM_all_mt$RaCH=="SF3B1m"]), cols.highlight = "tomato4",repel = T)

MDS_NBM_HSPC[["RaCH"]]<-ifelse(MDS_NBM_HSPC$bar_code %in% RaCH_df, "SF3B1m", "Unknown")

DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"), label=T,split.by = "Group",reduction = "umap", cells.highlight = names(MDS_NBM_all_mt$Mutation[MDS_NBM_all_mt$Mutation=="Mutant"]), cols.highlight = "tomato4",repel = T)
DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"), label=T,split.by = "Group",reduction = "umap", cells.highlight = names(MDS_NBM_all_mt$RaCH[MDS_NBM_all_mt$RaCH=="SF3B1m"]), cols.highlight = "tomato4",repel = T)
DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"), label=T,split.by = "Group",reduction = "umap", cells.highlight = names(MDS_NBM_all_mt$SF3B1_prediction[MDS_NBM_all_mt$SF3B1_prediction=="MT"]), cols.highlight = "tomato4",repel = T)
DimPlot(subset(MDS_NBM_all_mt,subset=Group=="MDS"), label=T,split.by = "Group",reduction = "umap", cells.highlight = names(MDS_NBM_all_mt$MT_comb[MDS_NBM_all_mt$MT_comb=="MT"]), cols.highlight = "tomato4",repel = T)

#Mutation call
 
#RaCHseq
MT_ratio<-subset(MDS_NBM_all_mt,subset=Group=="MDS")@meta.data%>%group_by(RaCH,type_re)%>%summarize(n=n())%>%group_by(type_re)%>%
  summarize(Mutation=RaCH,perc=n/sum(n)*100)

MT_ratio$type_re<-factor(MT_ratio$type_re,levels=c("BMSCs", "ECs", "HSC/MPPs","LMPPs","CLPs","MEPs","GMPs","Monoblasts",
                                                   "CD14+ Mono","CD16+ Mono", "DCs","Erythroid progenitors","Mks",
                                                   "CD4","CD8", "NK", "B", "Plasma cells"))
data.frame(MT_ratio)%>%complete(type_re,Mutation,fill = list(perc = 0))%>%filter(Mutation=="SF3B1m")%>%ggplot(aes(perc+0.01,type_re))+geom_col(fill="tomato4")+theme_bw()+labs(y="",x="Size")

#Splicosome
MDS_NBM_all_mt$SF3B1_prediction[is.na(MDS_NBM_all_mt$SF3B1_prediction)]<-"Unknown"


MT_ratio<-subset(MDS_NBM_all_mt,subset=Group=="MDS")@meta.data%>%group_by(SF3B1_prediction,type_re)%>%summarize(n=n())%>%group_by(type_re)%>%
  summarize(Mutation=SF3B1_prediction,perc=n/sum(n)*100)

MT_ratio$type_re<-factor(MT_ratio$type_re,levels=c("BMSCs", "ECs", "HSC/MPPs","LMPPs","CLPs","MEPs","GMPs","Monoblasts",
                                                   "CD14+ Mono","CD16+ Mono", "DCs","Erythroid progenitors","Mks",
                                                   "CD4","CD8", "NK", "B", "Plasma cells"))
data.frame(MT_ratio)%>%complete(type_re,Mutation,fill = list(perc = 0))%>%filter(Mutation=="MT")%>%ggplot(aes(perc+0.01,type_re))+geom_col(fill="tomato4")+theme_bw()+labs(y="",x="Size")

#Combine
MDS_NBM_all_mt[["MT_comb"]]<-ifelse(MDS_NBM_all_mt$SF3B1_prediction=="MT"|MDS_NBM_all_mt$RaCH=="SF3B1m"|MDS_NBM_all_mt$Mutation=="Mutant","MT","Unknown")

MT_ratio<-subset(MDS_NBM_all_mt,subset=Group=="MDS")@meta.data%>%group_by(MT_comb,type_re)%>%summarize(n=n())%>%group_by(type_re)%>%
  summarize(Mutation=MT_comb,perc=n/sum(n)*100)

MT_ratio$type_re<-factor(MT_ratio$type_re,levels=c("BMSCs", "ECs", "HSC/MPPs","LMPPs","CLPs","MEPs","GMPs","Monoblasts",
                                                   "CD14+ Mono","CD16+ Mono", "DCs","Erythroid progenitors","Mks",
                                                   "CD4","CD8", "NK", "B", "Plasma cells"))
data.frame(MT_ratio)%>%complete(type_re,Mutation,fill = list(perc = 0))%>%filter(Mutation=="MT")%>%ggplot(aes(perc+0.01,type_re))+geom_col(fill="tomato4")+theme_bw()+labs(y="",x="Size")


sum(Cells(MDS_NBM_all_mt)[MDS_NBM_all_mt$Mutation=="Mutant"] %in% Cells(MDS_NBM_all_mt)[MDS_NBM_all_mt$SF3B1_prediction=="MT"])

sum(Cells(MDS_NBM_all_mt)[MDS_NBM_all_mt$Mutation=="Mutant"] %in% Cells(MDS_NBM_all_mt)[MDS_NBM_all_mt$RaCH=="SF3B1m"])
sum(Cells(MDS_NBM_all_mt)[MDS_NBM_all_mt$SF3B1_prediction=="MT"] %in% Cells(MDS_NBM_all_mt)[MDS_NBM_all_mt$RaCH=="SF3B1m"])

sum(ifelse(MDS_NBM_all_mt$SF3B1_prediction=="MT"&MDS_NBM_all_mt$RaCH=="SF3B1m"&MDS_NBM_all_mt$Mutation=="Mutant",1,0))
sum(ifelse(MDS_NBM_all_mt$SF3B1_prediction=="MT"&MDS_NBM_all_mt$RaCH!="SF3B1m"&MDS_NBM_all_mt$Mutation!="Mutant",1,0))
sum(ifelse(MDS_NBM_all_mt$SF3B1_prediction!="MT"&MDS_NBM_all_mt$RaCH=="SF3B1m"&MDS_NBM_all_mt$Mutation!="Mutant",1,0))

sum(ifelse(MDS_NBM_all_mt$Mutation=="Mutant" & MDS_NBM_all_mt$RaCH!="SF3B1m" & MDS_NBM_all_mt$SF3B1_prediction!="MT",1,0))
sum(ifelse(MDS_NBM_all_mt$Mutation=="Mutant" & MDS_NBM_all_mt$RaCH!="SF3B1m" & MDS_NBM_all_mt$SF3B1_prediction!="MT",1,0))


#MT analysis on HSPCs


DimPlot(MDS_NBM_HSPC, reduction = "umap", label=T,pt.size = 0.25,split.by = "Group")

FeaturePlot(MDS_NBM_HSPC,split.by = "Group",features="CD34",min.cutoff = 0.25,order=T,pt.size=0.1)
DotPlot(MDS_NBM_HSPC,features=c("MAP3K8"),group.by = "RaCH",assay="RNA")

VlnPlot(MDS_NBM_HSPC,features=c("MAP3K8"),group.by = "SF3B1_prediction",assay="RNA",pt.size=0)
VlnPlot(MDS_NBM_HSPC,features=c("MAP3K8"),group.by = "RaCH",assay="RNA",pt.size=0)

VlnPlot(MDS_NBM_HSPC,features=c("BTG1"),group.by = "RaCH",assay="RNA",pt.size=0)



summary(factor(MDS_NBM_HSPC$SF3B1_prediction))

HSPC_MDS<-subset(MDS_NBM_HSPC,subset=Group=="MDS")
HSPC_MDS<-subset(MDS_NBM_HSPC,subset=Sample=="MDS_337",invert=T)

DimPlot(HSPC_MDS, reduction = "umap", label=T,pt.size = 0.25,split.by = "Sample")

HSPC_MDS[["MT_comb"]]<-ifelse(HSPC_MDS$SF3B1_prediction=="SF3B1m"|HSPC_MDS$RaCH=="SF3B1m"|HSPC_MDS$Mutation=="Mutant","SF3B1m","Unknown")

summary(factor(HSPC_MDS$RaCH))
summary(factor(HSPC_MDS$SF3B1_prediction))
summary(factor(HSPC_MDS$Mutation))
summary(factor(HSPC_MDS$MT_comb))

DimPlot(subset(HSPC_MDS,subset=Group=="MDS"),label = T,cells.highlight = names(HSPC_MDS$MT_comb)[HSPC_MDS$MT_comb=="SF3B1m"], cols.highlight = "tomato4",sizes.highlight = 0.4,pt.size=0.2,raster=F)


DimPlot(HSPC_MDS, reduction = "umap", label=T,pt.size = 0.25,split.by = "Group")


MDS_HSPC_deseq<-run_pseudobulkDE_new(data=HSPC_MDS,
                                 contrasts='condition,SF3B1m,Unknown',design='~condition+Extra',
                                 sample_col="Sample", condition_col="MT_comb",cluster_col="type_re",extr_col='Sample',do_prefiltering = TRUE)



MDS_HSPC_deseq["Gene"]<-row.names(MDS_HSPC_deseq)
MDS_HSPC_deseq["rank"]<- -sign(MDS_HSPC_deseq$log2FoldChange)*log2(MDS_HSPC_deseq$pvalue)
MDS_HSPC_deseq<-MDS_HSPC_deseq[order(MDS_HSPC_deseq$padj),]
MDS_HSPC_deseq_sig<-MDS_HSPC_deseq[MDS_HSPC_deseq$padj<0.05,]


MDS_HSPC_deseq["rank"]<- -sign(MDS_HSPC_deseq$log2FoldChange)*log2(MDS_HSPC_deseq$pvalue)
MDS_HSPC_deseq<-MDS_HSPC_deseq[order(MDS_HSPC_deseq$padj),]
MDS_HSPC_deseq_sig<-MDS_HSPC_deseq[MDS_HSPC_deseq$padj<0.05,]

write_xlsx(MDS_HSPC_deseq,"MDS_HSPCs_MT_vs_NA_Predict.xlsx")

MDS_HSPC_deseq%>%ggplot(aes(-log10(padj),log2FoldChange))+
  geom_point(alpha=0.2,color="grey80")+geom_vline(xintercept= -log10(0.05),linetype=3,cols="grey")+
  geom_point(data=MDS_HSPC_deseq_sig,mapping=aes(-log10(padj),log2FoldChange),alpha=0.2)+
  geom_text_repel(data=MDS_HSPC_deseq_sig,mapping=aes(-log10(padj),log2FoldChange,label=Gene),size=3,max.overlaps=30)+
  theme_bw()

pdf("MDS Figure/Figure 4/Vlnplot_mt_vs_unknown.pdf",width = 5, height =5)

MDS_HSPC_deseq%>%ggplot(aes(-log10(padj),log2FoldChange))+
  geom_point(alpha=0.2,color="grey80")+geom_vline(xintercept= -log10(0.05),linetype=3,cols="grey")+
  geom_point(data=MDS_HSPC_deseq_sig,mapping=aes(-log10(padj),log2FoldChange),alpha=0.2)+
  geom_text_repel(data=MDS_HSPC_deseq_sig,mapping=aes(-log10(padj),log2FoldChange,label=Gene),size=3,max.overlaps=30)+
  theme_bw()

dev.off()

#write_xlsx(MDS_HSPC_deseq,"single_cell_sequencing/MDS single cell/HSPC_MT_vs_NA.xlsx")

rank_MDS_HSPC<-MDS_HSPC_deseq$rank
names(rank_MDS_HSPC)<-MDS_HSPC_deseq$Gene
rank_MDS_HSPC<-rank_MDS_HSPC[is.finite(rank_MDS_HSPC)]
MDS_HSPC_Hall <- fgsea(hallmakers_set, rank_MDS_HSPC, minSize=15, maxSize = 500, nperm=1000)

head(MDS_HSPC_Hall[order(padj), ], n=10)
MDS_HSPC_Hall_sig<-MDS_HSPC_Hall%>%filter(padj<0.05)%>%arrange(NES)
MDS_HSPC_Hall_sig[["Enrichment"]]<-ifelse(MDS_HSPC_Hall_sig$NES<0,"Mutant","NA")
MDS_HSPC_Hall_sig[1]$leadingEdge
MDS_HSPC_Hall_sig[9]$leadingEdge
MDS_HSPC_Hall_sig%>%group_by(Enrichment)%>%top_n(n=10,wt=abs(NES))%>%
  mutate(pathway=factor(pathway, levels=MDS_HSPC_Hall_sig$pathway))%>%ggplot(aes(NES,pathway,fill= padj))+geom_col()+scale_fill_gradient(low="red",high="blue")+theme_classic()

pdf("MDS Figure/Figure 4/HM_GSEA_MT_vs_Unknown.pdf",width = 6, height =5)

MDS_HSPC_Hall_sig%>%group_by(Enrichment)%>%top_n(n=10,wt=abs(NES))%>%
  mutate(pathway=factor(pathway, levels=MDS_HSPC_Hall_sig$pathway))%>%ggplot(aes(NES,pathway,fill= padj))+geom_col()+scale_fill_gradient(low="red",high="blue")+theme_classic()


dev.off()

NBM_Mutant_HSPC_deseq["Gene"]<-row.names(NBM_Mutant_HSPC_deseq)



DefaultAssay(HSPC_NBM_MDS)<-"RNA"

MDS_NBM_HSPC_score<-AddModuleScore(HSPC_NBM_MDS,features=hallmakers_set[1],name="NFkB_via_TNF")
MDS_NBM_HSPC_score<-AddModuleScore(MDS_NBM_HSPC_score,features=hallmakers_set[10],name="Apoptosis")
MDS_NBM_HSPC_score<-AddModuleScore(MDS_NBM_HSPC_score,features=hallmakers_set[37],name="TP53_signaling")
MDS_NBM_HSPC_score<-AddModuleScore(MDS_NBM_HSPC_score,features=hallmakers_set["HALLMARK_INTERFERON_GAMMA_RESPONSE"],name="IFN_response")
MDS_NBM_HSPC_score<-AddModuleScore(MDS_NBM_HSPC_score,features=hallmakers_set["HALLMARK_E2F_TARGETS"],name="E2F_targets")

VlnPlot(MDS_NBM_HSPC_score,features=c("NFkB_via_TNF1","Apoptosis1","IFN_response1","E2F_targets1"),ncol=2,split.by = "Sample", pt.size=0)&theme(legend.position="bottom")

geneset="NFkB_via_TNF1"
data=MDS_NBM_HSPC_score
Heatmap_geneset<-function(data=MDS_NBM_HSPC_score, geneset="NFkB_via_TNF1"){
  data<-data@meta.data[,c("Group","Sample",geneset,"type_re")]
  names(data)<-c("Group","Sample","geneset","type_re")
  data<-data%>%group_by(Sample,type_re)%>%summarise(meanScore=mean(geneset))
  data["Group"]<-ifelse(grepl("MDS",data$Sample),"MDS","NBM")
  data$Sample<-factor(data$Sample,levels=c("NBM1","NBM2","NBM3","NBM4","MDS_10209","MDS_124","MDS_337","MDS_345","MDS_SF3B1"))
  data<-data%>%group_by(type_re)%>%summarise(Sample=Sample,ScaledScore=scale(meanScore))
  data<-data%>%filter(!type_re%in%c("MDS_Progs","CD16+ Mono"))
  data%>%ggplot(aes(Sample,type_re,fill=ScaledScore))+geom_raster()+
    scale_fill_gradient(low = "blue", high = "red")+theme_classic()+labs(title=geneset)+
    theme(axis.line = element_blank(), axis.text.x=element_text(angle=45, hjust=1,vjust=1))
}
Heatmap_geneset(data=MDS_NBM_HSPC_score, geneset="NFkB_via_TNF1")+theme(axis.title=element_blank())+NoLegend()+
  Heatmap_geneset(data=MDS_NBM_HSPC_score, geneset="Apoptosis1")+theme(axis.title=element_blank())+NoLegend()+
  Heatmap_geneset(data=MDS_NBM_HSPC_score, geneset="IFN_response1")+
  Heatmap_geneset(data=MDS_NBM_HSPC_score, geneset="E2F_targets1")+theme(axis.title=element_blank())+NoLegend()
  
#########################
#MEPs

MDS_MEP_deseq<-run_pseudobulkDE_new(data=subset(HSPC_MDS,ident="MEPs"),
                                     contrasts='condition,SF3B1m,Unknown',design='~condition+Extra',
                                     sample_col="Sample", condition_col="RaCH",cluster_col="type_re",extr_col='Sample',do_prefiltering = TRUE)

MDS_MEP_deseq<-run_pseudobulkDE_new(data=subset(HSPC_MDS,ident="MEPs"),
                                    contrasts='condition,MT,Unknown',design='~condition+Extra',
                                    sample_col="Sample", condition_col="SF3B1_prediction",cluster_col="type_re",extr_col='Sample',do_prefiltering = TRUE)


MDS_MEP_deseq["Gene"]<-row.names(MDS_MEP_deseq)
MDS_MEP_deseq["rank"]<- -sign(
  MDS_MEP_deseq$log2FoldChange)*log2(MDS_MEP_deseq$pvalue)
MDS_MEP_deseq<-MDS_MEP_deseq[order(MDS_MEP_deseq$padj),]
MDS_MEP_deseq_sig<-MDS_MEP_deseq[MDS_MEP_deseq$padj<0.05,]


MDS_MEP_deseq["rank"]<- -sign(
  MDS_MEP_deseq$log2FoldChange)*log2(MDS_MEP_deseq$pvalue)
MDS_MEP_deseq<-MDS_MEP_deseq[order(MDS_MEP_deseq$padj),]
MDS_MEP_deseq_sig<-MDS_MEP_deseq[MDS_MEP_deseq$padj<0.05,]

write_xlsx(MDS_MEP_deseq,"MDS_MEP_MT_vs_NA_Predict.xlsx")

MDS_MEP_deseq%>%ggplot(aes(-log10(padj),log2FoldChange))+
  geom_point(alpha=0.2,color="grey80")+geom_vline(xintercept= -log10(0.05),linetype=3,cols="grey")+
  geom_point(data=MDS_MEP_deseq_sig,mapping=aes(-log10(padj),log2FoldChange),alpha=0.2)+
  geom_text_repel(data=MDS_MEP_deseq_sig,mapping=aes(-log10(padj),log2FoldChange,label=Gene),size=3,max.overlaps=30)+
  theme_bw()



#write_xlsx(MDS_MEP_deseq,"single_cell_sequencing/MDS single cell/MEP_MT_vs_NA.xlsx")

rank_MDS_MEP<-MDS_MEP_deseq$rank
names(rank_MDS_MEP)<-MDS_MEP_deseq$Gene
rank_MDS_MEP<-rank_MDS_MEP[is.finite(rank_MDS_MEP)]
MDS_MEP_Hall <- fgsea(hallmakers_set, rank_MDS_MEP, minSize=15, maxSize = 500, nperm=1000)

head(MDS_MEP_Hall[order(padj), ], n=10)
MDS_MEP_Hall_sig<-MDS_MEP_Hall%>%filter(padj<0.05)%>%arrange(NES)
MDS_MEP_Hall_sig[["Enrichment"]]<-ifelse(MDS_MEP_Hall_sig$NES<0,"Mutant","NA")
MDS_MEP_Hall_sig[1]$leadingEdge
MDS_MEP_Hall_sig[9]$leadingEdge
MDS_MEP_Hall_sig%>%group_by(Enrichment)%>%top_n(n=10,wt=abs(NES))%>%
  mutate(pathway=factor(pathway, levels=MDS_MEP_Hall_sig$pathway))%>%ggplot(aes(NES,pathway,fill= padj))+geom_col()+scale_fill_gradient(low="red",high="blue")+theme_classic()



NBM_Mutant_MEP_deseq["Gene"]<-row.names(NBM_Mutant_MEP_deseq)
########################
